<!DOCTYPE html>
<html>
<!--#set var=TITLE value="Tour: runSingleTCW" -->
<!--#include virtual="../../ssi/head.html" -->
<body onLoad="init('runS')">
<a id=top></a>
<!--#include virtual="../ssi/start2t.html" -->

<div style="text-align: center"><big><ttp>runSingleTCW</ttp></big> - Build Annotated Database of Single Species</div>
<table  class="tablex" style="margin-left: auto;margin-right: auto;">
<tr><td class="white" style="text-align:left; background-color:#666666; padding-bottom:4px;">
<a href="#main" 	class="white">Main Window</a> 		<mark class="white">|</mark> 
<a href="#add" 		class="white">Additional Windows</a> <mark class="white">|</mark> 
<a href="#refs" 	class="white" >References</a>
</td></tr>
</table>

<p><ttp>runSingleTCW</ttp> takes as input transcriptome and expression level files, builds a 
database, and annotates the transcriptomes. 
See <a href="../../stcw/UserGuide.html" class="ext" target="_blank">sTCW UserGuide</a> for details. 

<a id=main></a>
<h3>Main window</h3>

<table class="tablex">
<tr>
	<td>
	1. <ttc>Build Database</ttc>
	<ol type="a">
		<li>The input may be transcripts with optional read counts, proteins with optional spectral counts, 
		or sequences to be assembled such as  Sanger ESTs, 454 reads, and/or transcript libraries. 
		<li>The <ttc>Add</ttc> window allows you to specify the sequence dataset (fasta file) 
		along with its conditions (e.g. tissue, treatment, etc),
		which will be listed under <ttc>Associated Counts</ttc> on the main window. 
		<li>If there are replicates, you may define them with the <ttc>Define Replicates</ttc> window.
	</ol>

	2. <ttc>Instantiate</ttc>
	<ol type="a">
		<li><ttc>Skip Assembly</ttc> - the sequences can be instantiated with no assembly.
		<li>Assembly - the sequences can be assembled. If Read sequences (i.e. ESTs) are mixed with
		transcripts that have read counts, the EST dataset 'counts' for the contig will be the number
		of ESTs in the given contig, and the transcripts will retain their counts from the input.
	</ol>

	3. <ttc>Annotate</ttc> 
	<ol type="a">
		<li>Use <a href="../runAS" class="ext" target="_blank">runAS</a> to download and format UniProts and the GO databases; these are input to (b) and (c).
		<li>Add one or more databases to search against, which can be protein or nucleotide.
		<li>In the <ttc>Options</ttc> window, define the GO database 
	</ol>

	4. <ttc>Add Remarks and Locations</ttc>
	<p> Remarks and Locations (i.e. chromosome, start, end, strand) can be added to sequences and queried in the viewSingleTCW.
	</td>
	<td>
		<a id="Fig1"></a>
		<a  href="img/runSingle.png">
		<img src="img/runSingle.png" alt="" style="border: 1px solid black; width: 500px"></a>
	</td>
	</tr>
</table>

<!--     Add          -->
<a id=add></a>
<h3>Additional windows</h3>

<table class="tablex">
<tr>
	<td class=top>
		<ttl>Sequence Datasets</ttl>
		<p><ttc>Add/Edit</ttc>:  Selecting <ttc>Add</ttc> shows this window. Selecting <ttc>Edit</ttc> shows
		a similar window, but only the Attribute values can be changed after the database has been created. 
		The counts can be in one file or generated with the <ttc>Build combined count file</ttc> option. 
		On <ttc>Save</ttc>, the condition names will be written in the <ttc>Associated Counts</ttc> table on the 
		main window.
		<p><ttc>Build combined count file file</ttc>:  (window not shown) If the counts are in multiple files, this button opens up a 
		window in which a directory of count files, or individual count files, can be entered. 
		It will generate a file called <tt>Combined_read_count.csv</tt> 
		where the columns are conditions with their respective counts.
	</td>
	<td>
		<a  href="img/runSingleLib.png"> <img src="img/runSingleLib.png" alt="" style="border: 1px solid black; width: 450px"></a>
	</td>
<tr>
	<td class="top">
		<ttc>Associated Counts</ttc>
		<p><ttc>Define Replicates</ttc>: If the <ttc>Associated Counts</ttc> table has replicates, 
		they can be defined in this window, which will update the table as shown above in the Main window.
		<p><ttc>Edit Attributes</ttc>: (window not shown) This window has the same attributes fields as the
		<ttc>Add or Edit a Sequence Dataset</ttc>, but will be applied to the selected condition. These can
		be edited after the database has been created.
	</td>
	<td>
		<a href="img/runSingleRep.png"><img src="img/runSingleRep.png" alt="" style="border: 1px solid black; width: 450px"></a>

<tr><td>&nbsp;

<tr>
	<td class="top">
	<ttl>AnnoDBs</ttl>
	<p><ttc>Add/Edit</ttc>: Databases to search against are referred to as "annoDBs". 
	Any FASTA file can be used an annoDB, though TCW gives special support to using UniProt taxonomic
	databases; besides allowing taxonomic specific querying in viewSingleTCW, 
	the UniProt .dat file is used to extract GO, KEGG, EC and PFam information. The <a href="../runAS" class="ext" target="_blank"><ttp>runAS</ttp></a>
	program provides this functionality.
	<p>
	As the UniProt TrEMBL databases increase in size, it has
	become impractical to search these with BLAST<sup>1</sup>. Fortunately, the 
	<a href="http://ab.inf.uni-tuebingen.de/software/diamond" class="ext" target="_blank">diamond</a><sup>2</sup>
	programs provide super fast results.
	</td>
	<td>
	<a  href="img/runSingleAnno.png"><img src="img/runSingleAnno.png" alt="" style="border: 1px solid black; width: 450px"></a>
<td>
</table>

<!-- REFS -->
<a id=refs></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h3>References</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<ol>
<li><ttp>BLAST</ttp> is used for assembly, annotation and interactive searching in viewSingleTCW. 
<br>Altschul SF, Madden TL, Schaffer AA, Zhang J, Zhang Z, et al. (1997) Gapped BLAST and PSI-BLAST: a new generation of protein database search programs. Nucleic Acids Res 25: 3389-3402.
<li><ttp>Diamond</ttp> can be used for annotation of blastx and blastp searches.
<br>Buchfink B, Xie C, Huson D (2015) Fast and Sensitive Protein Alignment 
using DIAMOND, Nature Methods, 12, 59-60 doi:10.1038/nmeth.3176.

<li><ttp>CAP3</ttp> is used for assembly. 
<br>Huang X, Madan A (1999) CAP3: A DNA sequence assembly program. Genome Res 9: 868-877.
<li><ttp>UniProt</ttp> is recommended for protein annotation as the GO and other information can 
be extracted by the TCW and added to the database. 
<br>Dimmer EC, Huntley RP, Alam-Faruque Y, Sawford T, O'Donovan C, et al. (2012) The UniProt-GO Annotation database in 2011. Nucleic Acids Res 40: D565-570.
<li><ttp>Gene Ontology</ttp> mySQL database is used for levels and descriptions. 
<br>GO Consortium (2012) The Gene Ontology: enhancements for 2011. Nucleic Acids Res 40: D559-564.
</ol>
	
<!--#include virtual="../../ssi/end.html" -->
</body>
</html>

