<!DOCTYPE HTML>
<html>
<head>
<title>GO Help</title>
<style>
.tablex    		{border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
.tablex caption {padding-bottom: 0.5em; font-family: Helvetica, verdana;}
.tablex th 		{padding: 3px; }
.tablex td 		{padding: 3px;}
.tabley    		{border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
.tabley caption {padding-bottom: 0.5em;  font-family: Helvetica, verdana;}
.tabley th 		{border: 1px  solid black; padding: 3px; }
.tabley td 		{border: 1px  solid black; padding: 3px; }
</style>
</head>
<body style="padding:20px;width:700px;font-family:Times">
<h1>GO Help</h1>
<p>The version of GO that was used for the TCW annotation is shown at the bottom
of the sTCWdb <b>Overview</b>.
Unless TCW was built with the latest GO database, there will be 
differences from what is found on amigo.geneontology.org as the
gene ontology is constantly being updated. The following example is dated 4-Mar-2021,
and screen shots are from:</p>
<table>
<tr><td><a href="http://amigo.geneontology.org/amigo/term/GO:0000166">Amigo</a></td>
<td>http://amigo.geneontology.org/amigo/term/GO:0000166</td>
<tr><td><a href="https://www.ebi.ac.uk/QuickGO/GTerm?class=GO:0000166">QuickGO</a></td>
<td>https://www.ebi.ac.uk/QuickGO/GTerm?class=GO:0000166</td>
</table>

<p>Proteins in the UniProt database have been <i>assigned</i> GO terms. 
TCW has the following mappings: 
<ul>
<li>Protein names to sequences (from a heuristic search program such as BLAST or DIAMOND), 
<li>GO terms to proteins (from UniProt), 
<li>GO terms to sequences (the GOs assigned to the proteins that hit the sequence). 
</ul>
When a protein is assigned a GO term, the ancestors of the GO term 
inherits the hit and sequences of the hit; this is because ancestors are
increasingly more general terms and all apply. Hence, the GO terms in TCW
are all the <i>assigned (direct)</i> and <i>inherited (indirect)</i> GO terms.  

<p>For example,  <tt>nucleotide binding</tt> (GO:0000166) 'is_a' <tt>nucleoside phosphate binding</tt>
(GO:190165) and 'is_a' <tt>small molecule binding</tt> (GO:0036094). This is often referred to s
GO:0000166 being the child of GO:190165 and GO:190165 being the parent of GO:0000166;
likewise for GO:0036094. GO:0000166 inherits four other GOs as shown in the figure below.

<p><table class="tablex">
<caption>Figure 2: Ancestor Paths and Graph for Nucleotide binding (GO:0000166)</caption>
<tr><th class=left>A. QuickGO Graph <th class=left>&nbsp;&nbsp;&nbsp;&nbsp;B. TCW GO Ancestor Path</tr>
<tr>
<td><img src=goPathsQuick.png alt="">
<td valign=top><img src=goPathsTCW.png alt="">
</tr>
</table>
In Figure 2A, the paths can be followed by following the arrows up the tree.
In Figure 2B, the numbers on the right of the paths table correspond to 'levels', 
however, a term may be in multiple levels, e.g. <tt>nucleotide binding</tt> 
is in level 4 for <u>Path3</u> but level 5 for <u>Path2</u>; it is only shown in 
its lowest level.

<p>To quote <a href="http://geneontology.org/docs/faq" class=ext target="_blank">http://geneontology.org/doc/faq</a>, 
<quote>
"GO terms do not occupy strict fixed levels in the hierarchy. Because GO is structured as a graph, terms would appear at different ‘levels’ if different paths were followed through the graph. 
This is especially true if one mixes the different relations used to connect terms."
</qoute>
Nevertheless, TCW computes levels where it assigns a GO its lowest level. 
<p><table class="tablex">
<caption>Figure 2: Neighborhood for Nucleotide binding (GO:0000166)</caption>
<tr><th class=left>A. Amigo Neighbors (Partial Children List) <th class=left>&nbsp;&nbsp;&nbsp;&nbsp;B. TCW Neighbors</tr>
<tr>
<td><img src=goNeighAmigoPartial.png alt="" border=1 width=600>
<td valign=top><img src=goNeighTCW.png alt=""  border=1 width=400>
</tr>
</table>

<p>
<b>NOTE:</b> In TCW, only GOs that have an assigned or inherited hit will be in the database,
hence, a GOs descendants will not necessarily all be in the TCW database (e.g. GO:0016502 is a child, but is not
list for the TCW display since it is not in the database). However, all ancestors will be in the database since
they are inherited by at least this GO.
<p>
The relation types used in TCW are 'is_a' and 'part_of'.

<h3>Graph terms</h3>
In computer science, a graph has <i>nodes</i> and <i>edges</i>. There are 
<i>paths</i> between nodes by following the edges. 
The graph is <i>directed</i> if the edges have arrows, where a <i>child</i> node points to the <i>parent</i> node.
In the gene ontology graphs, the GO terms are
represented by nodes; an edge from a child node to a parent node indicates
its a specialized term of the parent node.  
The gene ontology has three root nodes, <tt>biological process</tt>, <tt>cellular_component</tt>,
and <tt>molecular_function</tt>. GO is not strictly hierarchical as a node can have more than
one parent. See <a href="http://geneontology.org/page/ontology-structure" class=ext target="_blank">
http://geneontology.org/page/ontology-structure</a>  for a more in-depth discussion.
<p>TCW does not uses edges that cross the three trees. 

<h3>E-values</h3>
A GO term may be assigned or inherited by many protein hits in the TCW database. Each protein hit
has an e-value indicating how well it aligned to the sequence. The GO-protein hit is assigned the
best e-value over all the hit-sequence pairs. 

<h3>Examples</h3>
Say protein X (proX) aligns to sequence Y (seqY) with an e-value, and GO:0000166 is assigned to proX with an evidence code EC, 
then seqY is also said to be annotated with GO:0000166. 
<ul>
<li>Since proX is assigned GO:0000166, it also inherits all GO:0000166 ancestors.
<li>Since seqY has annotation of GO:0000166, it also inherits all GO:0000166 ancestors.
<li>Since GO:0000166 is assigned to proX, all its ancestors inherit proX.
<li>Since GO:0000166 annotates seqY, all its ancestors also annotate seqY.
<li>GO:0000166 has EC from proX and the e-value from the proX-seqY hit.
</ul>

<b>Searching in TCW</b>: If you search for GO:0000166 in a TCW database in the GO 
annotation basic search, it will be
shown in the table <b>IF</b> it has been assigned to a protein in the database or any of its
descendants have been assigned a protein. 

<p>Select GO:0000166 in the GO table followed by <b>View Sequences</b>; 
this will show all sequences that have been annotated by this GO, either assigned
or inherited.

<p>In the resulting table of sequences, select a sequence followed by <b>View Selected Sequence</b>; 
the <b>Sequence Detail</b> view will be shown for the sequence.
Select <b>Show GO for all hits</b> to see whether it is assigned (has EC) or inherited (no EC). 

<h3>Evidence codes (EC)</h3> 
They are shown in the pop-up hit table if there is a direct
assignment. They are also shown in the sequence GO table.

<table class=tabley>
<tr><th colspan=2>Experimental evidence			<th colspan=2>Computational analysis
<tr><td>EXP<td>Inferred from Experimental		<td>ISS<td>Inferred from Sequence or structural Similarity
<tr><td>IMP<td>Inferred from Mutant Phenotype	<td>ISA<td>Inferred from Sequence Alignment
<tr><td>IDA<td>Inferred from Direct Assay		<td>ISO<td>Inferred from Sequence Orthology
<tr><td>IGI<td>Inferred from Genetic Interaction<td>ISM<td>Inferred from Sequence Model
<tr><td>IPI<td>Inferred from Physical Interaction<td>IGC<td>Inferred from Genomic Context
<tr><td>IEP<td nowrap>Inferred from Expression Pattern<td>IBA<td nowrap>Inferred from Biological aspect of Ancestor
<tr><th colspan=2>Author statement<td>IBD<td>Inferred from Biological aspect of Descendent
<tr><td>TAS<td nowrap>Inferred from Traceable Author Statement<td>IKR<td>Inferred from Key Residues
<tr><td>NAS<td>Non-traceable Author Statement<td>IRD<td>Inferred from Rapid Divergence
<tr><th colspan=2>Curatorial judgement<td>RCA<td nowrap>Inferred from Reviewed Computational Analysis
<tr><td>IC<td>Inferred by Curator<th colspan=2>Automatically-assigned
<tr><td>ND<td>No biological Data available<td>IEA<td>Inferred from Electronic Annotation
</table>
<p>According to <a href="http://geneontology.org/page/guide-go-evidence-codes" class=ext target="_blank">http://geneontology.org/page/guide-go-evidence-codes</a>, 
all evidence codes are assigned by a curator except the last one. And the following
quote is from this site: "Evidence codes are <i>not</i> statements of the quality of the 
annotation. Within each evidence code classification, some methods produce annotations 
of higher confidence or greater specificity than other methods, in addition the way in 
which a technique has been applied or interpreted in a paper will also affect the quality 
of the resulting annotation. Thus evidence codes <i>cannot</i> be used as a measure of the 
quality of the annotation."
</body>
</html>
