<!DOCTYPE html>
<html>
<head>
<title>sTCW Reproduce</title>

<style>
.tabley    		{border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
.tabley td 		{border: 1px  solid black; padding: 3px; }
</style>

</head>

<body style="font-family: Arial, sans-serif;  font-size: 14px;">
<table style="width:70%; margin-left: auto; margin-right: auto;">
<tr><td>&nbsp;
<tr><td>&nbsp;

<tr><td>
<h2>Reproduce sTCW overview</h2>

This describes how to obtain the table of results corresponding to statistics
in the overview. The following short-hand is used:
<ul>
<li>The "Column:x" indicates that x should be selected for viewing in the table.
<li>#Seqs is the number of sequences, which is listed at the top of the overview.
<li>"Stats" is the "Show Column Stats" on the "Table..." drop-down.
</ul>
<b><i>Always clear filters before setting new ones!</i></b>
<h3>INPUT</h3>
Most of the INPUT section is data supplied by the user with runSingleTCW. The following
are computed:
<table class=tabley>
<tr><td colspan=3>Counts:
<tr><td>SIZE<td>All Seqs<td>Column:Counts for all conditions: Stats, column:Sum
<tr><td colspan=3>Sequences:
<tr><td>AVG-len<td>All Seqs<td>Column:Length; Stats, column:Average
<tr><td>MED-len<td>All Seqs<td>Column:Length; Stats, column:Median
<br>The median in the two cases may be slightly different because they are computed differently.
</table>

<h3>ANNOTATION</h3>
<h3>Hit statistics:</h3>

<table class=tabley>
<tr><th>Column<th>Search<th>Obtain number
<tr><td>Sequences with hits<td>Filters: Annotation: Annotated<td>Number of rows
<tr><td>Unique hits<td>AnnoDB Hits: Seq:None(slow)<td>Hits # above table
<tr><td>Total sequence hits<td>AnnoDB Hits: Seq:None(slow)<td>Pairs # above table
<tr><td>Bases covered by hit<td>AnnoDB Hits: Seq:Best Bits<td>Unselect "Group by Hit ID""; column:Align;
	<br>Stats, column:Sum; for NT, multiply by 3
<tr><td>Total bases for NT-sTCWdbs
	<br>and residues for AA-sTCWdbs
	<td>All Seqs<td>Column:Length; Stats, column:Sum
</table>

<h3>Annotation databases:</h3>
The first column is the DBtype-taxonomy of each annoDB (e.g. SP-plants: SP is SwissProt, plants is the taxonomy).
<table class=tabley>
<tr><th>Column<th>Search<th>Obtain number
<tr><td>ONLY
	<td>Filters: Annotation: Annotated, Best Bits,
	<br>Enter DBtype and Taxonomy for ANNODB,
	<br>General &lt;=1 annoDB
 	<td>Number of rows
<tr><td colspan=3><i>The following all use AnnoDB Hits panel with the correct ANNODB selected from the AnnoDBs panel.
</i>
<tr><td>BITS<td>Seq:Best Bits<td>Seqs # above table
<tr><td>ANNO<td>Seq:Best Anno<td>Seqs # above table

<tr><td>UNIQUE<td>Seq:None(slow)<td>Hits # above table
<tr><td>TOTAL<td>Seq:None(slow) <td>Pairs # above table

<tr><td>AVG %SIM<td>Seq:None(slow)<td>Unselect "Group by Hit ID"; column:%Sim; Stats, column:Average

<tr><td colspan=3>Rank=1 is the best hit for a sequence for a given annoDB.
<tr><td>HAS HIT<td>Seq:Rank=1<td>Seqs # above table; percentage of total #Seqs
<tr><td>AVG %SIM<td>Seq:Rank=1 <td>Uncheck "Group by Hit"; Column:%Sim; Stats, column:Average
<tr><td>Cover &gt;=N<td>Seq:Rank=1,%Sim&gt;=N,%HitCov<sup>*</sup>&gt;=N
	<td>Seqs # above table; percentage of HAS HIT
</table>
<sup>*</sup>HitCov is the difference between the hit stop and start coordinates divided by the length of the protein.

<h3>Top 15 species from total: N</h3>
The N is the number of unique species based on the first two words of the
species name. From "AnnoDB Hits":
<ul>
<li>Select "Species"", select "Two words"", enter first two words of species name next to "Find", select "Find", select the entry on the
left and add to the right.
<li>Select "Best Bits", "Best Anno" or "None" for the three numbers shown.
<li>BUILD TABLE
<li>Use the number listed beside "Pairs".
</ul>

<h3>Gene ontology statistics:</h3>
The counts of GO terms include assigned obsolete terms (level=0).
<table class=tabley>
<tr><th>Column<th>Search<th>Obtain number
<tr><td>Unique GOs
	<td>GO Annotation: no filters
	<td>Results number
<tr><td>Unique hits with GOs
	<td>AnnoDB Hits: Seqs:None; GO,etc:Has GO
	<td>Hits number at top of table
<tr><td>Sequences with GOs
	<td>Filters: Annotation: Annotated, Best with GO
	<td>Number of rows
<tr><td>Seq best hit has GOs
	<td>AnnoDB Hits: Seq:Best Bits; GO,etc:Has GO
	<td>Seqs number at top of table


<tr><td>biological_process
	<td>GO Annotation: Level: biological_process
	<td>Number GOs at top of table
<tr><td>molecular_function
	<td>GO Annotation: Level: molecular_function
	<td>Number GOs at top of table
<tr><td>cellular_component
	<td>GO Annotation: Level: cellular_component
	<td>Number GOs at top of table
<tr><td>is_a, part_of
	<td>GO Annotation:  no filters
	<td>Export..., Each GO's parents with relations, grep (see footnote<sup>*</sup>)
</table>
<sup>*</sup> From terminal, '<tt>grep is_a GOeachParents.tsv | wc</tt>'. Repeat with
<tt>is_a</tt> replaced with <tt>part_of</tt>.


<h2>EXPRESSION</h2>
The following sections may not exist if the input had no count files or the DE methods
were not executed.
<table class=tabley>
<tr><td>TPM:*
	<td>Filters: Counts and TPM: select Condition under Exclude; set "At Most" to 5000.
	<td>Continue using 1000, 100, 50, 5, 2 where the previous results are
subtracted from the current. The results are for intervals &gt;=N to &lt;M.
<tr><td>Differential Expression:
	<td>Filters:  Differential Expression: select the DE column, then enter the number, e.g. 1E-5.
	<td>These counts are accumulative.
<tr><td>GO enrichment:
	<td>GO Annotation: Enrich: enter the p-value threshold (e.g. 0.05); select green button next to "for" and select the p-value
	to filter on.
	<td>These counts are accumulative.
</table>
* If "RPKM" is at the top of the Overview instead of "TPM", then RPKM was computed instead of TPM.

<h2>SEQUENCES</h2>
If the sequences (e.g. ESTs) have been assembled, there are statistics on buried,
mate-pairs, etc. Most of them are columns, so select the column, then use the
"Show column stats" to view the sum. The following will only cover the three sections
that exists for all NT-sTCWdbs.
<table class=tabley>
<tr><td>Sequence lengths
	<td>Filters: General: Length: change &gt;= to &lt;=.
	Filter for each cutoff 100, 500, 1000, 2000, 3000, 4000, 5000.
	<td>The counts of the intervals to the left need to be subtracted from the number of rows.
<tr><td>ORF lengths
	<td>Filter: SNPs and ORFs: Has ORF &gt;= N.
	Filter for each cutoff 5001, 4001, 3001, 2001, 1001, 501, 101.
	<td>The counts of the intervals to the right need to be subtracted from the number of rows.
<tr><td>Quality
	<td>Filters: General: Has Ns: Yes; Columns: General: Ns
	<td>The resulting rows are the #n&gt;0; sort the rows in descending order on the Ns column,
	and scroll down to the row number that starts having #n&gt;10.
</table>

<h3>ORF stats:</h3>
<table class=tabley>
<tr><th>Column<th>Search<th>Obtain number
<tr><td colspan=3><i>The following use the Basic Sequence. Select "TCW" and enter
the Substring indicated.</i>
<tr><td>Is Longest ORF<td>!Lg<td>#Seqs minus Result number
<tr><td>Markov Best Score<td>!Mk<td>#Seqs minus Result number
<tr><td>All of the above<td>$<td>Result number
<tr><td>ORF=Hit<td>ORF=Hit<td>Result number
<tr><td>ORF=Hit with Ends<td>ORF=Hit+<td>Result number
<tr><td>Multi-frame<td>Multi<td>Result number
<tr><td>Stops in Hit<td>Stop<td>Result number
<tr><td>&gt;=9 Ns in ORF<td>9N<td>Result number

<tr><td colspan=3><i>The following use the Filters section: SNPs and ORFs.</i>
<tr><td>Has Hit<td>Protein confirmation<td>Number of rows
<tr><td>Both Ends<td>Has Ends (Start&Stop codons)<td>Number of rows
<tr><td>ORF&gt;=300<td>Has ORF &gt;= 300<td>Number of rows

</table>
<h3>GC content:</h3>
The only number reproducible is the GC Content, which is the %GC over the entire sequence.

<table class=tabley>
<tr><td>GC content<td>Seq Table<td>Column:%GC; Stats, column:Average
<br>Note, there will be some slight difference in the number due to round-off error.
</table>

<ul>
<li>The "Pos i" column is the percent of G or C in the ith positon of the CDS codons.
<li>The %GC is the percent of G's and C's over the sequence length.
<li>The CpG-O/E is ratio observed/expected [(#CpG/(#G*#C))*length].
<li>The UTRs can be viewed in the Sequence Detail alignments, but there is no column for it.
</ul>

</table>
</body>
</html>
