<!DOCTYPE html>
<html>
<head>
<title>TCW runAS</title>
<style>
	a:link {text-decoration: none;color:#6666CC;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.ext {font-weight: bold;}
	mark.red {color:red;background: none;}
	mark.green {color:green;background: none;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	ttp  {font-family: Arial, monospace; font-size: 14px; font-weight: 600; color: #556B2F; }
	ttl  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 600; color: DarkSlateGrey;}
	ttc  {font-family: 'Lucida Console', monospace; font-size: 14px; font-weight: 700; color: DarkSlateGrey;}
	.top {vertical-align: top;}
	.tx    {border: 0px; padding: 5px; border-spacing: 5px; }
	.tx td,th {border: 0px; padding: 5px; border-spacing: 5px}
	.ty    {border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
	.ty td, th {border: 1px  solid black; padding: 3px; }
	tt {font-size:13px}
	pre {font-size:13px}
	body {font-family:  Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="../img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial;  font-weight: bold;color:MidnightBlue">
					<font style="font-size: 20px;font-variant:small-caps;">runAS</font> -
					<font style="font-size: 15px;">Annotation Setup Guide</font>
				</td>

				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="../img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="../img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/TCW" class="white">TCW Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/TCW/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="../Docs.html" class="white">Docs</a> <mark class="white">|</mark>
						<a href="../tour" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->


<p>To prepare for annotation with <ttp>runSingleTCW</ttp>, it is necessary to download the databases to compared against.
The <ttp>runAS</ttp> program provides support for downloading the taxonomic and full UniProts along with mapping from the UniProt IDs to GO, KEGG, Pfam, EC, and InterPro.
<p><b>Tested</b>: <ttp>runAS</ttp> has been tested on Linux and MacOS. If you have any problems,
please let me know at <tt>tcw at agcol.arizona.edu</tt>.
<table style="width:90%">
	<tr>
		<td colspan="2"><b>Contents:</b>
		<tr>
			<td class="top">
			<ol style="line-height:1.4;">
				<li><a href="#over">Overview</a>
				<li><a href="#demo">Running the demo - <font color=green>Recommended</font></a>
				<li><a href="#run">Using <ttp>runAS</ttp></a>
				<li><a href="#details">Details and file structure</a>
				<li><a href="#clean">Cleanup</a>
				<li><a href="#mem">Memory and Time</a>
			</ol>
			</td>
			<td class="top">
			<ol start=7 style="line-height:1.4;">
				<li><a href="#best">What AnnoDBs to use</a>
				<li><a href="#other">Creating AnnoDBs from other databases (e.g. NCBI-nr)</a>
				<li><a href="#runstcw">Entering AnnoDBs and GOs into <ttp>runSingleTCW</ttp></a>
				<li><a href="#tax">Why use taxonomic databases</a>
				<li><a href="#obo">Parsing go-basic.obo</a>
				<li><a href="#link">Links to relevant databases</a>
			</ol>
			</td>
		</tr>
</table>

<a id="over"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Overview </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<b><i>Terminology</i></b>:
<p>The term "AnnoDB" refers to any database that will be used for annotation, i.e. the sequences in
TCW will be searched against all AnnoDB databases and the hits stored in the single TCW database
(sTCWdb) for query. </p>
<p><b><i>Requirements</i></b>: </p>

&rarr; <ttp>RunAS</ttp> uses <b><ttp>curl</ttp></b> for downloading annoDBs and the GO database.
<table class="tx">
	<tr>
		<td>You can get <ttp>curl</ttp> on most Linux machines with 'sudo yum install curl',
		and MacOS comes with it. If you cannot install it, <ttp>runAS</ttp> will prompt you as shown on the right;
		if you select <ttc>Continue</ttc>, it will perform the download with its own Java code, though it may take
		longer and is not as robust, i.e. could have potential problems due to network latency, etc. </td>

		<td><a href="img/runAScurl.png"><img src="img/runAScurl.png"
			alt="curl" style="border: 1px solid black; width: 300px"></a> </td>
	</tr>
</table>

<p><b><i>Processing steps</i></b>: The TCW <ttp>runAS</ttp> will perform the following: </p>
<ol>
	<li>Create the directory under <tt>projects/DBfasta</tt> for the downloads and generated FASTA files.
	<li>Download the selected Taxonomic UniProts <tt>.dat</tt> files and create FASTA files.
	<li>Download the selected full UniProt .dat file and create a FASTA file of the sequences not found in any the downloaded
	taxonomic files.
	<li>Create GO database, which contains mappings from UniProt IDs to GO, KEGG, EC, Pfam and InterPro.
	<ol type='a'>
		<li>Download <tt>go-basic.obo</tt> from <a href="http://current.geneontology.org/ontology/" class="ext" target="_blank"><tt>http://current.geneontology.org/ontology/</tt></a>
		<li>Create a local mySQL GO database (<tt>GOdb</tt>) with the information from this file.
		<li>Add information to the <tt>GOdb</tt> from the .fasta and .dat files in the UniProt directory.
	</ol>
	<li>Create the file <tt>projects/AnnoDBs_UniProt_&lt;date&gt;.cfg</tt> to be imported to <ttp>runSingleTCW</ttp>.
</ol>

<b><i>Important</i></b>:
<table class="tx">
	<tr>
		<td class="top"><a href="#mem">Memory and Time</a>
		<td>This can take a lot of memory and time, so make sure to read this section.
	</tr>
	<tr>
		<td  class="top"><a href="#best">What AnnoDBs to use</a>
		<td>To reduce the memory and time, make sure to read this section.
	</tr>
	<tr>
		<td  class="top"><a href="#other">Creating AnnoDBs from other...</a>
		<td  class="top">Other databases, such as NCBI nr, can be used for annotation but they will not have GO, KEGG, EC, Pfam, or InterPro.
	</tr>
</table>

<!--  ============= DEMO ======================== -->
<a id="demo"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Running the demo </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="tx" style="width:90%">
	<tr>
		<td class="top"> The TCW package provides subsets of UniProt for annotating the demo.
		In order to add GO annotations, a local GO mySQL 'demo' database needs to be created.
		<ol type="A">
			<li>From the TCW_4 directory, execute:
			<br>&nbsp;&nbsp;<tt>./runAS -d</tt>
			<br>The "-d" will cause it to enter the demo parameters, as shown on the right.
			The highlighted entries already exist. It is only necessary to build the GO database.
			<br>&nbsp; <br>&nbsp;

			<li>Execute <ttc>Build GO</ttc>.
			<br>&rarr; The GO tables are available for the demo, i.e. they will not be downloaded.
			<br>&rarr; When you select <ttc>Build GO</ttc>, a popup will say
				"<tt>GO files exist. Build GO database Only</tt>", click <ttc>Continue</ttc>.
			<br>&rarr; Building the GO database takes anywhere from 3-10 minutes.
		</ol>
		<td><a href="img/runASdemo.png"><img src="img/runASdemo.png" alt="demo" style="border: 1px solid black; width: 350px"></a>
</table>
<h4>
	Details about the Demo setup
</h4>
In the <tt>projects/DBfasta</tt> directory, there is the sub-directory <tt>UniProt_demo</tt> and <tt>GO_obodemo</tt>,
which contains the following:
<pre>
  GO_obodemo:
     go_basic.obo

  UniProt_demo:
    sp_bacteria/  sp_fungi/          sp_plants/       tr_plants/
    sp_full/      sp_invertebrates/  tr_invertebrates/
</pre>
Each taxonomic directory has a <tt>.dat</tt> and a <tt>.fasta</tt> file, which are very small subsets of the
original UniProt taxonomic <tt>.dat</tt> file.

<!-- ========== Java interface ================= -->
<a id="run"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Using <font style="font-weight: bold; font-variant: small-caps;">runAS</font></h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Typically, all you need to do is make sure you have an internet connection open and that you have enough disk space
		(see <a href="#mem">Memory</a>), then start the interface shown on the lower left by typing at the command line: <tt>./runAS</tt>

<table class="tx">
	<tr>
		<td>
		<ol>
			<li>The <ttl>TCW Annotation Directories</ttl> define where the files will be put. TCW provides defaults as shown on the right;
			it is recommended you use the defaults. <br>
			&nbsp;
			<li>Select the taxonomic databases you want to use, then select <ttc>Build Tax</ttc>,
			which downloads the respective .dat.gz files and creates FASTA files. <br>
			&nbsp;
			<li>Select the full databases you want to use, then select <ttc>Build Full</ttc>,
			which downloads the respective .dat.gz file and creates a subset FASTA file that only contains
			the sequence NOT in the downloaded taxonomic FASTA files. See <a href="#full">Full subsets</a> for more detail.
			<br>&nbsp;
			<li>Select <ttc>Build GO</ttc>, which downloads the GO database, creates a local mySQL GO database
			with a mapping of the UniProts from your downloaded set. This uses the information in <tt>HOSTS.cfg</tt>. <br>
			&nbsp;
			<li>Select <ttc>AnnoDB.cfg</ttc>, which writes a file called <tt>projects/AnnoDBs_UniProt_&lt;date&gt;.cfg</tt>
			that contains all the information downloaded; this can be used as input to <ttp>runSingleTCW</ttp>
			(see <a href="#runstcw">Import AnnoDBs</a>). <br>
			&nbsp;

		</ol>
		<td class="top"><a href="img/runAS.png"><img src="img/runAS.png" alt="runAS" style="border: 1px solid black; width: 350px"></a>
	</tr>
</table>
<ttc>Check</ttc>: The Check function automatically runs on startup and after any Build.
			It highlights everything that has been done.
			For example, the figure above shows that fungi and plant SwissProt have been downloaded and processed.
			<p><i>To force a check, or to view the UniProts in an existing goDB,  select the <ttc>Check</ttc>  button</i>.

<p><img src="../img/bullet_blue_small.gif" alt="bullet" style="border: 0px; width: 20px">
A log of the processing is written to <tt>projects/DBfasta/logs/runAS.log</tt>.
See the <a href="runAS.log.html" class="ext" target="_blank">log file</a> for an example.</p>
<br><u>Important points</u>:
<ul>
	<li><i><mark class="green"><ttp>runAS</ttp> will not replace an existing downloaded file</mark></i>:
	It will overwrite a .fasta file, but never a .dat file. If you want a .dat file downloaded again,
	you must remove it yourself. <br>
	&nbsp;
	<li><i><mark class="green">Build GOdb only after all desired taxonomic and full databases are downloaded</mark></i>: It is important that you create the GO database right after downloading
	the UniProt files so that they correspond. It is also important that you have downloaded all desired taxonomic and
	full UniProt databases. <br>


	&nbsp;
	<li><mark class="red"><i>Only download what you need!</i></mark>
	See <a href="#mem">Memory and Time</a> and <a href="#best">What AnnoDBs to use</a>.

	<br>&nbsp;
	<li><ttp>runMultitCW</ttp>: If multiple sTCWdbs are to be compared using <ttp>multiTCW</ttp>,
it is important they use the same set of AnnoDBs and GO database (see <a href="#runstcw">Entering AnnoDBs</a>).

</ul>
<a id=full></a>
<br><u>Full subsets:</u>
<table>
<tr><td class=top>When you select <ttc>Build Full</ttc>, a pop-up similar to the one on the right will be shown,
where only the taxonomic names will be shown that correspond to downloaded taxonomic .dat files.
This allows you to create different subsets. Typically, you will only want one subset, which is the
one corresponding to the taxonomic files downloaded.
<p>The FASTA file will have a suffix indicating what
subset it corresponds to. For example, the selection on the right would create the file <tt>uniprot_sprot_xBFxIxPxxV.fasta</tt>,
where the 10 characters represent the 10 taxonomic databases in alphabetic order, and the capital letters represent the
taxonomic sequences removed (Bacteria, Fungi, Invertebrate, Plant, Virus).
<p><i>Details:</i> You may unselect all entries and it will create a FASTA file of all sequences. When <ttp>runSingleTCW</ttp> loads UniProt IDs,
it only loads the first occurrence of a UniProt ID, so duplicates will not cause errors. However, by using the proper subset, processing is faster and
the e-values are lower since there are less sequences in the database.
You may create new subsets at any time as it does not effect the GOdb. Only one file will be shown in the <tt>AnnoDB.cfg</tt> file.
<td class=top><a href="img/runASsubset.png"><img src="img/runASsubset.png" alt="Full subset" style="border: 1px solid black; width: 200px"></a>
</table>


<a id="details"></a>
<!-- ============ Details ================ -->
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Details and file structure </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<b>Check</b>: Select to update the highlighting, as discussed below. The <ttc>Check</ttc> function is automatically run on startup,
and when any of the three "Builds" are executed.
<p>
	<b><i>Label Highlights</i></b>
</p>

<ul>
	<li>At the top:
	<ul>
	<li>If the <ttl>UniProt</ttl> directory label is highlighted in blue, it exists.
	<li>If the <ttl>GO</ttl> directory label is highlighted in pink, it exists but the GO OBO file has not been downloaded.
	<br>If the <ttl>GO</ttl> directory label is highlighted in blue, the GO OBO file has been downloaded.
	</ul>
	<li>On the middle right:
	<ul>
	<li>If the GO <ttl>Database</ttl> label is highlighted in blue, the GO database exists.
	</ul>
</ul>
<b><i>Taxonomic and Full UniProt Highlights</i></b>
<p>
	<u>Clear checkbox</u>: If a <ttl>Taxonomic</ttl> is clear, then neither the .dat file or .fasta file exists for it.
	When you check the box followed by <ttc>Build Tax</ttc>, you will need to confirm a popup that states "Download SP - xxx",
	where xxx will be the list of files to download. The download is always automatically followed by creating the .fasta files.
	The same applies to the <ttl>Full</ttl> checkboxes.
</p>
<p>
	<u>Pink checkbox</u>: If the .dat file exists, but the .fasta file does not, the checkbox will be highlighted pink.
	Check the pink box(s) and run <ttc>Build Tax</ttc> in order to create the .fasta file only.
	The same applies to the <ttl>Full</ttl> checkboxes.
</p>
<p>
	<u>Blue checkbox</u>: If both the .dat file and the .fasta file exists, the check box will be highlighted blue.
</p>
<p><b><i>File Structure</i></b> </p>
<p>For each taxonomic and full UniProt that you downloaded, a directory will be created under the <ttl>UniProt</ttl> directory.
For example, </p>
<pre>
  ./TCW/projects/DBfasta/UniProt_Dec2021%&gt; ls *
  sp_archaea:
  uniprot_sprot_archaea.dat.gz     uniprot_sprot_archaea.fasta

  sp_full:
  uniprot_sprot.dat.gz     uniprot_sprot_AxxxxxxxxV.fasta

  sp_viruses:
  uniprot_sprot_viruses.dat.gz      uniprot_sprot_viruses.fasta
</pre>
When you run the <ttp>BLAST</ttp> or <ttp>DIAMOND</ttp> search programs from <ttp>runSingleTCW</ttp>,
the formatted files will be placed in the corresponding directory.
<p><u>Compress Fasta</u>: If you plan on using <ttp>DIAMOND</ttp> as the search program, you may compress
the fasta files after download, e.g. </p>
<pre>
  cd projects/DBfasta/UniProt_&lt;date&gt;
  gzip */*.fasta
</pre>
<p><b><i>GO (Gene Ontology)</i></b></p>
<p>The <tt>go-basic.obo</tt> file is downloaded from
<a href="http://current.geneontology.org/ontology/" class="ext" target="_blank"><tt>http://current.geneontology.org/ontology/</tt></a>.</p>
<p><u>Database</u>: This text entry on the <ttp>runAS</ttp> interface is the name of the GO MySQL database that
will be created; you will enter this name in <ttp>runSingleTCW</ttp>. </p>
<p>The processing steps are as follows: </p>
<ol>
	<li>Download the GO Basic OBO file to <i>GO</i> directory.
	<li>Build a GO specific MySQL database (referred to as <tt>GOdb</tt>) with the contents of the file.
	<li>Add the UniProts from all subdirectories under the <tt>UniProt</tt> directory
	(e.g. <tt>projects/DBfasta/UniProt_Mar2021</tt>) to the <tt>GOdb</tt>.
</ol>



<a id="clean"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Clean up </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<ttp>runAS</ttp> does not remove files that are no longer necessary, which are the files downloaded from the internet:
<ul>
	<li>All "dat.gz" files in the UniProt directories, as the information has been transferred
	to the FASTA files and GO database.
	<li>The GO directory, as the information has been transferred to the GO database.
</ul>
These files can be removed, as <ttp>runSingleTCW</ttp> uses the FASTA files in the UniProt directories and
the GO mySQL database. However, if you do not have a space problem, keep them just for insurance; when UniProt
does the monthly update, your downloaded files will no be longer available on their site.
<p>For the FASTA files that you will be using <ttp>DIAMOND</ttp> to search against, you can gzip them as
<ttp>DIAMOND</ttp> can search against gzipped files.</p>
<p>When your calculating space, remember that the <ttp>BLAST</ttp> and <ttp>DIAMOND</ttp> programs will
format the .fasta file, which takes up even more space. For example: </p>
<pre>
  /TCW/projects/DBfasta/UniProt_Dec2021/sp_full% ls -hlG
  -rw-r--r--  1 cari  staff   597M Dec 20 07:07 uniprot_sprot.dat.gz
  -rw-r--r--  1 cari  staff    54M Dec 20 15:55 uniprot_sprot_xBFxIxPxxV.fasta
  -rw-r--r--  1 cari  staff    55M Dec 20 16:15 uniprot_sprot_xBFxIxPxxV.fasta.dmnd
</pre>

<!-- =============================================== -->
<a id="mem"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Memory and Time </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<b><i>Taxonomic</i></b>
<p>Downloads on 6-Jun-2021 onto a Linux machine with a ~500 Mbsp download connection and 128Gb of RAM
on a Sunday afternoon. Note, there can be considerable difference in download times.</p>
<table class="ty">
	<tr>
		<th>File
		<th>.dat Size
		<th>Download
		<th>.fasta Size<sup>1</sup>
		<th>Creation
	</tr>
	<tr>
		<td>uniprot_sprot_bacteria.dat.gz
		<td style="text-align: right">203Mb&nbsp;
		<td style="text-align: right">0m:27s
		<td style="text-align: right">150Mb&nbsp;
		<td style="text-align: right"> 0m:25s
	</tr>
	<tr>
		<td>uniprot_sprot_fungi.dat.gz
		<td style="text-align: right">49Mb&nbsp;
		<td style="text-align: right">0m:15s
		<td style="text-align: right">21Mb&nbsp;
		<td style="text-align: right">0m:04s
	</tr>
	<tr>
		<td>uniprot_sprot_invertebrates.dat.gz
		<td style="text-align: right">34Mb&nbsp;
		<td style="text-align: right">0m:05s
		<td style="text-align: right">14Mb&nbsp;
		<td style="text-align: right">0m:02s
	</tr>
	<tr>
		<td>uniprot_sprot_plants.dat.gz
		<td style="text-align: right">51Mb&nbsp;
		<td style="text-align: right">0m:10s
		<td style="text-align: right">21Mb&nbsp;
		<td style="text-align: right">0m:04s
	</tr>
	<tr>
		<td>uniprot_sprot_viruses.dat.gz
		<td style="text-align: right">16Mb&nbsp;
		<td style="text-align: right">0m:06s
		<td style="text-align: right">9Mb&nbsp;
		<td style="text-align: right">0m:01s
	</tr>
	<tr>
		<td>uniprot_sprot.dat.gz
		<td style="text-align: right">587Mb&nbsp;
		<td style="text-align: right">1m:09s
		<td style="text-align: right">55Mb<sup>2</sup>
		<td style="text-align: right">1m:43s
	</tr>
	<tr><td colspan=5>
	<tr>
		<td>uniprot_trembl_bacteria.dat.gz
		<td style="text-align: right"><mark class=red>87Gb&nbsp;</mark>
		<td style="text-align: right">1h:57m:02s
		<td style="text-align: right">64Gb&nbsp;
		<td style="text-align: right">2h:24m:45s
	</tr>
	<tr>
		<td>uniprot_trembl_fungi.dat.gz
		<td style="text-align: right">8.3Gb&nbsp;
		<td style="text-align: right">13m:41s
		<td style="text-align: right">7.4Gb&nbsp;
		<td style="text-align: right">13m:33s
	</tr>
	<tr>
		<td>uniprot_trembl_invertebrates.dat.gz
		<td style="text-align: right">7.8Gb&nbsp;
		<td style="text-align: right">12m:05s
		<td style="text-align: right">6.8Gb&nbsp;
		<td style="text-align: right">12m:50s
	</tr>
	<tr>
		<td>uniprot_trembl_plants.dat.gz
		<td style="text-align: right">12Gb&nbsp;
		<td style="text-align: right">16m:21s
		<td style="text-align: right">10.3Gb&nbsp;
		<td style="text-align: right">18m:57s
	</tr>
	<tr>
		<td>uniprot_trembl_viruses.dat.gz
		<td style="text-align: right">3.5Gb&nbsp;
		<td style="text-align: right">5m:20s
		<td style="text-align: right">2.3Gb&nbsp;
		<td style="text-align: right">5m:21s
	</tr>
</table>
<sup>1</sup><small>When TCW extracts the sequence into a FASTA file, it is not written in a gzipped format.
However, if you are going to use <ttp>DIAMOND</ttp>, you can zip them (the uniprot_trembl_bacteria.fasta zipped file is 31Gb).</small>
<br><sup>2</sup><small>The subset, i.e. full SwissProt minus all downloaded taxonomic entries.</small>

<p><b><i>GO database</i></b></p>
<p>It takes less than a minute to download the GO file. The time it takes to build the GO
	database is proportional to the number of UniProts to be processed. For example,</p>
<table class="ty">
	<tr>
		<th>Machine
		<th>AnnoDBs
		<th>Time
		<th>Database size
	</tr>
	<tr>
		<td>MacOS Catalina
		<td>SwissProt Plant and Full, TrEMBL Plant
		<td style="text-align: right">24m:47s
		<td style="text-align: right">2.7Gb
	</tr>
	<tr>
		<td>Linux (as specified above)
		<td>The 11 taxonomic and full listed above
		<td style="text-align: right">10h:42m:27s<sup>1</sup>
		<td style="text-align: right">26Gb<sup>1</sup>
	</tr>
</table>
<sup>1</sup><small>Most of the time, 8h:23m:30s, was for loading the uniprot_trembl_bacteria.dat,
which would also account for the database size.</small>

<!-- ================================= -->
<a id="best"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>What AnnoDBs to use </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<b><i>Strong suggestions:</i></b>
<ul>
	<li><i><mark class="red"> Only download what is relevant!</mark></i>
	<ul>
		<li>Download all relevant SwissProt files and the Full SwissProt UniProt.
		<li>Download only the most relevant TrEMBL files, and never the Full TrEMBL UniProt unless absolutely necessary.
	</ul>
	<li><mark class="red"><i>Do not perform constant downloads, it is a drain on the UniProt servers.</i></mark> <br>
	The UniProts do not change that fast, and it changes 'best' hits in TCW, which can disturb any on-going analysis.
</ul>
<b><i>Evidence</i></b>: In order to show that it is sufficient to just download the most relevant databases,
the following test was performed.
<p>The dataset used for the following tests is from <i>de novo</i> assembled sequences from
<i>Andropogon gerardii</i>, which is related to Sorghum. It was downloaded from
<a href="https://doi.org/10.5061/dryad.5bk4c" class="ext" target="_blank">Dryad</a> and published by
<a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12733" class="ext" target="_blank">Hoffman and Smith (2017)</a>.
The full dataset had &gt;60k transcripts, which was reduced to 27,085 for faster tests.
</p>
<p>Four annotations were compared: </p>
<table class="ty">
	<tr>
		<th>Annotation
		<th>AnnoDBs
		<th>#Annotated
	</tr>
	<tr>
		<td>#1
		<td>sp_plants, tr_plants, sp_ful
		<td>25,049 (92.5%)
	</tr>
	<tr>
		<td>#2
		<td>#1 + sp_virus, sp_fungi, sp_invertebrate, sp_bacteria
		<td>25,052 (92.5%)
	</tr>
	<tr>
		<td>#3
		<td>#2 + tr_virus, tr_fungi, tr_invertebrate, tr_bacteria, tr_full
		<td>25,070 (92.6%)
	</tr>
	<tr>
		<td>#4
		<td>#1 + nr
		<td>25,160 (92.9%)
	</tr>
</table>
<br>
Using only sp_plants, tr_plants, and sp_ful, 92.5% of the transcripts were annotated compared
with 92.9% using the entire NR database.
If your organism is not closely related to any model organism, then there will likely be a bigger difference.

<!-- ============================================= -->
<a id="other"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Creating AnnoDBs from other databases </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

UniProt and NCBI-nr descriptor lines works with TCW. For other databases, you will need to make sure they
have a TCW accepted descriptor line.
<p><b>Description lines</b>
<p>The description line is the ">" line that describes the subsequent sequence in a FASTA file.
From it, <ttp>runSingleTCW</ttp> extracts:

<ul>
	<li><tt>DBtype</tt>: used in naming the tab output file and is used in <ttp>viewSingleTCW</ttp>
	to aid in identifying where the hitID is from.
	<li><tt>hitID</tt>: the unique identifier of the hit.
	<li><tt>description</tt>: generally the functional description
	<li><tt>species</tt>: the species
</ul>
<br><b>UniProt</b>
<pre>
  >sp|Q9V2L2|1A1D_PYRAB Putative 1-ami OS=Pyrococcus abyssi GN=PYRAB00630 PE=3 SV=1v
</pre>
<ul>
	<li>The 'sp' is the <tt>DBtype</tt>. For TrEMBL, the first two characters would be 'tr'.
	<li>The third entry of the first string is the identifier (e.g. 1A1D_PYRAB)
	<li>The string up to the OS is the description.
	<li>The string after the "OS=" is the species.
</ul>
<a id="nr"></a>
<p><b>NCBI nr</b> (See <a href="#link">Download NR</a>)</p>
<pre>
  >XP_642837.1 hypothetical protein DDB_G0276911 [Dictyostelium discoideum AX4]
</pre>
<ul>
	<li>The first entry is the identifier (e.g. XP_642837.1). Note, there is no longer a way to detect
	the database origin within the file, hence, the <tt>DBtype</tt> will be the generic 'PR' for protein.
	<li>The text from the first space to the first "[" is the description.
	<li>The text within the "[]" is the species.
</ul>
As it does not have an "type code", its type will default to "PR". If the taxonomy is given as "nr", the
TCW abbreviation for this database will be <u>PRnr</u>.

<!--  ============== -->
<p><b>Generic</b></p>
<p>If you have a file other than UniProt or nr, make the descriptor names as follows: </p>
<pre>
  >CC|ID description OS=species
</pre>
<ul>
	<li>CC is the type code, and will be used as the <tt>DBtype</tt> in TCW.
	<li>ID is the unique identifier
	<li>Everything up to the OS is the description
	<li>Everything after the OS is the species
</ul>
<p><u>Example 1</u>: The TCW perl script <tt>scripts/formatPlantTFDB.pl</tt> takes as input a file from
<a href="http://planttfdb.gao-lab.org/download.php" class="ext" target="_blank">PlantTFDB</a>,
which has header lines like: </p>
<pre>
  >KFK36254.1 Arabis alpina|G2-like|G2-like family protein
</pre>
and converts them to header lines:
<pre>
  >tf|G2_like_1 G2-like family protein {KFK36254.1} OS=Arabis alpina
</pre>
The DBtype will be "tf"". If the taxonomy entered into <ttp>runSingleTCW</ttp> is "plants"",
the abbreviation for this database will be <u>TFpla</u>.
<p><u>Example 2</u>: The TCW python script <tt>scripts/formatNCBIrna.py</tt> takes as
input an RNA file from NCBI, which has header lines like: </p>
<pre>
  >XM_002436391.2 PREDICTED: Sorghum bicolor GDP-mannose 4,6 dehydratase 1 (LOC8069086), mRNA
</pre>
and converts them to header lines:
<pre>
  >XM_002436391.2 GDP-mannose 4,6 dehydratase 1 (LOC8069086), mRNA OS=Sorghum bicolor
</pre>
As this does not have a type code at the beginning, its type will default to "NT".
If the taxonomy is entered as "sb", the abbreviation for this database will be <u>NTsb</u>.
The script can be modified to add a type code.

<!-- ========================================= -->
<a id="runstcw"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Entering AnnoDBs and GOs into runSingleTCW </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

Execute <tt>./runSingleTCW</tt> and select your project.
<p>&rarr; Select <ttc>Import Anno</ttc>, a file chooser will popup. Select either of the following to enter the names of the
	UniProt in the <ttl>AnnoDB</ttl> table and the GO database:
	<table class="ty">
	<tr><td><tt>projects/AnnoDBs_UniProt_&lt;date&gt;.cfg</tt><td> This will use the AnnoDBs &amp; GO written by <ttl>AnnoDB.cfg</ttl>.
	<tr><td><tt>projects/&lt;project-name&gt;/sTCW.cfg</tt><td> This will use the AnnoDBs &amp; GO used by another project.
	</table>
<p>&rarr; Now you are ready to run <ttc>Annotate</ttc> with the UniProt and GO you just downloaded.
<p>
<a href="img/runSingleStep3.png"><img src="img/runSingleStep3.png" alt="" style="border: 1px solid black; width: 80%"></a>
<p>AnnoDBs can be entered using the <ttc>Add</ttc> button, where the taxonomy is defined. They can also be changed with <ttc>Edit</ttc>.
<p>The GO database and GO slim category are defined or changed in the <ttc>Options</ttc> menu. <a id="tax"></a> </p>

<!-- ====================================== -->
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Why use taxonomic databases instead of the full UniProt </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<ttp>viewSingleTCW</ttp> refers to the annoDBs by the 'DBtype' and 'taxonomy', with them
combined into 'DBtax'. The DBtype and taxonomy can be queried on and columns of the data viewed.
The "sp" is SwissProt and the "tr" is "TrEMBL".
<p>The following shows an example of a set of hit proteins: </p>
<p><a href="img/runASview.png"><img src="img/runASview.png" alt="" style="border: 1px solid black;"></a> </p>
<p>The following shows a table of sequences: </p>
<p><a href="img/runASview.png"><img src="img/runASview2.png" alt="" style="border: 1px solid black;"></a> </p>
<p>The following shows the details of a specific sequence: </p>
<a href="img/runASview.png"><img src="img/runASview3.png" alt="" style="border: 1px solid black;"></a>


<!-- ============================================= -->
<a id="obo"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Parsing go-basic.obo</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<i>The following is an example record in the OBO file:</i>
<pre>
  [Term]
  id: GO:0000785
  name: chromatin
  namespace: cellular_component
  alt_id: GO:0000789
  alt_id: GO:0000790
  alt_id: GO:0005717
  def: "The ordered and organized complex of DNA, protein, ....
  comment: Chromosomes include parts that are not part of  ....
  synonym: "chromosome scaffold" RELATED []
  synonym: "cytoplasmic chromatin" NARROW []
  synonym: "nuclear chromatin" NARROW []
  xref: NIF_Subcellular:sao1615953555
  is_a: GO:0110165 ! cellular anatomical entity
  relationship: part_of GO:0005694 ! chromosome
</pre>

<i>TCW parses for the following keywords:</i>
<table class="ty">
<tr><th>Keyword		<th>AmiGO term<th>TCW term<th>Example
<tr><td>id			<td>Accession	<td>GO ID		<td>GO:0000785
<tr><td>name		<td>Name		<td>Description	<td>chromatin
<tr><td>namespace	<td>Ontology	<td>Domain		<td>cellular_component
<tr><td>is_a		<td>is_a		<td>is_a		<td>GO:0110165
<tr><td>relationship: part_of<td>?	<td>part_of		<td>GO:0005694
<tr><td>alt-id		<td>Alternate ID<td>Alternate ID	<td>GO:0000790
<tr><td>&nbsp;		<td>replaced by	<td>Replaced by<td>GO:0000785
<tr><td>is_obsolete: true<td>Name: obsolete<td>Description: obsolete<td>obsolete replicative cell aging
</table>

<p><i>Views in AmiGO and TCW:</i></p>
<table class="tx">
<tr>
<th>AmiGO<th>TCW
<tr>
<td class=top><a href="img/runASgo785.png"><img src="img/runASgo785.png" alt="" style="border: 1px solid black; width: 450px"></a>
<td class=top><a href="img/runASgo785t.png"><img src="img/runASgo785t.png" alt="" style="border: 1px solid black; width: 350px"></a>
<tr>
<td><a href="img/runASgo790.png"><img src="img/runASgo790.png" alt="" style="border: 1px solid black; width: 250px"></a>
<td><a href="img/runASgo790t.png"><img src="img/runASgo790t.png" alt="" style="border: 1px solid black; width: 350px"></a>
</tr>
</table>

<i><u>NOTES:</u></i>
<ol>
<li>UniProt occasionally uses the Alternate IDs and has a few Obsolete GO terms.
<li>I cannot guarantee that AmiGO always treats "alt_id" as specified here.
</ol>

<!-- ============================================= -->
<a id="link"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Links to relevant databases </h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

To download the UniProt files without <ttp>runAS</ttp>:
<ul>
	<li>Go to <a href="https://www.uniprot.org/downloads" class="ext" target="_blank">UniProt Downloads</a>.
	<li>In the second line from the top, it says "For downloading complete data sets we recommend
	using <b>ftp.uniprot.org</b>." Click the <b>ftp.uniprot.org</b>.
	<li>This brings up the UniProt download directories in a Finder window. You may view it as "Guest".
	<li>Click "Current_release", "knowledgebase". Here you will see "complete" and "taxonomic_divisions".
</ul>
The NCBI-nr database can be downloaded:
<ul>
	<li><a href="https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/" class="ext" target="_blank">NCBI nr</a>
	(https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA).
	<li>As of 24-Jan-21, it is 89GB and took 1h:45m to download.
	<li>It is called <tt>nr.gz</tt>; since the File Chooser requires a FASTA suffix,
	rename it: <tt>mv nr.gz nr.fa.gz</tt>
</ul>
GO Basic OBO file: <a href="http://geneontology.org/docs/download-ontology/" class="ext" target="_blank">
<tt>http://geneontology.org/docs/download-ontology/</tt></a></p>
	<!---- END BODY -->

	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:tcw@agcol.arizona.edu">tcw@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>
