<!DOCTYPE html>
<html>
<head>
<title>Search</title>
<style>
.tabley    		{border: 1px  solid black; border-spacing: 0px;  border-collapse: collapse;}
.tabley td 		{border: 1px  solid black; padding: 3px; }
</style>
</head>
<body style="padding:20px;width:650px;font-family:Times">

<h2>Searching</h2>
Search sequence(s) against sTCW NT-sequences, sTCW AA-ORFs, or a selected database
of protein sequences. 
<ol>
<li><b>Query:</b> Input one or more sequence in FASTA format into the entry box. 
<ul>
<li>If your sequence contains characters A,G,C,T, or N, it will
be treated as a nucleotide sequence, else a protein sequence.
<li>From the Sequence Detail page, you can use the <b>Copy...</b> command to copy 
a hit sequence or the seqID sequence, and <b>Paste</b> it here.
</ul>
<p>
<li><b>Subject:</b> 
<br>For NT-sTCW, there will be the following options: 
<ul>
<li>NT-Seq (nucleotide sequences in the database)
<li>AA-ORF (translated ORFs from the nucleotide sequences)
<li>AA-DB  (selected protein file with FASTA suffix: .fa, .fasta, .fna, .ffn, .faa, or .frn)
</ul>
For AA-sTCW, the options will be:
<ul>
<li>AA-Seq (amino acid sequences in the database) 
<li>AA-DB (a selected protein database)
</ul>
<p>
<li>You have a choice to use <tt>BLAST</tt> and <tt>DIAMOND</tt>. 
<tt>DIAMOND</tt> can only be used with an AA subject (blastp or blastx mode).

<p> 
<li>You may enter/change
parameters, though you must <i>make sure they are correct</i> since TCW does not verify them.
A list of some of the possible parameters are shown in the next section.

<p>
<li>Select <b>RUN SEARCH</b> to execute the search.
The results are shown in the bottom half of the panel, and are also written
to a file in the "/hitResults" directory. 

<p>
<li>If tabular format is chosen, then you can
select a row followed by <b>View Selected Sequence</b>, which displays the detail view of the sequence. 

</ol>

<b>Delete Files</b>: Files written by the TCW search command will go into the "/ResultHits"
directory. These files will end with ".fa" or be formated files. These files will be
deleted when you execute this command. Result files will not be deleted. If you select a protein database that is
not in the "/ResultHits" directory, the formatted files are not deleted. 

<p>
<b>Reset</b>: Sets the display back to all defaults.
 

<h4>No Results Found</h4>
If you do not get the results you expect:
<ol>
<li>Remember that the original annotation is typically 1000's of sTCW sequences compared with
1000's of annotation sequences, whereas this compares just a few sequences against 1000's, which
causes the e-values to be different.
<br><br>
<li>Select <b>Trace</b> and <b>RUN SEARCH</b> again. The command is written to 
the terminal; copy it and run it from the terminal. If there was
any error (e.g. in the parameters), it will tell you the error. 
</ol>

<h4>Original Search Results</h4>
When the database was built, the sequences were annotated with one or more "AnnoDBs" (e.g. UniProt).
In order to view the search results, select a given sequence for the query, then the appropriate
AnnoDB (usually in <tt>/projects/DBfasta</tt>) for the subject. Note that the E-value and Bitscore will be different,
and there is a chance it will not ever be shown since the input is one sequence versus 1000's.

<h4>Relevant parameters</h4>
<b>BLASTX</b>:
Execute 'blastx -help' or 'blastp -help' from the command line to view the full set of parameters.

<table class=tabley>
<tr><th>Parameter<th>Default<th >Description
<tr><td nowrap>-evalue<td>10 (1e+01)<td>Maximum e-value to report alignments
<tr><td  nowrap>-max_target_seqs<td>500<td> Maximum number of aligned sequences to keep (outfmt=6)
<tr><td>-num_threads<td>1<td>Number of threads (CPUs) to use in the BLAST search

<tr><td>-word_size<td>??<td>Integer Word size
<tr><td>-gapopen<td>??<td>Integer Cost to extend a gap 
<tr><td>-matrix<td>BLOSUM62<td>Scoring matrix name

<tr><td>-use_sw_tback<td>N/A<td> Compute locally optimal Smith-Waterman alignments 

<tr><td>-dbsize<td>Size of input<td>Effective length of the database
<tr><td>-max_hsps<td>--<td>Set maximum number of HSPs per subject sequence to save for each query
</table>


<p><b>DIAMOND</b>: Execute 'diamond help' from the command line to view the full set of parameters.
<table class=tabley>
<tr><td>Parameter<td>Default<td>Description
<tr><td>--evalue (-e)<td>0.001 (1e-03)<td>maximum e-value to report alignments
<tr><td nowrap>--max-target-seqs (-k)<td>--<td>Maximum number of target sequences to report alignments for
<tr><td>--threads (-p)<td>1<td>Number of threads (CPUs) to use in the BLAST search
<tr><td>--id<td>--<td> Minimum identity% to report an alignment
<tr><td>--sensitive<td>off<td>       Enable sensitive mode (slower)
<tr><td>--more-sensitive<td>off<td>  Enable more sensitive mode (even slower)
<tr><td>--matrix  <td>BLOSUM62<td>score matrix for protein alignment
<tr><td>--masking<td>1<td>enable masking of low complexity regions (0/1=default)
<tr><td>--max-hsps<td>--<td>Setting this to 1 causes only one HSP for each query/subject pair.
<tr><td>--min-orf (-l)<td>--<td>Ignore translated sequences that do not contain an open reading 
frame of at least this length. By default this feature is disabled for sequences of length below 30, 
set to 20 for sequences of length below 100, and set to 40 otherwise. Setting this option to 1 
will disable this feature.
<tr><td>--query-cover #<td>--<td>Report only alignments above the given percentage of query cover.
<tr><td>--subject-cover #<td>--<td>Report only alignments above the given percentage of subject cover.
</table>
</body>
</html>
