<html>
<body style="padding:20px;width:650px;font-family:Times">
<title>GO Help</title>
<h1>GO Help</h1>
<p>The version of GO that was used for the TCW annotation is shown on the Overview.
Unless TCW was built with the latest GO database, there will be 
differences from what is found on amigo.geneontology.org as the
gene ontology is constantly being updated. The following examples come from GO dated
201608 (i.e. August 2016).

<p>Proteins in the UniProt database have been "assigned" GO terms. 
TCW has the following mappings: 
<ul>
<li>Protein names to sequences (from a heursitic search program such as Blast or Diamond), 
<li>GO terms to proteins (from UniProt), 
<li>GO terms to sequences (the GOs assigned to the proteins that hit the sequence). 
</ul>When a protein is assigned a GO term, the ancestors of the GO term's 
inherits the hit and sequences of the hit; this is because ancestors are
increasingly more general terms and all apply. Hence, the GO terms in TCW
are all the <i>assigned (direct)</i> and <i>inherited (indirect)</i> GO terms.  

<p><b>Figure 1: Ancestor List for nucleotide binding (GO:0000166).</b> 
<p>
<table>
<tr><th align="left">A. Ordered Ancestors (by Dist)<th align="left">B. Ancestors (by Level)
<tr><td>
<pre>
GO term    Dist Description 
GO:0003674  4   molecular_function 
GO:0005488  3   binding 
GO:1901363  2   heterocyclic compound binding 
GO:0097159  2   organic cyclic compound binding 
GO:1901265  1   nucleoside phosphate binding 
GO:0036094  1   small molecule binding 
</pre>
<td>
<pre>
GO term    Level Description 
GO:0003674   1   molecular_function
GO:0005488   2   binding
GO:1901363   3   heterocyclic compound binding
GO:0097159   3   organic cyclic compound binding
GO:0036094   3   small molecule binding
GO:1901265   4   nucleoside phosphate binding
</pre>
</table>
<br>Any protein hit assigned to 'nucleotide binding' will be inherited these 6 terms,
e.g. if its nucleotide binding, it is also 'small molecule binding'.
<p>The "ordered" list corresponds to the ordering shown in Amigo for the "Inferred Tree". 
The level-based list is provided as its faster to compute in TCW. See section on "Level versus 
Distance".

<p><b>Figure 2: Ancestor Paths and Graph for Nucleotide binding (GO:0000166)</b>
<p><table>
<tr><th align="left">A. TCW GO Ancestor Path<th align="left">B. AmiGO Graph
<tr>
<td><table border=1 width=>
<tr><th>&nbsp;<th>Path1<th>Path2<th>Path3
<tr><td>1<td align=center bgcolor="#d3d3d3 " colSpan=3>molecular function
<tr><td>2<td align=center bgcolor="#d3d3d3 " colSpan=3>binding
<tr><td>3<td width=100>heterocyclic compound binding<td width=100>organic cyclic compound binding(1)<td width=100>small molecule binding(1)
<tr><td>4<td width=100>nucleoside phosphate binding<td width=100>nucleoside phosphate binding<td>&nbsp;
<tr><td>5<td align=center bgcolor="#d3d3d3 " colSpan=3>nucleotide binding
</table>
<br>The numbers on the right of the paths table correspond to 'levels',
however, a term may be in multiple levels, e.g.
nucleotide binding is in level 4 for Path3 but level 5 for Path2; it is only shown in its lowest
level.

<br>&nbsp;<br>The paths can be followed in Figure 2B by following the arrows.
<td><img src=GO166.png></img>
</table>


<p><b>Figure 3: Descendant and Neighbor List for Nucleotide binding (GO:0000166) </b>
<p><table>
<tr><th align="left">A. TCW GO Descendent List<th align="left">B. GO Neighbor List
<tr><td>
<pre>
GO term    Level Description 
GO:0051287   6   NAD binding
GO:0050661   6   NADP binding
GO:0017076   6   purine nucleotide binding
GO:0032553   6   ribonucleotide binding
GO:0030554   7   adenyl nucleotide binding
GO:0010181   7   FMN binding
GO:0019001   7   guanyl nucleotide binding
etc (18 terms total)

*Descendents are only shown if they have a hit
</pre>
<td>
<pre>
Parents:
GO:0036094  is_a  mole  small molecule binding
GO:1901265  is_a  mole  nucleoside phosphate binding

Children*:
GO:0017076  is_a  mole  purine nucleotide binding
GO:0032553  is_a  mole  ribonucleotide binding
GO:0050661  is_a  mole  NADP binding
GO:0051287  is_a  mole  NAD binding

*Children are only shown if they have a hit
</pre>
</table>
Any hit assigned to the descendants will be inherited by 'nucleotide binding'.
<p>
<b>NOTE:</b> In TCW, only GOs that have an assigned or inherited hit will be in the database,
hence, a GOs descendants will not necessarily all be in the TCW database, but its
ancestors will be.
<p>
The relation types used in TCW are 'is_a' and 'part_of'.

<h3>Graph terms</h3>
In computer science, a graph has "nodes" and "edges". There are 
"paths" between nodes by following the edges. 
The graph is 'directed' if the edges have arrows, where a 'child' node points to the 'parent' node.
In the gene ontology graphs, the GO terms are
represented by nodes; an edge from a child node to a parent node indicates
its a specialized term of the parent node.  
The gene ontology has three root nodes, "biological process", "cellular_component",
and "molecular_function". GO is not strictly hierarchical as a node can have more than
one parent. See http://geneontology.org/page/ontology-structure for 
a more in-depth discussion.
<p>TCW does not uses edges that cross the three trees. 

<h3>Level versus distance</h3>
Distance and level are the reverse of each other. Figure 1 shows distance and Figure 2A shows
levels. For example, the molecular_function is at level 1 and has a distance of 3 and 4; 
Nucleotide binding has a level of 4 and 5 and a distance of 0 (to itself).
<p>Gene Ontology only uses distance. For convenience, TCW (and some other software) allows
queries on level, where TCW uses the maximum level for queries (e.g. nucleotide binding
would be considered at level 5).

<h3>E-values</h3>
A GO term may be assigned or inherited by many protein hits in the TCW database. Each protein hit
has an e-value indicating how well it aligned to the sequence. The GO-protein hit is assigned the
best e-value over all the hit-sequence pairs. 

<h3>Examples</h3>
Say protein X (proX) aligns to sequence Y (seqY) with an e-value, and GO:0000166 is assigned to proX with an evidence code EC, 
then seqY is also said to be annotated with GO:0000166. 
<ul>
<li>Since proX is assigned GO:0000166, it also inherits all GO:0000166 ancestors.
<li>Since seqY has annotation of GO:0000166, it also inherits all GO:0000166 ancestors.
<li>Since GO:0000166 is assigned to proX, all its ancestors inherit proX.
<li>Since GO:0000166 annotates seqY, all its ancestors also annotated seqY.
<li>GO:0000166 has EC from proX and e-value from proX-seqY hit.
</ul>

<b>Searching in TCW</b>: If you search for GO:0000166 in a TCW database in the GO 
annotation basic search, it will be
shown in the table <b>IF</b> it has been assigned to a protein in the database or any of its
descendants have been assigned a protein. 

<p>Select GO:0000166 in the GO table followed by "View Sequences"; 
this will show all sequences that have been annotated by this GO, either assigned
or inherited.

<p>In the resulting table of sequences, select a sequence followed by "View Selected Sequence"; 
the "Sequence Detail" view will be shown for the sequence.
Select "Show GO for all hits" to see whether it is assigned (has EC) or inherited (no EC). 

<h3>Evidence codes (EC)</h3> 
They are shown in the pop-up hit table if there is a direct
assignment. They are also shown in the sequence GO table.

<table border=1 cellpadding=4>
<tr><th colspan=2>Experimental evidence			<th colspan=2>Computational analysis
<tr><td>EXP<td>Inferred from Experimental		<td>ISS<td>Inferred from Sequence or structural Similarity
<tr><td>IDA<td>Inferred from Direct Assay		<td>ISO<td>Inferred from Sequence Orthology
<tr><td>IMP<td>Inferred from Mutant Phenotype	<td>ISA<td>Inferred from Sequence Alignment
<tr><td>IGI<td>Inferred from Genetic Interaction<td>ISM<td>Inferred from Sequence Model
<tr><td>IPI<td>Inferred from Physical Interaction<td>IGC<td>Inferred from Genomic Context
<tr><td>IEP<td nowrap>Inferred from Expression Pattern<td>IBA<td nowrap>Inferred from Biological aspect of Ancestor
<tr><th colspan=2>Author statement<td>IBD<td>Inferred from Biological aspect of Descendent
<tr><td>TAS<td nowrap>Inferred from Traceable Author Statement<td>IKR<td>Inferred from Key Residues
<tr><td>NAS<td>Non-traceable Author Statement<td>IRD<td>Inferred from Rapid Divergence
<tr><th colspan=2>Curatorial judgement<td>RCA<td nowrap>Inferred from Reviewed Computational Analysis
<tr><td>IC<td>Inferred by Curator<th colspan=2>Automatically-assigned
<tr><td>ND<td>No biological Data available<td>IEA<td>Inferred from Electronic Annotation
</table>
<p>According to http://geneontology.org/page/guide-go-evidence-codes, 
all evidence codes are assigned by a curator except the last one. And the following
quote is from this site: "Evidence codes are <b>not</b> statements of the quality of the 
annotation. Within each evidence code classification, some methods produce annotations 
of higher confidence or greater specificity than other methods, in addition the way in 
which a technique has been applied or interpreted in a paper will also affect the quality 
of the resulting annotation. Thus evidence codes <b>cannot</b> be used as a measure of the 
quality of the annotation."
</html>
