<!DOCTYPE html>
<html>
<head>
<title>singleTCW Instructions</title>
<style>
.tablex    		{border: 0px; padding: 3px; border-spacing: 3px; }
.tablex td 		{border: 0px; padding: 3px; border-spacing: 3px; }
</style>
</head>
<body style="padding:20px;width:650px;font-family:Times;  width:600px">

<h1>Instructions</h1>
Terminology:
<table class=tablex>
<tr><td valign=top><i>Sequence</i></td>
<td>The input may be TCW-assembled transcripts, transcript sequences, proteins or genes.</td>
</tr>
<tr><td valign=top><i>Conditions</i></td>
<td>A condition may be a tissue type, treatment, etc.</td>
</tr>  
<tr><td valign=top><i>AnnoDB</i></td>
<td>A database in FASTA format against which the sTCW sequences were searched against
for annotation. The search used  <tt>BLAST</tt> or <tt>DIAMOND</tt>.</td>
</tr>

<tr><td valign=top><i>DB hit</i></td>
<td>An alignment of an entry from an annoDB (e.g. UniProt) to an sequence. A given
sequence can <i>hit</i> an annoDB multiple times (have alignments to multiple entries) and
it can hit multiple different annoDBs (have alignments in more than one annoDB). </td>
</tr>
</table>

<table class=tablex>
<tr><td colspan=2><u>The following are characterizations assigned to DB hits</u></td>
</tr>
<tr><td><i>Best Bits (BS/BITS)</i></td>
<td>The hit with the best bitscore. 
If there is a tie, a SP (SwissProt) or TR (TrEMBL) takes precedence.
If there is still a tie, the best e-value takes precedence.</td>
</tr>

<tr><td  valign=top nowrap><i>Best Anno (AN/ANNO)</i></td>
<td>The hit with the best
bitscore from the list of informative descriptions, where an informative description
is one that does not contain phrases such as  "uncharacterized protein". 
It also looks for names such as "Os02g0768200 protein", though this means it sometimes 
gets it wrong, e.g "B1 protein" is detected as non-informative as it has the general composition of a name.
(See "Special Case" below).
</td>
</tr>
<tr><td valign=top nowrap><i>Best with GO (WG)</i></td>
<td>From a list sorted on bitscore, the first one with at least one GO ID assigned.</td>
</tr>

<tr><td valign=top><i>Rank</i></td>
<td>The rank is assigned based on the bitscore sort followed by the 
e-value sort, i.e. when the bitscore is tied, the order is
based on the e-value.
Note that the <tt>Blast</tt> and <tt>Diamond</tt> default order is the e-value sort followed by the
bitscore. </td>
</tr>
</table>
NOTE: The bitscore takes precedence over the e-value since the e-value takes into account the size of the database, wherease
the bitscore will be the same regardless of the size of the database.
<p>Special Case - <i>Best Anno (AN/ANNO)</i>:  The database may having been built with the "SwissProt precedence" option set, in which
case, this will be stated at the end of the <b>Overview</b> panel. When this is selected, a SwissProt hit will be selected
if its bitscore is sufficiently close the non-SP best annotation.

<hr>
<p>Detailed Help is provided on each page; the following provides an overview
of what is available.

<h2>General</h2>
<b>&gt;&gt;Instructions</b> - this page
<p><b>&gt;&gt;Overview</b> - Summary of everything in database.
<p><b>&gt;&gt;Decimal Display</b> - Specify how to display decimal numbers and highlight p-values.
<p><b>&gt;&gt;Find hits</b> - Search a sequence against the sequences in the database or a user
selected protein database. This requires the <tt>Blast</tt> and optional <tt>Diamond</tt> paths to 
be in the HOSTS.cfg file. 

<h2>Sequences</h2>

<b>&gt;&gt;Show All</b> - A table of all sequences in the database. 
<ul>
<li>To view a sequence from the table, select the sequence and then "View Selected Sequence".
<li>Change the columns in the table using "Columns" under "Sequence.
</ul>

<b>&gt;&gt;Columns</b> - Menu of columns for the sequence table.
<ul>
<li>Select or deselect columns.
<li>The columns are not immediately changed for existing tables. Either select "Refresh All Existing Tables"
or redisplay the table (listed on the left panel) and select "Refresh columns".
</ul>

<b>&gt;&gt;Filters</b> - A menu with specific filters for sequences.
<ul>
<li>Select one or more filters, then select "View Filtered Sequences". The sequences that pass all filters will
be displayed in a table. 
<li>If you are not seeing the columns you desire, go to the "Columns", change
the columns, come back to the table (it will be listed on the left), and click "Refresh Columns".
</ul>

<b>&gt;&gt;Results</b> - List all Filter query results.
<ul>
<li>All queries that have been performed will be listed in a table. 
<li>Selecting a label will redisplay it on the right, or select it from this Results table.
<li>To remove a label, select the 'x' beside it on the left panel, or remove it from the Results
table.
</ul>

<h2>Basic Queries</h2>
<b>&gt;&gt;Sequences</b> - Search on Seq ID, TCW Remarks or User Remarks.  
<ul>
<li> 
This builds a table from the results of the selected filters.
<li>The resulting sequences can be shown as "View Selected Sequence". 
<li>Columns can be changed with "Select Columns". 

</ul>

<b>&gt;&gt;AnnoDB Hits</b> - Search on the Hit Id (e.g. UniProt ID) or description. 
<ul>
<li>It works like the basic Sequences.
</ul>

<b>&gt;&gt;GO Annotation</b> - Search on the GO ID or description. 
<ul>
<li>It works like the basic Sequences.
</ul>


<hr>
<h2>Similar Sequence Pairs</h2>
This section will only exist if the database was built with the <b>Similar Pairs</b> option.
All options are the same as above, except apply to sequence pairs. 

</BODY>
</HTML>
