<!DOCTYPE html>
<html>
<!--#set var=TITLE value="Annotation Setup" -->
<!--#include virtual="../ssi/head.html" -->
<body bgcolor="#FFFFFF" text="#000000" onLoad="init('set')">
<!--#include virtual="../ssi/start2.html" -->

<!--------------- Body of document follows ------------------------------->

<a name=top>

<p>
To prepare for annotation with <code>runSingleTCW</code>, it is necessary
to download the databases to compared against. TCW provides support
for downloading the taxonomic and full UniProts
and mapping from the UniProt IDs to GO, KEGG, Pfam, EC, and InterPro.
<p><b>Tested</b>:
<code>runAS</code> has been tested on Linux, MacOS 10.9 and 10.15.  
If you have any problems, please let me know at <tt>tcw at agcol.arizona.edu</tt>.

<p><b>Contents:</b>
<ul>
 <li><a href="#over">Overview</a>
  <br>&nbsp;
 <li>Annotation Setup
 <ul>
 <li><a href="#run">Using Java graphical interface -- <code>runAS</code>
 
  <li><a href="#clean">Cleanup</a>
 <li><a href="#mem">Memory and time</a>
 </ul>

 <li><a href="#demo">Demo annotation setup</a>
 <br>&nbsp;
 <li><a href="#tax">Why use taxonomic databases instead of the full UniProt</a>
  <br>&nbsp;
 <li><a href="#other">Using other databases (e.g. NCBI-nr) for annotation</a>
  <br>&nbsp;
 <li><a href="#runstcw">Entering this data into <code>runSingleTCW</code></a>
</ul> 

<a name=over></a>
<table width='100%'><tr><td align='left' >	
<h2>Overview</h2>
</td><td align='right' valign='top' class='smtxt'><a href=#top>Go to top</a></td></tr></table>
<b><i>Terminology</i></b>: 
<p>The term "AnnoDB" refers to any database that will be used for annotation,
i.e. the sequences in TCW will be searched against all AnnoDB databases and the hits stored
in the TCW database for query. For a TCW created with nucleotide sequences, an annoDB may be nucleotide or protein. For
a TCW created with protein sequences, an annoDB can only be protein. 
<p>If multiple species are to be compared using <code>multiTCW</code>, it is important they all
use the same set of AnnoDBs and GO database.

<p><b><i>Requirements</i></b>: 

<p>The setup uses <code>curl</code> for downloading annoDBs and the GO database.
Note -- if you do not have <code>curl</code>, <code>runAS</code> will still work, though may not be
as reliable.

<p>For mySQL, the command <tt>mysqladmin</tt> is used, so
you may need to define its path, e.g. on Mac, 
<pre>
	alias mysqladmin '/usr/local/mysql/bin/mysqladmin'    #tcsh
	alias mysqladmin='/usr/local/mysql/bin/mysqladmin'    #bash
</pre>

<p>With MySQL v8, to add the GO database, you may need to set following MySQL variable:
<pre>
	SET GLOBAL local_infile = 1;
</pre>

<p><b><i>Processing steps</b></i>:
<ol>
<li>Create the directory for the downloads and generated FASTA files. 
<br>The default location is the TCW sub-directory <tt>projects/DBfasta</tt>, but it can be put elsewhere.
<li>Download Taxonomic UniProts ".dat" files and create FASTA files. 
<li>Download full UniProt, remove entries from taxonomic databases, and create FASTA files.
<li>Create GO database, which contains mappings from UniProt IDs to GO, KEGG, EC, Pfam and InterPro.
<ol type='a'>
<li>Download  <tt>go_&lt;date&gt;-termdb-tables.tar.gz</tt>.
<li>Create a local mySQL GO database with the information from this tar file. 
<li>Add information to the local GO database from the .fasta and .dat files in the UniProt directory.
</ol>
</ol>

<p>Other databases can be used for annotation, as discussed in <a href=#other>using other databases</a>,
but they will not have GO, KEGG, EC, Pfam, or InterPro support.
<!------------------------------------------------------------->

<a name=run></a>
<table width='100%'><tr><td align='left' >	
<h2>Java graphical interface -- <code>runAS</code></h2>
</td><td align='right' valign='top' class='smtxt'><a href=#top>Go to top</a></td></tr></table>

<table>
<tr><td>
Typically, all you need to do is make sure you have an internet connection open
and that you have enough disk space (see <a href=#mem>Memory</a>),
then:
<ol>
<li>The "TCW Annotation Directories" define where the files will be put (see <a href=#details>Details</a> below);
TCW provides defaults as shown on the right.
<br>&nbsp;
<li>Select the taxonomic databases you want to use, then select <tt class=c>Build Tax</tt>,
which downloads the respective files and creates FASTA files.
<br>&nbsp;
<li>Select the full databases you want to use, then select <tt class=c>Build Full</tt>,
which downloads the respective files, creates a subset of the full by removing
all entries found in the downloaded taxonomic databases, and creates a FASTA file.
<br>&nbsp;
<li>Select <tt class=c>Build GO</tt>, which downloads the GO database,
creates a local mySQL GO database with a mapping of the UniProts from your downloaded
set. This uses the information in <tt>HOSTS.cfg</tt>. 
See the <a href=runAS.log.html class="ext" target="_blank">log file</a> for the UniProt and GO processing.
<br>&nbsp;
<li>Select <tt class=c>TCW.anno</tt>, which writes a file <tt>projects/TCW.anno.&lt;date&gt;</tt>.
<br>&nbsp;
<li>Run <tt>./runSingleTCW</tt>.
<ul>
<li>Select <tt class=c>Import Anno</tt>, a file chooser will popup, select 
<tt>projects/TCW.anno.&lt;date&gt;</tt>.
<li>Select <tt class=c>Options</tt> in the annotation section, and select the GO database.
<li>Now you are ready to <tt class=c>Exec Annotate Sequences</tt> with the UniProt and GO you 
just downloaded.
</ul>
</ol>
The TrEMBL Bacteria and Full databases are 
<font color=red>very very large</font> (41GB and 61GB as of
Nov 2017), so do not download these unless your really really need them. Also, <i>they take a lot of memory
to process</i>.  

<td><a href="img/runAS.png"><img src="img/runAS.png" width=400 border=1>
</table>

<font color=green><b>Note:</b></font> I have not downloaded the Full TrEMBL UniProt in years; I find it sufficient to
download the SwissProt of all relevant taxonomies and the Full SwissProt database, as these are the best annotation;
then I include the TrEMBL of the most relevant taxonomies.
<p><tt class=c>Check</tt>: Selecting this button highlights everything that has been done.
For example, the figure on the upper right shows that the directory
<tt>UniProt_Jun2020</tt> has been created and only Archaea SwissProt has been downloaded and processed. 
The <tt class=c>Check</tt> automatically runs on startup.

<table>
<tr><td>
The download uses <code>curl</code>. 
You can get this on most Linux machines with 'sudo yum install curl'. 
If you cannot install it, <code>runAS</code> will prompt you as shown on
the right; if you select <tt>Continue</tt>, it will perform the download with its own Java code, 
though it may take longer and is not as robust, i.e. could have potential problems due to network latency, etc.
<td><a href="img/runAScurl.png"><img src="img/runAScurl.png" width=450 border=1>
</table>

<a name=details>
<h4>Details</h4>
<p>The rest of this section provides details:

<p>The default directory for annoDBs is <tt>projects/DBfasta</tt>.
<p><tt class=c>UniProt</tt> directory is where the UniProt files will be downloaded. If the
directory does not exist, it will be created. As shown in the image above, the default
name is <tt>UniProt_&lt;date&gt;</tt>, though it does not have to be this name.
<p><tt class=c>GO</tt> directory is where the GO file will be downloaded. If the directory
does not exist, it will be created. As shown in the image above, the default
name is <tt>go_tmp&lt;date&gt;</tt>, though it does not have to be this name. 
<p><tt class=c>Swiss</tt> and <tt class=c>TrEMBL</tt> headings indicate that selecting
a check box under <tt>Swiss</tt> will download the SwissProt database for the corresponding
taxonomic database (label to the right). Similarly, selecting the check box 
under <tt>TrEMBL</tt> will download the TrEMBL taxonomic database.
<p><tt class=c>Full UniProt</tt> will download the full SwissProt or TrEMBL database.
It expects that you will have downloaded at least one taxonomic database, i.e. the one
that corresponds with your species.

<p><tt class=c>GO (Gene Ontology)</tt>: a tar file containing the schema and data is downloaded
from <a href=http://archive.geneontology.org><tt>http://archive.geneontology.org/</tt></a>. 
The most current file is go_daily-termdb-tables.tar.gz

<p><tt class=c>Database</tt>: This is the name of the GO database that will be created; you will enter
this name in <code>runSingleTCW</code>.

<table>
<tr><td><b><code>runAS</code> will not replace an existing downloaded file</b>. If you select <tt>Continue</tt>
on the prompt on the right, it will skip the download but perform the rest of the processing. This
is necessary if you have (1) run <tt class=c>Build Full</tt> to create the subset, or (2)
run <tt class=c>Build GO</tt> to create the GO database, but then download another Taxonomic UniProt;
<i>these two steps need to be re-run</i>.
<td><a href="img/runASnoDL.png"><img src="img/runASnoDL.png" width=450 border=1>
</table>
<p>For each of the three download steps, there will be an initial prompt to ensure that you meant to 
select the download.


<!------------------------------------------------------------------------>
<a name=clean></a>
<table width='100%'><tr><td align='left' >
<h2>Clean up</h2>
</td><td align='right' valign='top' class='smtxt'><a href=#top>Go to top</a></td></tr></table>

<code>runAS</code> does not remove files that are no longer necessary, which are
the files downloaded from the internet:
<ul>
<li>All "dat.gz" files in the UniProt directories, as the information has
been transferred to the FASTA files and GO database. 
<li>The GO directory, as the information has been transferred to the GO database. 
</ul>
These files can be removed, as <code>runSingleTCW</code>
uses the FASTA files in the UniProt directories and the GO mySQL database. However, if you
do not have a space problem, keep them just for insurance. 
When UniProt does the monthly update, your downloaded files will no longer available
on their site. 
<p>For the FASTA files that you will be using 
<code>DIAMOND</code> to search against, you 
can gzip them as <code>DIAMOND</code> can search against gzipped files.

<!------------------------------------------------------------------------>
<a name=mem></a>
<table width='100%'><tr><td align='left' >
<h2>Memory and Time</h2>
</td><td align='right' valign='top' class='smtxt'><a href=#top>Go to top</a></td></tr></table>
To view the current sizes of the UniProt taxonomic divisions and complete database:
<ul>
<li>Go to <a href=https://www.uniprot.org/downloads class="ext" target="_blank">UniProt Downloads</a>.
<li>In the second line from the top, it says "For downloading complete data sets we recommend using <b>ftp.uniprot.org</b>."
Click the <b>ftp.uniprot.org</b>. 
<li>This brings up the UniProt download directories in a Finder window. You may view it as "Guest".
<li>Click "Current_release", "knowledgebase". Here you will see "complete" and "taxonomic_divisions".
<li>View these two directories along with the sizes.
</ul>

When TCW extracts the sequence into a FASTA file, it is not written in a gzipped format.

<p>Strong suggestions:
<ul>
<li><font color=red>Only download what is relevant!</font> 
<br>For example, download both SwissProt and TrEMBL taxonomic databases
for your species of interest, plus any related taxonomies, then only download the full SwissProt.
<li><i>Do not perform constant downloads, it is a drain on the UniProt servers.</i> 
<br>The UniProts do not change that fast, and it changes 'best' hits in TCW, which can disturb any on-going analysis.
</ul>

<p><b>Times</b>: The following shows the times for downloading 2 SwissProt and 1 TrEMBL files 
using a ~500 Mbsp download connection and 128Gb of RAM. 
<pre>
	curl complete ./projects/DBfasta/UniProt_Aug2019/sp_fungi/uniprot_sprot_fungi.dat.gz      0m:5s
	curl complete ./projects/DBfasta/UniProt_Aug2019/sp_plants/uniprot_sprot_plants.dat.gz    0m:6s
	curl complete ./projects/DBfasta/UniProt_Aug2019/tr_plants/uniprot_trembl_plants.dat.gz   8m:24s

	34,106 written to uniprot_sprot_fungi.fasta                     0m:4s  (3Mb)
	42,597 written to uniprot_sprot_plants.fasta                    0m:4s  (3Mb)
	9,534,192 written to uniprot_trembl_plants.fasta                9m:25s (3Mb)
</pre>
<p><b>Time for creating GO database:</b> The following is the times for build the GO database for 4 SwissProt, Full SwissProt, and 2 TrEMBL.
<pre>
	Complete creating GO database go_Aug2019                        32m:36s (10Mb)
</pre>
NOTE: It takes much longer if you have many large TrEMBL taxononic databases downloaded, where it may
take over night.

<!--------------------------------------------->
<a name=demo></a>
<table width='100%'><tr><td align='left' >
<h2>Demo annotation setup</h2>
</td><td align='right' valign='top' class='smtxt'><a href=#top >Go to top</a></td></tr></table>
In order to add GO annotations, a local GO mySQL 'demo' database needs to be created.

<table>
<tr><td valign="top">
<ol type="A">
<li>From the TCW_2 directory, execute:
<pre class=tt>
./runAS -d
</pre>
The "-d" will cause it to enter the demo parameters, as shown on the right.
The highlighted entries already exist. Its only necessary to build the GO database,
which takes about 10 minutes. 
<br>&nbsp;
<br>&nbsp;
<br>&nbsp;
<li>Execute <tt class=c>Build GO</tt>.
<br>Enter "Confirm" on the first prompt to continue.
<br>The GO tables are available for the demo, i.e. they will not be downloaded. This is evident from the purple GO label.
<br>Building the GO database takes anywhere from 10 to 60 minutes.
</ol>
<td><a href="img/runASdemo.png"><img src="img/runASdemo.png" width=400 border=1>
</table>


<h4>Details about the Demo setup</h4>
In the <tt>projects/DBfasta</tt> directory, there is the sub-directory <tt>UniProt_demo</tt> and <tt>GO_tmpdemo</tt>, 
which contains the following:
<pre>
	GO_tmpdemo:
	go_daily-termdb-tables.tar.gz

	UniProt_demo:
	sp_bacteria/	sp_fungi/	   sp_plants/	      tr_plants/
	sp_fullSubset/	sp_invertebrates/  tr_invertebrates/
</pre>
Each taxonomic directory has a <tt>.dat</tt> and a <tt>.fasta</tt> file, 
which are very small subsets of the original UniProt taxonomic <tt>.dat</tt> file. 
<p>
See <a href=#over>Overview</a> for more details.


<a name=tax></a>
<table width='100%'><tr><td align='left' >
<h2>Why use taxonomic databases instead of the full UniProt</h2>
</td><td align='right' valign='top' class='smtxt'><a href=#top >Go to top</a></td></tr></table>

<ol>
<li>TCW refers to the annoDBs by the 'type' and 'taxonomy', e.g. sp_fungi has 
a type of 'sp' and a taxonomy of 'fungi'. Within TCW, you can query by taxonomy.
Also, all results will show this information so you can easily see which taxonomic database a hit is from.
The following shows an example:
<br><a  href="img/runASview.png"><img src="img/runASview.png" border=1></a>

<li>You only need annotate against the taxonomies of interest. The bacterial
TrEMBL is especially large and should be left out if not specifically desired.  

<li>The complete UniProt has everything but bacteria. There is an option to download the complete UniProt, then make
a subset to search against that has everything but what is in the taxonomic databases. This will be 
referred to as the "Full" database below. In the above figure, "SPful" refers to the SwissProt full subset
database and "TRful" refers to the TrEMBL full subset.
</ol>

<a name=other></a>	
<table width='100%'><tr><td align='left' >
<h2>Using other databases for annotation</h2>
</td><td align='right' valign='top' class='smtxt'><a href=#top >Go to top</a></td></tr></table>

NCBI-nr works with TCW. For other databases, you will need to make sure they have a TCW
accepted descriptor line. 

<h4>Description lines</h4> 
The description line is the ">" line that describes the subsequent sequence in a FASTA file. 
From it, execAnno extracts 
<ul>
<li><tt>DB type</tt>: used in naming the blast output file and is used in <code>viewSingleTCW</code> to aid
in identifying where the hitID is from.
<li><tt>hitID</tt>: the unique identifier of the hit.
<li><tt>description</tt>: generally the functional description
<li><tt>species</tt>: the species 
</ul>

<h4>UniProt:</h4>
<pre>
	>sp|Q9V2L2|1A1D_PYRAB Putative 1-ami OS=Pyrococcus abyssi GN=PYRAB00630 PE=3 SV=1v
</pre>
<ul>
<li>For TrEMBL, the first two characters would be 'tr'. The 'sp' or 'tr' are the <tt>DB type</tt>
<li>The third entry of the first string is the identifier (e.g. 1A1D_PYRAB)
<li>The string up to the OS is the description.
<li>The string after the "OS=" is the species.
</ul>

<h4>NCBI nr</h4>
The NCBI-nr database can be downloaded using the following link:
<a href="ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz" class="ext">NCBI nr<a> 
(ftp.ncbi.nih.gov/blast/db/FASTA).
Zipped, this is 32Gb as of Nov2017. 
Using the <code>DIAMOND</code> program can search this database quickly; 
however, there will be no GO, KEGG, EC, or Pfam information associated with these hits. 

<p>In the summer of 2016, NCBI changed the format of the subject line. TCW can parse the old or new format,
though the recent NCBI-nr file (Oct 2016) has some badly formed entries, which are ignored. The descriptor line
may have multiple entries separated by "CNTL+A"; only the first entry is used.
<p>

<p>New format:
<pre>
	>XP_642837.1 hypothetical protein DDB_G0276911 [Dictyostelium discoideum AX4]
</pre>
<ul>
<li>The first entry is the identifier (e.g. XP_642837.1). 
Note, there is no longer a way to detect the database origin within the file, hence, the <tt>DB type</tt> will be the generic 'PR' for protein.
<li>The text from the first space to the first "[" is the description.
<li>The text within the "[]" is the species.
</ul>
<br>Old format:
<pre>
	>gi|66818355|ref|XP_642837.1| hypothetical protein DDB_G0276911 [Dictyostelium discoideum AX4]
</pre>
<ul>
<li>The 'gi' is the <tt>DB type</tt>.
<li>The fourth entry of the first string is the identifier (e.g. XP_642837.1)
<li>The text from the last "|" to the first "[" is the description.
<li>The text within the "[]" is the species.
</ul>

<h4>Generic:</h4> 
If you have a file other than UniProt or nr, make the descriptor names as follows:
<pre>
	>CC|ID description OS=species
</pre>
<ul>
<li>CC is the type code, and will be used as the <tt>DB type</tt> in TCW.
<li>ID is the unique identifier
<li>Everything up to the OS is the description
<li>Everything after the OS is the species
</ul>
You want the CC type code + taxonomy (entered through the <code>runSingleTCW</code> interface)
 to be unique for each annoDB. The type code + first three letters of the taxonomy are used to
name the blast output, and is also used in <code>viewSingleTCW</code> to easily determine
what annoDB a hit came from.

<a name=runstcw></a>
<table width='100%'><tr><td align='left'>
<h2>Entering this data into <code>runSingleTCW</code></h2>
</td><td align='right' valign='top' class='smtxt'><a href=#top>Go to top</a></td></tr></table>

<a  href="img/runSingleStep3.png"><img src="img/runSingleStep3.png" border=1></a> 
<p>The AnnoDBs can be entered using the "Add" button, where the taxonomy is defined.
Alternatively, use the "Import AnnoDBs" to add the databases from an existing <tt>sTCW.cfg</tt>
file or from the <tt>TCW.anno</tt> file created from selecting "TCW.anno" on <code>runAS</code>. Note,
the full path of the AnnoDBs is not shown in this image. 
<p>The GO database is defined in the "Options" menu.



<p><a href="#top">Go to Top</a>

<!--#include virtual="../ssi/end.html" -->	
</html> 

