<!DOCTYPE html>
<HTML>
<!--#set var=TITLE value="sTCW Differential Expression Guide" -->
<!--#include virtual="../ssi/head.html" -->
<body onLoad="init('de')">
<a id=top></a>
<!--#include virtual="../ssi/start2.html" -->

<ttp>RunDE</ttp> computes differential expression (DE) between conditions for an existing singleTCW database,
as follows:
<ol>
<li>DE per sequence: R scripts for <ttp>EdgeR</ttp>, <ttp>DESeq2</ttp>. 
<ul>
<li>It is <i>important</i> that you study the documentation for the respective DE method to determine the
best approach for your data. You may either use one of the existing scripts,
customize an existing script, or create a new one. 
<li>The <a href="#scripts">R-scripts for DE</a> sections explains how <ttp>runDE</ttp> writes the data to R variables, 
executes the R-script, then reads the results into sTCWdb.
</ul>

<li>GO enhancement: R script for <ttp>GOseq</ttp>. 
<ul>
<li>Same comments above. See <a href="#scripts">R-scripts for GO</a>.
</ul>
</ol>
This document uses the "<ttx>demoTra</ttx>" project as an example. The
<a href="UserGuide.html" class="ext" target="_blank">User Guide</a> explains how to built it.

<p><b>Contents:</b>
<table style="width:50%">
<tr><td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<td class="top">
<ol type="1">
    <li><a href="#install">Installing R and packages</a>
    <li><a href="#demo">Example: demoTra</a>
    	<ul>
		<li><a href="#execDE">Execute DE output</a>
		<li><a href="#execGO">Execute GO output</a>
		</ul>
    <li><a href="#libDE">DE between conditions</a>
	<ol type="A">
	<li><a href="#methods">Methods and Options</a>
		<ul>
		<li><a href="#script1">R-script</a>
		<li><a href="#reps">Use of replicates</a>
		<li><a href="#filter">Pre-filtering</a>
		<li><a href="#disp">Fixed dispersion</a>
		</ul>
	<li><a href="#exec">Execute</a>
		<ul>
		<li><a href="#all">All Pairs for Group 1</a>
		<li><a href="#groups">Group 1 - Group 2</a>
		<li><a href="#file">All Pairs from File</a>
		<li><a href="#pfile">P-values from File</a>
		</ul>
	</ol>
    <li><a href="#go">GO enhancement</a>
    
    <li><a href="#rm">Remove</a>
   
    <li><a href="#rscript">R-scripts</a>
     <ul>
     <li><a href="#DEscript">R-scripts for DE</a>
    <li><a href="#GOscripts">R-scripts for GO</a>
    <li><a href="#fname">Naming R-scripts</a>
    </ul>
    <li><a href="#Rperms">R directory permissions</a>
</ol>
</table>

<a id="install"></a>
<h2>1. Installing R and packages</h2>

<h4>Installing R</h4>
<ul>
<li>See <a href="https://www.r-project.org" class="ext" target="_blank">R Project</a> for the latest instructions on installing R.
<li>See <a href="../Mac-OSX.html" class="ext" target="_blank">MacOS</a> for installation that worked on MacOS 10.15.
<li>See <a href="#Rperms">R directory permissions</a> for some permission suggestions.
</ul>

<h4>Installing rJava and R packages</h4>
<ul>
<li>See <a href="../Install.html#R" class="ext" target="_blank">rJava and R packages</a>.
</ul>

<a id="demo"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>2. Example: demoTra</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="tablex">
<tr>
	<td class="top" >
	From the command line, type:
	<pre>
    >runDE tra
	</pre>
	Or, if you type  <ttp>runDE</ttp> with no arguments, the sTCW graphical chooser will appear
	with all sTCW databases listed; select the <ttx>demoTra</ttx> database. 
	Either way, this brings up the <ttp>runDE</ttp> interface shown at right.
	<p>
	Define a pairwise comparison by selecting "Stem" for <ttl>Group 1</ttl> and "Root" for <ttl>Group 2</ttl>, as shown. 
	<p>
	Check the box <ttl>Save results in p-value column</ttl>; the text box will display the column name "StRo", which can be changed.
	<p>
	Assuming you have installed <ttp>edgeR</ttp>, leave the 
	<ttl>Differential Expression R-script</ttl> as <ttl>edgeRglm.R</ttl> . 
	<p>
	Execute <ttc>Group 1 - Group 2</ttc>. See <a href=execGO>Execute DE</a> for the output.
	
	<p>The <ttl>GO enrichment R-script</ttl> will be <ttc>goSeqBH.R</ttc> by default, which
	performs BH multiple hypothesis testing correction. However, 
	this results in 0 p-value&lt;0.05 for the demo,
	hence, <i>for the demo only</i>, change this to <ttc>goSeqNoFDR.R</ttc>. Then select
	<ttc>Execute</ttc>. See <a href=execGO>Execute GO</a> for the output.
	 
</td>
	<td class="top">
		<a href="img/DErun.png"><img src="img/DErun.png" alt="" style="border: 1px solid black; width: 500px"></a>
	</td>
</tr>
</table>

<a id="execDE"></a>
<h3>Execute DE output</h3>
<pre>
> ****************************************************************
******** Start DE execution for column: StRo *********
****************************************************************
Replicates:
      5 Stem
      5 Root
Collecting count data (may take several minutes)
Using CPM filter > 1 for >= 2
      1 filtered sequences
Assigning R variables
   gc: GC values of sequences
   rowNames: sequences (row) names
   grpNames: group (column) names
   repNames: replicate names
   counts: counts of sequences
   countData <- array(counts,dim=c(210,10))
   rm(counts)
   rownames(countData) <- rowNames
   nGroup1 <- 5
   nGroup2 <- 5

Start R-script
   source('R-scripts/edgeRglm.R')
Loading required package: limma
Using traditional glm (quasi-likelihood F-tests)
   results
   rowNames
R-script done

Number of DE results:
   <1e-5 <1e-4 <0.001 <0.01 <0.05
       6    17    32    65    97

Saving 210 scores for StRo
   Adding column to database...
                            .
Finished DE execution for StRo             0m:0s  (8Mb)
Complete all Group1-Group2 for tra
The console is in R, you may run R commands -- q() or Cntl-C when done, or perform another Execute.</pre>
</pre>	
The R console remains open and you can now manually explore the data using R, if desired. However, it easiest to explore it be
selecting "Close" on <ttp>runDE</ttp>, which closed the java program but not R. Then do the following:
<pre>
	>q()
	Save workspace image? [y/n/c]: y
	R
</pre>
This saves all variables written by <ttp>runDE</ttp> and the scripts (in this case <ttp>edgeRglm.R</ttp>). By restarting R, you can now
explore the data, e.g.
<pre>
	> library(edgeR)
	> plotBCV(y)
	> plotMDS(y, labels=grpNames)
</pre>

<h4>View DE Results</h4>
<table class="tablex">
<tr>
	<td>
	To view the resulting p-values, run '<ttx>viewSingleTCW tra</ttx>' from the command line,  
	and click <ttc>&gt;&gt; Columns</ttc>. The column options has a section 
	<ttx>Differential Expression</ttx> with the DE columns listed, e.g. the DE columns for <ttx>demoTra</ttx> are shown on the right. There is also
	a section for <ttx>Conditions</ttx>, which has the "Stem", "Root, and "Oleaf" column options. Select the desired columns (
	e.g. "Stem", "Root", "StRo"); you may then select a filter (e.g. all sequences that have an absolute p-value &lt; 0.05).

	Note in the below figure that if Stem &lt; Root, the p-value StRo is negative, else it is positive. 
	<td>
		<ttp>viewSingleTCW</ttp> DE section with all pairs computed.
		<a href="img/DEcolumn.png"><img src="img/DEcolumn.png" alt="" style="border: 1px solid black; width: 250px; padding-left:20px"></a>
</tr>
<tr><td colspan=2>
<a href="img/DEtable.png"><img src="img/DEtable.png" alt="" style="border: 1px solid black; width: 700px"></a>
</table>

<a id="execGO"></a>
<h3>Execute GO output</h3>
<pre>
****************************************************************
******** 1. Start GO enrichment for column: StRo *********
****************************************************************
Assigning R variables
   208  Sequences with GOs (from 208)
   seqNames: sequence names
   seqLens:  sequence lengths
   nSeqs <- 208
   seqDEs: DE binary vector (97 seq p-value < 0.05)
   names(seqDEs) <- seqNames
   seqGOs <- vector("list",nSeqs)
   For all n: seqGOs[[n]] <- c(gonum list)
   names( seqGOs) =  seqNames                       ...

Start R-script
   source('R-scripts/goSeqNoFDR.R')
Using manually entered categories.
Calculating the p-values...
   oResults
   goNums
R-script done

Saving 6651 values to database
   Adding column to go_info table
    311 GO p-values < 0.05
Finished GO enrichment for StRo                                          0m:2s
Complete GO enrichment for tra
The console is in R, you may run R commands -- q() or Cntl-C when done, or perform another Execute.
</pre>
The R console remains open and you can now manually explore the data using R, if desired. However, it easiest to explore it be
selecting "Close" on <ttp>runDE</ttp>, which closed the java program but not R. Then do the following:
<pre>
	>q()
	Save workspace image? [y/n/c]: y
	R
</pre>
This saves all variables written by <ttp>runDE</ttp> and the scripts (in this case <ttp>goSeqNoFDR.R</ttp>). By restarting R, you can now
explore the data.

<h4>View GO Enhancement Results</h4>
<table class="tablex">
<tr>
	<td>
</table>
	
<!--  ========= DE  ==============   -->
<a id="libDE"></a>
<h2>3. DE between conditions</h2>

<a id="methods"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h3>Methods & Options</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a id="script1"></a>
<h4>R-script</h4>
The default R-script is <ttc>edgeR.R</ttc>, however, there are other scripts in the <ttx>R-script</ttx>, or the
user can supply there own (see <a href="#DEscript">R-script for DE</a>).

<a id="reps"></a>
<h4>Use of replicates</h4>

Biological replicates are valuable for RNA-seq to provide the baseline estimate of variability.
This is incorporated into the models (typically as dispersion of the negative-binomial distribution)
and makes a big difference in the result DE scores. 
<p>
If replicates are provided for a condition, they will automatically be used for the applicable
R packages (e.g., <ttp>edgeR</ttp>). All of the methods also work without replicates, but are likely to
overestimate DE. 
<p>
Replicates should be biological replicates, i.e. sequencing from distinct individuals having the
same treatments, rather than technical replicates, i.e. re-sequencing of the same individuals. 
Technical replicates can be input to TCW and will be used in the same way, but do not provide
a useful estimate of variability. 

<a id="filter"></a>
<h4>Pre-filter</h4>
This example uses a database with 26k sequences and 5 replicates for root and stem.
The three filtering options were run with defaults (table on lower left)
before input to <ttp>edgeR</ttp> (results in figure on lower right). 
The first option filtered 1713 sequences and the second option filtered 7518; 
with <ttp>edgeR</ttp>, the number of analyzed sequences effects the resulting DE value.

<table>
<tr>
	<td style="width: 50%; vertical-align: top;">
		<table class="tabley">
			<tr><td><i>Filter</i><td><i>Description</i><td><i>Column</i><sup>1</sup>
			</tr>
			<tr><td>Count >[n]
				<td>Filter all sequences that do not have any replicates with a count &gt; n (default 1) 
				<td>RoSt1
			</tr>
			<tr><td class="nw">CPM<sup>2</sup> >=[n] for >[m]
				<td>Filter all sequences that do not have more than m replicates with CPM >= n (default 2 and 1)
				<td>RoSt2
			</tr>
			<tr><td>None<td> No filtering <td>RoSt3
			</tr>		
		</table>
		<small>
		<sup>1</sup> Column in table on right.
		<br><sup>2</sup> CPM (Counts per million).
		</small>
	<td style="width: 50%; vertical-align: top; padding-left:10px;">
		<a href="img/DEfilter.png"><img src="img/DEfilter.png" alt="" style="width:400px; border: 1px solid black; "></a>	
		<small>
		<br>The DE value (RoSt) is set to negative if Root&lt;Stem.
		<br>The columns prefixed with '#' are the counts, which are used by the filters.
		<br>The two columns at the end (Root and Stem) are the RPKM values. 
		</small>
</table>

In the figure on the right:
<ol>
	<li>DE value 3.00: the filtered sequences have a value of 3.
	<li>DE value 2.00: the sequences that received an NA from the DE computation have a value of 2.
</ol>
  
<p>
To view the first 50 filtered sequences that do not have 'all' zero count values, 
execute <ttx>runDE</ttx> with the ID or database name and a -v, 
<br>e.g. <ttx>runDE NnR -v</ttx>.
<p>
Using Count > 1, the following are the Counts of the filtered sequences.
<pre>
	NnR_00003    0    0    0    1    0    0    0    0    0    0
	NnR_00016    0    0    0    0    0    0    1    0    0    0
	NnR_00032    0    0    0    0    0    1    0    0    0    0
	NnR_00090    1    0    0    1    0    1    0    0    0    0
	NnR_00095    1    1    1    1    0    1    0    0    0    0
	NnR_00112    0    0    0    1    0    0    0    0    0    0
	...
	1713 filtered sequences
</pre>
Using CPM >=2 for >1 replicates, the following are the Counts:CPM values of the filtered sequences.
<pre style="font-size: 12px">
	NnR_00002   6   10   13    8   21   20   16    9   13    3  :  0.2  0.4  0.4  0.3  0.6  0.8  0.5  0.3  0.4  0.1
	NnR_00003   0    0    0    1    0    0    0    0    0    0  :  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
	NnR_00004  19   27   18   15   17   16   17    4   22   10  :  0.8  1.0  0.6  0.5  0.5  0.6  0.5  0.1  0.7  0.3
	NnR_00007  34   23   29   29   31   12    6    1    7    2  :  1.4  0.9  0.9  1.0  0.9  0.5  0.2  0.0  0.2  0.1
	NnR_00009  45   23   33   22   46   17   26    7   22    3  :  1.8  0.9  1.1  0.8  1.3  0.6  0.8  0.3  0.7  0.1
	NnR_00010  19   16   29   23   26    5    4    3    4    1  :  0.8  0.6  0.9  0.8  0.7  0.2  0.1  0.1  0.1  0.0
	NnR_00014  28   21   38   26   31   13   13   15   14   11  :  1.1  0.8  1.2  0.9  0.9  0.5  0.4  0.5  0.5  0.4
	NnR_00016   0    0    0    0    0    0    1    0    0    0  :  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
	NnR_00017   5    4    4   10   11    0    0    2    1    0  :  0.2  0.2  0.1  0.4  0.3  0.0  0.0  0.1  0.0  0.0
	...
	7518 filtered sequences
</pre>

<a id="disp"></a>
<h4>Fixed dispersion</h4>
If there are not biological replicates, then it is necessary to set the dispersion. If it is not set, TCW will use 0.1 by default.

<a id="exec"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h3>Execute</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

The most meaningful DE comparison is between exactly two conditions, as illustrated in the example above. 
Groups of more than one condition may also be selected, however 
the conditions in each group will be treated as replicates, leading (probably) to an overestimation
of biological variability and underestimation of DE. 
<p>
If multi-condition testing is desired, e.g. to find sequences reduced in condition A compared to
any or all of conditions B,C,D, then the recommended process is to compute the individual
pair DEs A-B, A-C, A-D, and then use filter options in <ttp>viewSingleTCW</ttp>
to perform the screen. 

<p>There are four options to add DE values to the sTCW database, as explained in the following four sections.
<a id="groups"></a>

<a id="all"></a>
<h4>All Pairs for Group 1</h4>
<ul>
	<li>Only check conditions in Group 1.
	<li>Results will be saved in the database with auto-generated column names.
	<li>Execute: Each condition selected in group 1 will be pairwise compared; any selections
	in Group 2 are ignored.
</ul>
For example, if all three conditions (Stem, Root, Oleaf) were selected, then it would 
compute:
<pre>
	Stem  Root   StRo
	Stem  Oleaf  StOl
	Root  Oleaf  RoOl
</pre>
where the last column is the generated column name.

<h4>Group 1 - Group 2</h4> 
<ul>
	<li>Select at least one condition from each group.
	<li>If you want to store the results in the database, check <ttl>Save results in P-value column</ttl>.
	If the groups have been selected, a column name will be auto-generated which you can change.
	When you change the groups for a subsequent execution, uncheck-then-check the column box
	to auto-generated a name for the new groups.
	<li>Execute: The condition(s) in Group 1 will be compared with the condition(s) in group 2.
</ul>

<a id="file"></a>
<h4>All Pairs from File</h4>
<ul>
	<li>Any checked conditions in <ttl>Group 1</ttl> and <ttl>Group 2</ttl> are ignored; the conditions to
	compare are specified in the file.
	<li>Results will be saved in the database with the provided column names.
	<li>Execute: All pairs from the file will be compared with the selected method
	and options.
</ul>

<i>File format:</i> Each line has 3 columns which represent Group 1, Group 2 and the column name to use, respectively. 
For example:
<pre>
	Root     Stem  RoSt
	Root     Oleaf RoOl
	Stem     Oleaf StmLea
	Root:Stem Oleaf RoSt_Ol
</pre>
The last row shows how to select two or more from Group 1; no spaces are allowed between the ":".
Group 2 can only have one condition.
<p>
Keep the column names short but meaningful. A file similar to the above is in the <ttx>projects/demoTra</ttx> directory,
filename <ttx>traDE.tsv</ttx> to allow user to try this option.

<a id="pfile"></a>
<h4>P-values from File</h4>
The first line defines the two groups and the p-value column name, using the same format as the "Get Pairs from File". All remaining lines have a seqID followed by the p-value. For example,
<pre>
	#Example p-value file 
	root stem RoSt
	tra_001	0.14
	tra_002	0.14
	tra_003	0.65
</pre>
The p-values are read as absolute values since runDE makes values negative if Cond1&lt;Cont2.

<!-- ========== GO ==================== -->
<a id="go"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>4. GO enrichment (over-represented categories)</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

If GO annotations are loaded (see <a href="UserGuide.html#anno" class="ext" target="_blank">User Guide</a>), you can
compute p-values for enrichment of DE in GO categories (terms/IDs). 
<ul>
<li><u>R-script</u>: Use the default <ttc>goSeqBH.R</ttc> or supply your own. 
<li><u>P-value columns</u>: If one or more DE p-values have been added to the database, 
the &lt;---&gt; box will say <u>All p-values</u>;
this can be changed to compute GO enrichment for just one p-value column.
<li><u>P-value cutoff</u>: A <ttl>p-value</ttl> of 0.05 is the default; this applies to the DE p-value so that 
all sequences with a p-value&lt;0.05 will be considered DE (which assumes FDR correction was applied).
Since the DE p-values still contain considerable
uncertainty as absolute measures of probability, there is an option to defined DE
in terms of ranking, i.e., top 10% (the ranking is ordered by the p-values).
</ul>The meaning of a low p-value for a given GO is that the sequences assigned the
GO have a higher percentage of DE than the sequences as a whole. For example, if GO:000123 is assigned to
1000 sequences, of which 500 are DE (50%), while only 25% of sequences as a whole are
DE, then GO:000123 will receive a very low p-value. Note that the direction of the DE for
each sequence (elevated or suppressed) is not relevant. 
<p>
The result is p-value columns for the GO, having
the same names as the corresponding sequence DE columns, accessible through the <ttc>Basic GO Query</ttc> function of 
<ttp>viewSingleTCW</ttp>.


<a id="rm"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>5. Remove</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<!-- ============= R-scripts ============== -->
<a id="rscripts"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>5. R-scripts</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<a id="DEscript"></a>
<h3>R-scripts for DE</h3>

When the <b>Execute</b> is selected, TCW performs the following:
<ol>
<li>Writes the necessary information to the R environment based on the options selected.
<li>Runs the R script using the <ttx>source</ttx> command.
<li>Reads the <ttx>results</ttx> variable from the R environment, which should contain the p-values.
</ol>
TCW writes the following variables to the R environment (the following assignments are from the above demo example):
<pre>
	> gc: GC values of sequences
		rowNames: sequences (row) names
		grpNames: group (column) names
		repNames: replicate names
		counts: counts of sequences
		countData &lt;- array(counts,dim=c(210,10))
		rm(counts)
		rownames(countData) &lt;- rowNames
		nGroup1 &lt;- 5
		nGroup2 &lt;- 5
</pre>
The variables displayed with a ":" after them are assigned with the <ttx>assign</ttx> function, whereas the others are the exact assignment.
<br>The most important variables are <ttx>countData</ttx> and <ttx>grpNames</ttx>, e.g.
<pre>
	> head(countData)
			[,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
	tra_001 1017  594 1222 1209 1315  378 1002 1649  826  1195
	tra_002  272  239  431  400  368  101  206  151  109   185
	tra_003 3830 5185 4847 4857 5451 1859 2506 1334 1541  2307
	tra_004 1707 1088 2429 2210 2334  529  919 1103  810  1427
	tra_005  479  369  439  444  565  114  192  151  232   351
	tra_006 1122  923 1381 1320 1482  632  839  905  670   823
	> grpNames
	 [1] "Stem" "Stem" "Stem" "Stem" "Stem" "Root" "Root" "Root" "Root" "Root"
</pre>
The R script should put the p-values in the <ttx>results</ttx> variables, e.g. 
<pre>
	> head(results)
	[1] 1.537634e-07 1.695981e-07 2.478791e-07 2.478791e-07 4.026519e-06
	[6] 4.600047e-06
</pre>
The order should correspond to the input order; if changed, then change the order in <ttx>rowNames</ttx> to correspond.
<p>An example script is:
<pre>
	# edgeR gln method unless no replicates
	library(edgeR)
	y &lt;- DGEList(counts=countData,group=grpNames)
	y &lt;- calcNormFactors(y)
	if (nGroup1==1 &amp;&amp; nGroup2==1) {
		writeLines("Using classic with fixed dispersion")
		et &lt;- exactTest(y, dispersion=disp)
		res &lt;- topTags(et, n=nrow(et), adjust.method="BH")
	} else {
		writeLines("Using traditional glm")
		design &lt;- model.matrix(~grpNames)
		y &lt;- estimateDisp(y, design)
		fit &lt;- glmQLFit(y,design)
		qlf &lt;- glmQLFTest(fit,coef=2)
		res &lt;- topTags(qlf, n=nrow(qlf), adjust.method="BH")
	} 
	# Columns are:  logFC    logCPM        F       PValue          FDR
	results &lt;- res$table$FDR
	rowNames &lt;- rownames(res)
</pre>

<a id="GOscript"></a>
<h3>R-scripts for GO</h3>
This works like the <a href=""#DEscript"">R-scripts for DE</a>, with the variables shown below in the GO enrichment output:
<pre>
****************************************************************
******** 1. Start GO enrichment for column: StRo *********
****************************************************************
Assigning R variables
   208  Sequences with GOs (from 208)
   seqNames: sequence names
   seqLens:  sequence lengths
   nSeqs <- 208
   seqDEs: DE binary vector (97 seq p-value < 0.05)
   names(seqDEs) <- seqNames
   seqGOs <- vector("list",nSeqs)
   For all n: seqGOs[[n]] <- c(gonum list)
   names( seqGOs) =  seqNames                       ...

Start R-script
   source('R-scripts/goSeqBH.R')
Using manually entered categories.
Calculating the p-values...
   oResults
   goNums
R-script done

Saving 6651 values to database
    0 GO p-values < 0.05
Finished GO enrichment for StRo                                          0m:2s
</pre>
See <tt>R-scripts/goSeqBH.R</tt> for an example. The p-values to be loaded must be in "oResults" and
the associated GO numbers are in "goNums" (note that TCW removes the "GO:" prefix and just used the number).

<h4>seqGOs</h4>
The GOs assigned to each sequence are both the direct and indirect.

<h4>FDR</h4>
According to the GOseq Manual: "Having performed the GO analysis, you may now wish to 
interpret the results. If you wish to identify categories significantly enriched/unenriched 
below some p-value cutoff, it is necessary to first apply some kind of multiple hypothesis 
testing correction. For example, GO categories over enriched using a .05 FDR cutoff 
[Benjamini and Hochberg, 1995] are:"
<pre>
> enriched.GO=GO.wall$category[p.adjust(GO.wall$over_represented_pvalue, + method="BH")<.05]
</pre>
Hence, the values loaded into "oResults" for the TCW supplied <ttc>GOseqBH.R</ttc>:
<pre>
oResults <- p.adjust(GO.wall$over_represented_pvalue, method="BH")
</pre>
The p.adjust methods are: c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY",  "fdr", "none")
<p>
<u>demoTra</u>: The TCW supplied demo result in 0 p-values&lt;0.05. 
In order to view some p-values&lt;0.05 results
for the demo, use <ttc>GOseqNoFDR.R</ttc>, which has no multiple testing correction.


<a id="fname"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h3>Naming R-scripts and Overview</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
If you use your own R-script, make sure the name reflects its function. The R-script names are
shown in the TCW <ttc>Overview</ttc>, as illustrated below:
<pre>
DE (Differential Expression) computation: 
      Column       Method               Conditions
      StRo         edgeRglm.R           Stem : Root 
      StOl         edgeRglm.R           Stem : Oleaf 
      RoOl         edgeRglm.R           Root : Oleaf 

   GO enrichment computation: 
      Column       Method               Cutoff
      StRo         goSeqNoFDR.R         0.0500
      StOl         goSeqNoFDR.R         0.0500
      RoOl         goSeqNoFDR.R         0.0500
</pre>

<a id="Rperms"></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>6. R Directory Permissions</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
The following was written in 2014, so may be dated. 

<p>Unless you plan to run as root (not recommended), 
or will be the only one running <ttp>runDE</ttp>, you will also want to set up R so that one user
can install packages to be used by all users. Note: you
do NOT need R to view the DE results with <ttp>viewSingleTCW</ttp>.

<p>When you install an R package, it can go either to a shared location or to a user-specific location.
Installing to the shared location is better, as others can then use the package, however in the default
installation of R the shared directories are owned by the root user and their permissions must be
changed to allow ordinary users to install shared packages. This is done as follows (unless you have 
<ttx>sudo</ttx> access, you will need an administrator to do this):
<ol>
	<li> Run R (by typing "R")
	<li> In R, use <ttx>Sys.getenv("R_HOME")</ttx> to find the R installation directory (usually /usr/lib64/R)
	<li> Exit from R with command "q()"
	<li> <ttx>ls /usr/lib64/R</ttx> (or wherever the R_HOME was)
	<li> The directory listing should have a <ttx>library</ttx> subdirectory; if not, R is not installed right
	<li> Change permissions on the <ttx>library</ttx> directory
		(and everything under it) such that ordinary users can write to it. 
		Ideally there is an appropriate user group and you can use <ttx>chgrp -R &lt;group&gt; library</ttx>
		and <ttx>chmod -R g+w library</ttx> to give this group write permissions. Otherwise, 
		<ttx>chmod -R 777 library</ttx> will be necessary. 
</ol>

Also, you can check the <ttx>R-devel</ttx> installation using <ttx>ls &lt;R_HOME&gt;/etc</ttx>. There
should be a file <ttx>Renviron</ttx> in this <ttx>etc</ttx> directory; if not, or if <ttx>etc</ttx> is
not there, then <ttx>R-devel</ttx> was not installed. 


<!--#include virtual="../ssi/end.html" -->
</BODY>
</HTML>
