<html>
<title>tra overview</title>
<body>
<center>
<h2>TCW overview for tra </h2>
<table width=700 border=1><tr><td>
<pre>
Project: tra   #Seqs: 211   #Hits: 8,516   #GOs: 5,002    RPKM  DE  GO-DE   Pairs
 
INPUT
   Counts: 
      SEQID  ID          SIZE  #REPS  
      tra    Stem   1,865,201      5  
      tra    Root     955,097      5  
      tra    Oleaf    474,551      5  

   Sequences: 
      SEQID  SIZE  TITLE    AVG-len  MED-len  
      tra     211  demoTra    1,079      861  

ANNOTATION
   Annotated with TCW release: 3.0.3   Annotation date: 13-Jun-20

   Annotation Statistics:
      Sequences with hits     209  (99.1%)     Bases covered by hit  179,766  (79.0%)  
      Unique hits           8,516              Total bases           227,683           
      Total sequence hits  12,005                                                      
                                                                                       
      Sequences with GOs      207  (98.1%)     Best hit with GOs         199  (94.3%)  
      Unique GOs            5,002              Unique hits with GOs    8,088  (95.0%)  
      Has goslim_plant         94                                                      

   annoDBs (Annotation databases): 7   (see Legend below)
      ANNODB            ONLY  EVAL  ANNO  UNIQUE  TOTAL   AVG  HIT-SEQ (#SEQ)  BEST  AVG  COVER  COVER  
                                                         %SIM                  HIT  %SIM   >=50   >=90  
      SP-plants            0     2     2     747  1,248  65.3     193 (91.5%)       69.7  62.2%   3.6%  
      SP-fungi             0     0     0     566    947  47.3     112 (53.1%)       47.2  28.6%   0.9%  
      SP-invertebrates     0     0     0     285    482  50.1     121 (57.3%)       47.3  30.6%   0.8%  
      SP-bacteria          0     0     0     532    819  46.6      59 (28.0%)       46.8  28.8%     0%  
      SP-fullSubset        0     0     0     686  1,361  49.7     134 (63.5%)       48.8  33.6%   0.7%  
      TR-plants            9   204   204   4,056  4,723  76.6     209 (99.1%)       81.0  69.9%   8.6%  
      TR-invertebrates     0     3     3   1,644  2,425  51.0     163 (77.3%)       49.7  31.3%   0.6%  

   Top 15 species from total: 992
      SPECIES (25 char)      EVAL   ANNO  TOTAL     SPECIES                 EVAL   ANNO   TOTAL  
      Musa acuminata          125     51    341     Apostasia shenzhenica      2      3      80  
      Phoenix dactylifera      33     73    359     Elaeis guineensis          2      3      20  
      Ananas comosus           11     20    163     Macleaya cordata           2      2      76  
      Zingiber officinale       7      7     24     Pinus tabuliformis         2      2       6  
      Anthurium amnicola        4     10    111     Cucumis melo               1      3      31  
      Dendrobium catenatum      3      3     95     Zea mays                   1      2     227  
      Vitis vinifera            3      1     84     Aquilegia coerulea         1      1      60  
      Nelumbo nucifera          2      7    146     Other                     10     21  10,182  

   Total ORFs: 211 (100.0%)
      Has Hit             209 (99.1%)    ORF>=300         191 (90.5%)    MultiFrame     3  (1.4%)  
      Is Longest ORF      200 (94.8%)    Has Start&Stop    53 (25.1%)    >=9 Ns in ORF  0    (0%)  
      Markov Best Score  211 (100.0%)    Has Start|Stop   164 (77.7%)    Stops in Hit   5  (2.4%)  
      All of the above    199 (94.3%)    Average ORF len         860         

   GC Content: 48.65%
      Pos1  18.3%              5UTR    CDS   3UTR              5UTR     CDS    3UTR  
      Pos2  13.7%    %GC      43.91  48.78  36.27    Length     22k    181k     25k  
      Pos3  16.6%    CpG-O/E   0.88   0.71   0.63    AvgLen   102.3   859.7   117.1  
                                    
   Similar pairs:
      Total pairs:                18  
      with Nucleotide selfblast:  18  

EXPRESSION
   RPKM: (% of 211)
                 <2     2-5    5-10    10-50  50-100   100-1k     1k-5k     >=5k  
      Stem   3 (1%)  2 (1%)  1(<1%)   5 (2%)  5 (2%)  41(19%)  101(48%)  53(25%)  
      Root   1(<1%)  0 (0%)  1(<1%)   1(<1%)  0 (0%)  48(23%)  102(48%)  58(27%)  
      Oleaf  6 (3%)  0 (0%)  2 (1%)  11 (5%)  3 (1%)  40(19%)   92(44%)  57(27%)  

   Differential expression: (% of 211)
             <1E-5    <1E-4   <0.001    <0.01     <0.05  
      StRo  6 (3%)  17 (8%)  32(15%)  65(31%)   97(46%)  
      StOl  0 (0%)   2 (1%)  12 (6%)  65(31%)  101(48%)  
      RoOl  1(<1%)  16 (8%)  37(18%)  74(35%)  110(52%)  

   Over-represented GOs: (% of 5,002)
             <1E-5   <1E-4  <0.001    <0.01     <0.05  
      StRo  0 (0%)  0 (0%)  3(<1%)  39 (1%)  209 (4%)  
      StOl  0 (0%)  0 (0%)  0 (0%)   1(<1%)   26 (1%)  
      RoOl  0 (0%)  0 (0%)  1(<1%)  13(<1%)  141 (3%)  

SEQUENCES
   Build date: 13-Jun-20
   Sequences loaded from external source

   Sequence lengths:
      <=100  101-500  501-1000  1001-2000  2001-3000  3001-4000  4001-5000  >5000  
      0(0%)  37(18%)   84(40%)    72(34%)      8(4%)      8(4%)      2(1%)  0(0%)  

   ORF lengths:
      <=100  101-500  501-1000  1001-2000  2001-3000  3001-4000  4001-5000  >5000  
      0(0%)  61(29%)   92(44%)    48(23%)      5(2%)      5(2%)      0(0%)  0(0%)  

LOCATIONS
   Sequences with location:          12  unique locations: 12
   Sequences on positive strand:      7  negative strand:  5
   Sequences per group:
      1  2  3-4  5-7  8-10  11-20  21-30  >30  
      0  0    3    0     0      0      0    0  


-------------------------------------------------------------------
PROCESSING INFORMATION:
   AnnoDB Files:
      Type  Taxo           FILE                                DB DATE    ADD DATE   EXECUTE                                     
      sp    plants         uniprot_sprot_plants.fasta          20-Aug-18  13-Jun-20  diamond  --masking 0 --max-hsps 1 --top 20  
      sp    fungi          uniprot_sprot_fungi.fasta           20-Aug-18  13-Jun-20  diamond  --masking 0 --max-hsps 1 --top 20  
      sp    invertebrates  uniprot_sprot_invertebrates.fasta   20-Aug-18  13-Jun-20  diamond  --masking 0 --max-hsps 1 --top 20  
      sp    bacteria       uniprot_sprot_bacteria.fasta        20-Aug-18  13-Jun-20  diamond  --masking 0 --max-hsps 1 --top 20  
      sp    fullSubset     uniprot_sprot_fullSubset.fasta      20-Aug-18  13-Jun-20  diamond  --masking 0 --max-hsps 1 --top 20  
      tr    plants         uniprot_trembl_plants.fasta         20-Aug-18  13-Jun-20  diamond  --masking 0 --max-hsps 1 --top 20  
      tr    invertebrates  uniprot_trembl_invertebrates.fasta  20-Aug-18  13-Jun-20  diamond  --masking 0 --max-hsps 1 --top 20  

   Over-represented GOs: go_202005-termdb-tables.tar.gz    [GOs added with sTCW v3.0.3 (14-June-20)]
   GO Slim: goslim_plant

   ORF finder:
      Use ATG only for start site
      Rule 1: Use Good hit: E-value <=1E-10 or Sim >= 20%
      Rule 2: Use longest ORF if Log Len Ratio > 0.5
      Rule 3: Use best Markov score
              Train using best hits
      Good coverage: Hit overlap >= 95% with Sim 60% (internal params)

   DE (Differential Expression) computation: 
      DE column    Conditions                     Method
      StRo         Stem : Root                    source(edgeRglm.R)
      StOl         Stem : Oleaf                   source(edgeRglm.R)
      RoOl         Root : Oleaf                   source(edgeRglm.R)

   GOseq: RoOl:0.05,StOl:0.05,StRo:0.05

-------------------------------------------------------------------
LEGEND:
   annoDB:
      ANNODB    is DBTYPE-TAXO, which is the DBtype and taxonomy
      ONLY      #Seqs that hit the annoDB and no others
      EVAL      #Seqs with the best E-value from the annoDB
      ANNO      #Seqs with the best annotation from the annoDB 
      UNIQUE    #Unique hits to the annoDB
      TOTAL     #Total seq-hit pairs for the annoDB
      AVG %SIM  Average percent similarity of the total seq-hit pairs
      HIT-SEQ   Percent of #Seqs that have at least one hit from the annoDB
      BEST HIT  The following columns refer to the best hit (Rank=1):
         AVG %SIM  Average percent similarity of the best hit seq-hit pairs
         Cover>=N  Percent of HIT-SEQ where the best hit has similarity>=N% and hit coverage>=N%

   #Seqs is listed at top of overview
   Best Annotation:
      Descriptions may not contain words such as 'uncharacterized protein'

</pre>
</body>
</html>
