<html>
<title>tra overview</title>
<body>
<center>
<h2>TCW overview for tra </h2>
<table width=700 border=1><tr><td>
<pre>
Project:  tra   #Seqs: 211   #Hits: 17,080   #GOs: 6,651    TPM  Seq-DE  GO-Enrich   Pairs
 
05-Jul-21 Build Database  sequences loaded from external source
05-Jul-21 Last Annotation with sTCW v3.2.6

INPUT
   Counts: 
      SEQID  ID          SIZE  TITLE            #REPS  
      tra    Stem   1,865,201  Tissue Stem          5  
      tra    Root     955,097  Tissue Root          5  
      tra    Oleaf    474,551  Tissue Old Leaf      5  

   Sequences: 
      SEQID  SIZE  TITLE                             AVG-len  MED-len  
      tra     211  Demo of assembled Illumina reads    1,079      861  

ANNOTATIONS
   Hit Statistics:
      Sequences with hits     209  (99.1%)     Bases covered by hit  178,710  (78.5%)  
      Unique hits          17,080              Total bases           227,683           
      Total sequence hits  23,642                                                      

   annoDBs (Annotation databases): 9   (see Legend below)
      ANNODB            ONLY  BITS  ANNO  UNIQUE  TOTAL   AVG  HIT-SEQ (#SEQ)  BEST  AVG  COVER  COVER  
                                                         %SIM                  HIT  %SIM   >=50   >=90  
      SP-plants            0     0     0   1,320  2,954  53.7     193 (91.5%)       69.9  61.7%   3.6%  
      SP-invertebrates     0     0     0     608  1,248  42.5     126 (59.7%)       47.5  30.2%   0.8%  
      SP-fungi             0     0     0     850  1,451  43.4     118 (55.9%)       47.0  26.3%   0.8%  
      SP-bacteria          0     0     0     787  1,121  44.2      63 (29.9%)       45.8  25.4%     0%  
      SP-fullSubset        0     0     0   1,259  2,288  44.7     137 (64.9%)       48.7  34.3%   0.7%  
      TR-plants            0   156   126   4,569  5,216  76.0     209 (99.1%)       81.8  68.4%   8.6%  
      TR-invertebrates     0     3     3   2,914  3,716  50.0     179 (84.8%)       54.7  38.5%   2.8%  
      PR-nr                0    50    80   4,569  5,219  76.8     209 (99.1%)       82.6  67.5%  12.4%  
      TF-tf                0     0     0     204    429  44.6     123 (58.3%)       51.0   6.5%     0%  

   Top 15 species from total: 1,618
      SPECIES (25 char)         BITS   ANNO  TOTAL     SPECIES                 BITS   ANNO   TOTAL  
      Musa acuminata              95    108    806     Cocos nucifera             2      3     134  
      Musa balbisiana             49     35    610     Anthurium amnicola         2      2      96  
      Ensete ventricosum          14      9    532     Macleaya cordata           2      2      75  
      Elaeis guineensis            8      9    690     Curcuma alismatifolia      2      2       2  
      Phoenix dactylifera          7      7    615     Vitis riparia              1      2      35  
      Zingiber officinale          7      7     46     Pinus tabuliformis         1      2       8  
      Meloidogyne enterolobii      3      0     18     Asparagus officinalis      1      1     194  
      Ananas comosus               2      3    843     Other                     13     17  18,938  

   Gene Ontology Statistics:
      Unique GOs          6,651              Unique hits with GOs  10,701  (62.7%)  
      Sequences with GOs    208  (98.6%)     Seq best hit has GOs     148  (70.1%)  
      Has goslim_plant       95                                                     
                                                                                    
      biological_process  4,878  (73.3%)     is_a                  10,778           
      molecular_function  1,027  (15.4%)     part_of                1,176           
      cellular_component    746  (11.2%)                                            

EXPRESSION
   TPM: (% of 211)
               <2.0     2-5    5-10    10-50  50-100   100-1k     1k-5k     >=5k  
      Stem   3 (1%)  2 (1%)  1(<1%)   6 (3%)  7 (3%)  45(21%)  100(47%)  47(22%)  
      Root   1(<1%)  0 (0%)  1(<1%)   1(<1%)  0 (0%)  60(28%)  101(48%)  47(22%)  
      Oleaf  6 (3%)  1(<1%)  1(<1%)  11 (5%)  6 (3%)  40(19%)   96(45%)  50(24%)  

   Differential expression:  (% of 211)
             <1E-5    <1E-4   <0.001    <0.01     <0.05  
      StRo  6 (3%)  17 (8%)  32(15%)  65(31%)   97(46%)  
      StOl  0 (0%)   2 (1%)  12 (6%)  65(31%)  101(48%)  
      RoOl  1(<1%)  16 (8%)  37(18%)  74(35%)  110(52%)  

   Gene ontology enrichment: (% of 6,651)
             <1E-5   <1E-4  <0.001    <0.01     <0.05  
      StRo  0 (0%)  0 (0%)  2(<1%)  58 (1%)  311 (5%)  
      StOl  0 (0%)  0 (0%)  0 (0%)   1(<1%)   48 (1%)  
      RoOl  0 (0%)  0 (0%)  2(<1%)  21(<1%)  172 (3%)  

SEQUENCES
   Sequence lengths:
      <=100  101-500  501-1000  1001-2000  2001-3000  3001-4000  4001-5000  >5000  
      0(0%)  37(18%)   84(40%)    72(34%)      8(4%)      8(4%)      2(1%)  0(0%)  

   ORF lengths:
      <=100  101-500  501-1000  1001-2000  2001-3000  3001-4000  4001-5000  >5000  
      0(0%)  61(29%)   92(44%)    48(23%)      5(2%)      5(2%)      0(0%)  0(0%)  

   Total ORFs: 211 (100.0%)
      Has Hit            209  (99.1%)    Average ORF len          858                                
      Is Longest ORF     200  (94.8%)    ORF>=300         191  (90.5%)    MultiFrame     4   (1.9%)  
      Markov Best Score  211 (100.0%)    Has Start&Stop    52  (24.6%)    >=9 Ns in ORF  0     (0%)  
      All of the above   199  (94.3%)    Has Start|Stop   161  (76.3%)    Stops in Hit   6   (2.8%)  
                        

   GC Content: 48.65%
      Pos1  18.3%              5UTR    CDS   3UTR              5UTR     CDS    3UTR  
      Pos2  13.7%    %GC      44.20  48.75  36.26    Length     22k    181k     25k  
      Pos3  16.6%    CpG-O/E   0.87   0.71   0.63    AvgLen   103.7   858.2   117.1  
                                    
   Similar pairs: 449
      Nucleotide              20  
      Translated nucleotide  415  
      Translated ORFs        445  


-------------------------------------------------------------------
PROCESSING INFORMATION:
   AnnoDB Files:
      Type  Taxo           FILE                                DB DATE    ADD DATE   EXECUTE                  
      sp    plants         uniprot_sprot_plants.fasta          27-Jan-21  05-Jul-21  Load tra_SPpla.dmnd.tab  
      sp    invertebrates  uniprot_sprot_invertebrates.fasta   27-Jan-21  05-Jul-21  Load tra_SPinv.dmnd.tab  
      sp    fungi          uniprot_sprot_fungi.fasta           27-Jan-21  05-Jul-21  Load tra_SPfun.dmnd.tab  
      sp    bacteria       uniprot_sprot_bacteria.fasta        27-Jan-21  05-Jul-21  Load tra_SPbac.dmnd.tab  
      sp    fullSubset     uniprot_sprot_fullSubset.fasta      27-Jan-21  05-Jul-21  Load tra_SPful.dmnd.tab  
      tr    plants         uniprot_trembl_plants.fasta         27-Jan-21  05-Jul-21  Load tra_TRpla.dmnd.tab  
      tr    invertebrates  uniprot_trembl_invertebrates.fasta  27-Jan-21  05-Jul-21  Load tra_TRinv.dmnd.tab  
      pr    nr             nr.fa                               25-Jan-21  05-Jul-21  Load tra_PRnr.dmnd.tab   
      tf    tf             PlantTFDB_TCW.fa                    25-Oct-20  05-Jul-21  Load tra_TFtf.dmnd.tab   

   Gene Ontology: go-basic.obo-Feb2021  GOdb: go_demo  [GOs added with sTCW v3.2.6]
   GO Slim: goslim_plant

   ORF finder:
      Use ATG only for start site
      Rule 1: Use Good hit: E-value <=1E-10 or Sim >= 20%
      Rule 2: Use longest ORF if Log Len Ratio > 0.5
      Rule 3: Use best Markov score
              Train using best hits
      Good coverage: Hit overlap >= 95% with Sim 60% (internal params)

   DE (Differential Expression) computation: 
      Column       Method               Conditions
      StRo         edgeRglm.R           Stem : Root 
      StOl         edgeRglm.R           Stem : Oleaf 
      RoOl         edgeRglm.R           Root : Oleaf 

   GO enrichment computation: 
      Column       Method               Cutoff
      StRo         goSeqNoFDR.R         0.0500
      StOl         goSeqNoFDR.R         0.0500
      RoOl         goSeqNoFDR.R         0.0500

-------------------------------------------------------------------
LEGEND:
   annoDB:
      ANNODB    is DBTYPE-TAXO, which is the DBtype and taxonomy
      ONLY      #Seqs that hit the annoDB and no others
      BITS      #Seqs with the overall best bitscore from the annoDB
      ANNO      #Seqs with the overall best annotation from the annoDB 
      UNIQUE    #Unique hits to the annoDB
      TOTAL     #Total seq-hit pairs for the annoDB
      AVG %SIM  Average percent similarity of the total seq-hit pairs
      HIT-SEQ   Percent of #Seqs that have at least one hit from the annoDB
      BEST HIT  The following columns refer to the best hit (Rank=1):
         AVG %SIM  Average percent similarity of the best hit seq-hit pairs
         Cover>=N  Percent of HIT-SEQ where the best hit has similarity>=N% and hit coverage>=N%

   #Seqs is listed at top of overview
   Best Annotation:
      Descriptions may not contain words such as 'uncharacterized protein'

</pre>
</body>
</html>
