<html>
<body style="padding:20px;width:650px;font-family:Times;">

<title>Cluster Methods</title>

<h2>Clustering Methods</h2>
 

<p>From the drop-down menu, choose the clustering method to use. On "Keep", it 
will be written to the "Cluster Set" table on the main panel. The clusters will
be computed when the <b>Add New Clusters</b> button is pressed.

<p><b>Prefix (required)</b> :
This is the prefix used for the clusters, e.g. if the 
cluster method is prefixed with "T1" and has three clusters, they will be named T1_1, T1_2, T1_3.
There can be multiple instances of the same "Method" in the databases, e.g. built with different
parameters; the prefix allows them to be queries separately in <tt>viewMultiTCW</tt>.

<h3>BBH (Bi-directional Best Hit)</h3>
This methods creates the bi-directional best hit. 
<p>
<table>
<tr><td nowrap valign=top><u>Hit File Type</u><td>Either the amino acid search results or the nucleotide search results can be
used, where the nucleotide is only valid if there are at least two nucleotide
singleTCW in the multiTCW database.
<tr><td valign=top><u>%Coverage</u><td>This is computed as the percentage of align/length.
The %coverage must be greater than this number for  'either' or 'both' sequences. 
For example, if
the align length is 100, sequence 1 has length 100, and sequence 2 has length 300, then the
coverage will be 100% and 33%, respectively; if the cutoff is 90,
the pair will pass for the 'either' option, but not the 'both' option.
<tr><td valign=top><u>%Similarity</u><td>This is the column named "%Identity" in the Hit file, 
where the %similarity must be greater or equal this value.
</table>

<p><i>If your mTCWdb only has two sTCWdbs as input, then ignore the rest of this section.</i>
<p>For an mTCWdb with &gt;2 sTCWdbs as input, this method works in the following ways:
<ol>
<li>Specify two sTCWdbs to create BBH clusters from the two datasets, which uses the
traditional BBH algorithm. 
<li>If you specify more then N (N&gt;2) sTCWdbs, it will create BBH clusters with N members 
where each pair of a cluster must be a BBH pair. 

</ol>
<h3>Closure</h3>

The parameters are the same as described for BBH, except that there  is no option to select the
sTCW databases as all sequences are considered regardless of origin. 
<p>For a sequence to be added to a Closure cluster, it must have a search hit with every
other sequence in the cluster, and 
it must pass the %similarity and %coverage filter
with at least one sequence in the cluster.

<h3>Best Hit</h3>
Clusters will be created from the best annotation hit description or hitID. If the hit description is used,
a cluster can contain multiple different hitIDs. To compare descriptions:
<ol>
<li>If the description is un-informative (e.g. contains "uncharacterized protein"), then the sequence can only
be part of a cluster with the same hitID.
<li>Before comparing the descriptions, each is converted to lower case, and if
there is text at the end starting with "{", it is removed.
</ol>
Clusters are formed as follows (the terms <i>protein hit</i> or <i>pair hit</i> indicates a DIAMOND/BLAST result):
<ol>
<li>All sequences share the same best <u>HitID</u> or best <u>Description</u>.
<li>Each sequence in the cluster has a protein hit with 
<u>%Similarity</u>&gt;=N and <u>%Coverage</u>&gt;=M to the representative cluster protein. 
<li>A sequence must have a pair hit with <u>%Similarity</u>&gt;=N and <u>%Coverage</u>&gt;=M to all/any other sequences in the cluster.
<ul>
<li>All: If the <u>All Hit</u> filter is "Yes", then each sequence in a cluster must pass rule #3 with all other
sequences in the cluster, 
<li>Any: If the <u>All Hit</u> filter is "No", each sequence only needs to pass rule #3 with one other sequence in cluster.
</ul>
If the parameters are too lax, this can create some bad clusters. Rule of thumb:
<br>Yes - you can reduce %similarity and %coverage.
<br>No -  you should increase %similarity and %coverage significantly.
</ol>


<h3>OrthoMCL</h3>

OrthoMCL is an ortholog-detection package which has been included 
with TCW. Its
clustering is governed by the <u>Inflation</u> parameter; see www.orthomcl.org for more detail.
Larger inflation values lead to smaller clusters.

<p><b>NOTE</b>: OrthoMCL (orthomcl.org) is an external program run from the /Ext directory. 
For it to work, Perl and DBD::MySQL must be installed. 

<h3>User Defined</h3>

Here you load a file defining your own groups. The format is as follows:
<pre>
D1: bar|bar_072 foo|foo_210 fly|fly_141 foo|foo_204
D2: bar|bar_002 foo|foo_045 fly|fly_031
...
</pre>
Each line has a group number followed by a space-separated list of the sequences in the group. 
Each group member starts with the project prefix followed by "|" (e.g. <tt>bar|</tt>), where
the prefix for a sTCW database is defined when the database is added.  In the project
directory, there is a <tt>methods</tt> directory where you can see example of cluster files added by one
of the above methods.
