<!DOCTYPE html>
<HTML>
<!--#set var=TITLE value="TCW Documentation Index" -->
<!--#include virtual="./ssi/head.html" -->
<body onLoad="init('doc')">
<a id=top></a>
<!--#include virtual="./ssi/start1.html" -->

<center>
<table style="width: 80%">
 
<tr><td colspan=3> Getting started</td></tr>
<tr>	<td>&nbsp;&nbsp;<td><a href="Summary.html">Summary</a>		<td>A summary of the input, processing and suite of programs.

<tr>	<td>&nbsp;&nbsp;<td><a href="Install.html">Installation</a><td>Installing TCW and other necessary and optional software.
<tr>	<td colspan=3>&nbsp;&nbsp;

<tr><td colspan=3> Major documents with step-by-step details using the Demo dataset</td></tr>
<tr>	<td>&nbsp;&nbsp;<td><a href="stcw/AnnoDBs.html"><ttp>runAS</ttp></a></td>
			<td>Annotation Setup<sup>1</sup> - Getting UniProt, GO, nr, etc for use with <ttp>runSingleTCW</ttp></td></tr>

<tr>	<td>&nbsp;&nbsp;<td><a href="stcw/UserGuide.html"><ttp>runSingleTCW</ttp></a></td>
			<td>Single species - Building a single species annotated TCW database
			 
<tr>	<td>&nbsp;&nbsp;<td class="top"><a href="stcw/DiffExp.html"><ttp>runDE</ttp></a></td>
			<td>Differential Expression - Add DE to the singleTCW database
			</td></tr> 
<tr>	<td>&nbsp;&nbsp;<td><a href="mtcw/UserGuide.html"><ttp>runMultiTCW</ttp></a></td>
			<td>Multiple species - Building a comparison database from single TCW databases
			
<tr>	<td>&nbsp;&nbsp;<td><ttp>viewSingleTCW</ttp> and <ttp>viewMultiTCW</ttp></td>
			<td>Viewing results - the user manual is built into the programs Help pages; 
			also see 
			<a href="./tour/viewSingle/index.html">Single Tour</a> and 
			<a href="./tour/viewMulti/index.html">Multi Tour</a>. 
			
<tr>	<td colspan=3>&nbsp;&nbsp;
<tr><td colspan=3> Details</td></tr>
<tr>	<td>&nbsp;&nbsp;<td><a href="stcw/AnnoGuide.html">Annotate Details</a></td>
			<td>Removing, adding, updating annotation; pruning hits</td></tr>
		<tr>	<td>&nbsp;&nbsp;<td><a href="stcw/ORF.html">ORF finder</a></td>
			<td>Description of options, algorithm, outputs, and results. </td></tr>
<tr>	<td>&nbsp;&nbsp;<td><a href="stcw/AssemGuide.html">Assembly Details</a></td>
			<td>Not typically necessary<sup>2</sup></td> </tr>

<tr>	<td colspan=3>&nbsp;&nbsp;
<tr><td colspan=3> Other docs</td></tr>
<tr>	<td>&nbsp;&nbsp;<td><a href="External.html">External Software</a></td><td>Packages used by TCW</td></tr>
<tr>	<td>&nbsp;&nbsp;<td><a href="./tour">TCW Tour</a> </td> 	<td>Snapshots of different TCW screens</td></tr>

<tr>	<td>&nbsp;&nbsp;<td><a href="Mac-OSX.html">MacOS</a></td>	<td>Installing MySQL, etc on MacOS</td></tr>
<tr>	<td>&nbsp;&nbsp;<td><a href="./Trouble.html">Trouble Shooting</a></td>
			<td>Potential technical difficulties.</td></tr>
<tr>	<td>&nbsp;&nbsp;<td><a href="./Release.html">
				Release Notes</a></td><td>Release detail since v1.0</td></tr>
<tr>	<td>&nbsp;&nbsp;<td><font color=red>Help</font></td><td>All TCW programs have Help pop-ups on every display.
	
<tr><td colspan=3><td>&nbsp;&nbsp;</td>
<tr><td colspan=3><sup>1</sup><font size="2">The sequences that are loaded into TCW (e.g. transcripts, proteins) are searched against one or more databases for annotation;
these annotation databases are referred to as "AnnoDBs". When UniProt is used for annotation, TCW provides GO, KEGG,
EC, Interpro and Pfam annotations.</font></td>

<tr><td colspan=3><sup>2</sup><font size="2">TCW can be used with already assembled transcripts, proteins or mRNA (e.g. predicted genes with introns removed), 
i.e. the assembly is not usually necessary now.
Though the TCW assembly does not work for high-throughput sequencing reads, it is still occasionally useful for Sanger
ESTs or assembling multiple sets of transcripts.</font></td>
</table>
</center>
<!--#include virtual="./ssi/end.html" -->
</body>
</HTML>


