<!DOCTYPE html>
<HTML>
<!--#set var=TITLE value="TCW Summary" -->
<!--#include virtual="./ssi/head.html" -->
<body>
<a id=top></a>
<!--#include virtual="./ssi/start1.html" -->
TCW provides two sets of programs: <ttp>singleTCW</ttp> for annotating and viewing 
single species transcriptions, and <ttp>multiTCW</ttp> for 
annotating and viewing comparative transcripts.

<p>Contents:
<ul>
<li><a href="#task">Programs</a>
<li><a href="#mach">Machines tested on</a>
<li><a href="#time">Timings</a>
<li><a href="#ref">References</a>
</ul>
<a id=task></a>
<h2>Programs</h2>
<b><i><u>singleTCW contains four major programs:</u></i></b>
<table class="tablex">
 <tr>
 	<td class="top">
	1. <a href="./stcw/AnnoDBs.html" class="ext" target="_blank"><ttp>runAS</ttp></a> -- annotation setup for input to <ttp>runSingleTCW</ttp>
	<ul>
		<li>Downloads the UniProt taxonomic and/or full SwissProt and TrEMBL .dat files
		<li>Create fasta files from .dat files for searching against TCW sequences
		<li>Download the GO mysql database and augment it with UniProt information about GO, KEGG, Pfam, and InterPro
	</ul>
	Note: <ttp>runSingleTCW</ttp> can take as input other databases such as Genbank nr, 
	but these results will not have associated GO annotations.
	<td  class="top"><a href="./tour/runAS/img/runAS.png"><img src="./tour/runAS/img/runAS.png" alt="" style="border: 1px solid black; width: 100px"></a>
</table>

<table class="tablex">
 <tr><td class="top">
	2. <a href="./stcw/UserGuide.html" class="ext" target="_blank"><ttp>runSingleTCW</ttp></a> -- builds a single-species TCW database (sTCWdb)
	<ul>
	   <li> Input: sequences and optional counts, where <i><b>any</b></i> of the following are valid (
	   the first three are NT-sTCWdbs and the last is a AA-sTCWdb):
		<ol>
		<li>Load RNA-seq transcripts and count data with optional replicates.
		<li>Load sequences with location information (e.g. predicted genes).
		<li>Assemble up to ~1M sequences, such as: transcript sets, 
		paired-end Sanger ESTs, or a mix of transcripts and ESTs. 
		<li>Load protein sequences and spectra (count) data with optional replicates.
		</ol>
	   <li> Annotation:
		<ol>
		<li>Annotate sequences with one or more nucleotide or protein databases (called annoDBs). 
		UniProt should be downloaded with the <ttp>runAS</ttp> program. The searching may be done with 
		the super-fast
		<ttp>DIAMOND</ttp> or the standard <ttp>BLAST</ttp>.
		<li>If UniProt is used, GO annotations along with EC, KEGG and Pfam identifiers are extracted 
		from the GO database and entered into the sTCWdb (The GO database is set up with the <ttp>runAS</ttp>).
		<li>Compute ORFS and GC content.
		</ol>
	   <li> All data and results are stored in a MySQL database. 
	</ul>
	<td  class="top"><a href="./tour/runSingle/img/runSingle.png"><img src="./tour/runSingle/img/runSingle.png" alt="" style="border: 1px solid black; width: 100px"></a>
</table>
	
<table class="tablex">
 <tr><td class="top">
	3. <a href="./stcw/DiffExp.html" class="ext" target="_blank"><ttp>runDE</ttp></a> -- run Differential Expression analysis
	<ul>
		<li>An interface to several R packages (<ttp>EdgeR, DESeq</ttp>) for calculating differential
			of sequences. Additionally, it can execute a user-provided R script for DE calculation.
		<li>If the sTCWdb contains GOs, the <ttp>GOseq</ttp> R program can be used to compute differential GO terms.
	</ul>
	<td  class="top"><a href="./tour/runDE/img/runDE.png"><img src="./tour/runDE/img/runDE.png" alt="" style="border: 1px solid black; width: 100px"></a>
</table>

<table class="tablex">
 <tr><td class="top">
	4. <a href="./tour/viewSingle" class="ext" target="_blank"><ttp>viewSingleTCW</ttp></a> -- view a single-species TCW database
	<ul>
		<li>Query and view the results. There are various filters, for example,
	   	filters are provided specific to taxonomic databases, trimmed GOs, filter by annotation, etc.
	   	The initial view is the <a href="./ov/sDemo.html">Overview</a>, which summarizes the results.
	</ul>
	<td  class="top"><a href="./tour/viewSingle/img/mainTable.png"><img src="./tour/viewSingle/img/mainTable.png" alt="" style="border: 1px solid black; width: 100px"></a>
</table>
	
<b><i><u>multiTCW contains two major programs</u></i></b>:
<table class="tablex">
 <tr><td class="top">
	1. <a href="./mtcw/UserGuide.html" class="ext" target="_blank"><ttp>runMultiTCW</ttp></a> -- builds a multi-species comparison TCW database (mTCWdb)
	<ul><li>Builds a database from multiple sTCWdbs, which can be NT-sTCWdbs, AA-sTCWdbs
		or a mix. This has been tested with up to 5 input sTCWdbs.
		<li>Runs <ttp>DIAMOND</ttp> or <ttp>BLAST</ttp> to compare the sequences from the input TCWs. Add the resulting pairs to the database.
	 	<li>Clusters the pairs into ortholog groups. They can be clustered with 
	 	TCW Closure, TCW BBH (best bi-directional), TCW shared hit, <ttp>OrthoMCL</ttp>, and/or user-supplied clusters can be uploaded. 
		Multiple ortholog clustering can be in the database for query.
		<li>If the input is from two or more NT-sTCWdbs, coding statistics are calculated. Additionally,
		alignment files are output for input to <ttp>KaKs_calculator</ttp>, and the results of running
		the KaKs_calculator are input to <ttp>runMultiTCW</ttp>.
		<li>MSA (multiple sequence alignments) are computed for each cluster and scored with Trident.
	 </ul>
<td  class="top"><a href="./tour/runMulti/img/runMultiTCW.png"><img src="./tour/runMulti/img/runMultiTCW.png" alt="" style="border: 1px solid black; width: 100px"></a>
</table>

<table class="tablex">
 <tr><td class="top">
	2. <a href="./tour/viewMultiTCW" class="ext" target="_blank"><ttp>viewMultiTCW</ttp></a> -- view a multi-species TCW database
	 	<ul><li>Query and view the results. The results can be filtered on various attributes. A cluster
	 	can be viewed graphically with the results of <ttp>MAFFT</ttp> or <ttp>MUSCLE</ttp> 
	 	MSA or pairwise alignment
	 	(e.g. see graphical views <a href="./tour/viewMulti/img/MSAdb.png">MSA</a> and <a href="./tour/viewMulti/img/MSAdb.png">Pair</a>, respectively).
	 	</ul>
	<td  class="top"><a href="./tour/viewMulti/img/pairTable.png"><img src="./tour/viewMulti/img/pairTable.png" alt="" style="border: 1px solid black; width: 100px"></a>
</table>

<hr>	
<a id=mach></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Machines tested on</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<table class="tabley">
<tr><th>OS<th>Architecture<th>Purchased<th>Database<th>Java<th>Test
<tr><td>Linux x86.64 (Centos 7) <td>3.2 Ghz AMD 24-Core, 128Gb<td style="text-align:center">2011<td>MariaDB v10.4.12<td>v1.8<td>Build and View
<tr><td>Linux x86.64 (Centos 7)  <td>2.0 Ghz AMD 4-Core, 20Gb<td style="text-align:center">2008<td>MariaDB v5.5.60<td>v1.7<td>View
<tr><td>MacOS (Catalina 10.15.4)<td>3.2 GHz Intel 6-Core i7, 64Gb<td style="text-align:center">2020<td>MySQL v8.0.17<td>v14.0<td>Build and View
<tr><td>MacOS (Maverick 10.9.5) <td>2.4 Ghz Intel 2-Core i5, 16Gb<td style="text-align:center">2012<td>MySQL v5.6.21<td>v1.7<td>View
</table>


<a id=time></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h2>Timings</h2>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<h3>singleTCW</h3>
The timings were run on the MacOS (row #3) and Linux (row #1) of the above machine table.
The "Mb" numbers are maximum memory (when available), and are approximate.
	<table class="tabley">
	<tr><th>Task<th>Count<th colspan=2>Mac Mini<th colspan=2>Linux
	<tr><td>Load<td>64,930 seqs
		<td style="text-align:right">1m:26s
		<td style="text-align:right">20Mb
		<td style="text-align:right">2m:17s
		<td style="text-align:right">19Mb
	<tr><td>Instantiate<td>64,930 seqs
		<td style="text-align:right">47s
		<td style="text-align:right">4Mb
		<td style="text-align:right">1m:16s
		<td style="text-align:right">1Mb
	<tr><td>Diamond<sup>1</sup><td>5G-6G file size
		<td style="text-align:right">12m:09s
		<td style="text-align:right">---
		<td style="text-align:right">11m:54s
		<td style="text-align:right">---
	<tr><td>Annotate<sup>2</sup><td>461,118 unique hits
		<td style="text-align:right">18m:14s
		<td style="text-align:right">2374Mb
		<td style="text-align:right">56m:7s
		<td style="text-align:right">2384Mb
	<tr><td>ORF<td>64,930 ORFs
		<td style="text-align:right">3m:54s
		<td style="text-align:right">150Mb
		<td style="text-align:right">9m:36s
		<td style="text-align:right">246Mb
	<tr><td>GO<td>10,724 GOs
		<td style="text-align:right">28m:42s
		<td style="text-align:right">1030Mb
		<td style="text-align:right">52m:04s
		<td style="text-align:right">1350Mb
	</table>
	<small>
	<sup>1</sup>TR-plants; The May2020 file (5G) was used on Mac, 
	and the Oct2020 file (6G) was used on Linux, which influenced the increased times on Linux. 
	<br><sup>2</sup>SP-plants, TR-plants, SP-full; times do not include Diamond. Linux had 462,883 unique hits.
	</small>
	
	<h3>multiTCW</h3>
	Using the same two computers as above, the mTCW database was built from a dataset of 
	28,392 and another of 26,685 sequences. Both datasets were annotated with SP-plants, TR-plants, and SP-full,
	where the Mac versions were downloaded May2020 and the Linux on Oct2020.
	<table class="tabley">
	<tr><th>Task<th>Count<th colspan=2>Mac Mini<th colspan=2>Linux
	<tr><td>Build Database<td>55,077 seq and 595,698 unique hits<sup>1</sup>
	     <td style="text-align:right">&nbsp;6m:28s
	     <td style="text-align:right">150Mb
	     <td style="text-align:right">9m:31s
	     <td style="text-align:right">152Mb
	<tr><td>GO<td>21,693 GOs
	    <td style="text-align:right">16m:48s
	    <td style="text-align:right">11Mb
	    <td style="text-align:right">30m:26s
	    <td style="text-align:right">506Mb
	<tr><td>Add Pairs with Hits<td>392,919 AA and 43,485 NT pairs
	    <td style="text-align:right">5m:03s
	    <td style="text-align:right">61Mb
	    <td style="text-align:right">15m:54s  
	    <td style="text-align:right">68Mb
	<tr><td>Add 3 cluster sets<sup>2</sup><td>22,376 total clusters
		<td style="text-align:right">2m:56s
		<td style="text-align:right">33Mb
		<td style="text-align:right">8m:10s 
		<td style="text-align:right">31Mb
	<tr><td colspan=6><b>Run Stats</b>
	<tr><td>PCC<td>410,401 Pairs
		<td style="text-align:right">1m:32s
		<td style="text-align:right">68Mb
		<td style="text-align:right">2m:16s 
		<td style="text-align:right">63Mb
	<tr><td>Pair Stats<td>95,628 Pairs to align
		<td style="text-align:right">44m:33s
		<td style="text-align:right">624Mb
		<td style="text-align:right">2h:5m:19s 
		<td style="text-align:right">661Mb
	<tr><td>MSA<sup>3</sup> Stats<td>22,376 clusters
		<td style="text-align:right">1h:20m:11s
		<td style="text-align:right">---
		<td style="text-align:right">4h:34m:23s
		<td style="text-align:right">---
	</table>
	<small>
	<sup>1</sup>The Linux database had 606,845 unique hits.
	<br><sup>2</sup>The methods added were BBH, Closure and BestHit. Closure used the most memory.
	<br><sup>3</sup>This step uses the external program MAFFT.
	</small> 
	<p>The Linux times are much slower than the Mac, even though they comparable architectures; the Linux
	may not be optimized well and is a much older machine. It is important with MariaDB to make sure
	the variables are set right; execute 
	<ttx>
	./runSingleTCW -v
	</ttx>
	to print out some variables that seem to effect performance with MariaDB on Linux.
	<h3>Linux only for TCW v3.0</h3>
The following are from the supplement 1 of Soderlund (2019). The singleTCW was 26,856 sequences.
The UniProts were five SwissProt, the full SwissProt and five TrEMBL databases (but not the full).
The multiTCW was created from three sTCWdbs for a total of 198,702 sequences.
There is not much speed difference between
TCW v3 and later versions.
<table class="tabley">
<tr><th>Program<th>Step<th>Time<th>Note
<tr><td>runAS<td>	Download and create UniProt FASTA files	<td style="text-align:right">5h:33m:24s           	<td>a
<tr><td>&nbsp;	<td>Download SwissProt and create subset FASTA file	<td style="text-align:right">4m:34s	<td>a
<tr><td>&nbsp;	<td>	Download and build GO database	<td style="text-align:right">8h:59m:55s	<td>b
<tr><td>runSingleTCW	<td>Build sTCW_rhi_NnR of 26,685 transcripts	<td style="text-align:right">3m:13s<td>	
<tr><td>&nbsp;	<td>	Annotate sequences	<td style="text-align:right">2h:15m:22s	<td>c
<tr><td>&nbsp;	<td>	ORF-finding	<td style="text-align:right">6m:09s	<td>
<tr><td>&nbsp;	<td>	Add GO annotations	<td style="text-align:right">50m:26s	<td>
<tr><td>runDE	<td>edgeR for 12 condition pairs	<td style="text-align:right">0m:20s	<td>d
<tr><td>&nbsp;	<td>	GOseq for each of the 12 DE results	<td style="text-align:right">10m:56s	<td>d
<tr><td>runMultiTCW	<td>Build mTCW_rhi of 198,702 transcripts	<td style="text-align:right">20m:56s	<td>
<tr><td>&nbsp;	<td>	Add GO annotation for 23,516 unique GOs	<td style="text-align:right">56m:02s	<td>
<tr><td>&nbsp;	<td>	Add 1,099,986 pairs after searching	<td style="text-align:right">28m:20s	<td>e
<tr><td>&nbsp;	<td>	Build Clusters (total)	<td style="text-align:right">22m:00s	<td>f
<tr><td>&nbsp;	<td>	&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;BBH (OlR-Os) 	<td style="text-align:right"> :04s	<td>
<tr><td>&nbsp;	<td>	&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Closure 	<td style="text-align:right">:08s	<td>
<tr><td>&nbsp;	<td>	&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;OrthoMCL	<td style="text-align:right">8m:04s	<td>
<tr><td>&nbsp;	<td>	Compute PCC for 1,099,986 pairs	<td style="text-align:right">3m:37s	<td>
<tr><td>&nbsp;	<td>	Pair-align and analyze 354,280 pairs	<td style="text-align:right">2h:27m:08s	<td>
<tr><td>&nbsp;	<td>	Multi-align and analyze 71,339 clusters	<td style="text-align:right">12h:14m:43s	<td>g
<tr><td>&nbsp;	<td>	Add KaKs values	<td style="text-align:right">1m:1s	<td>h
</table>
<small>
<br>a There can be considerable variation in download times.
<br>b The bulk of the time was from loading uniprot_trembl_bacterial.dat.gz, which took 6h:56m:58s; the compressed file was 57G and contained 96,592,456 entries. 
<br>c This did not include the DIAMOND formatting but did include the time for searching; the longest time for searching was against TrEMBL bacterial,  which took 35m:20s.
<br>d The time for executing the R code and loading the results.
<br>e The time does not include searching all sequences against themselves. The protein self-comparison took DIAMOND 3m22s and nucleotides self-search took Blastn 5m:13s.
<br>f The time for computing and adding all clusters; the majority of the time was for loading the clusters into the database. The following three individual cluster times are only for computing the clusters.
<br>g This includes the time to run MAFFT in parallel on each cluster.
<br>h This does not include the time to execute the KaKs_Calculator, which is run by the user before the add.
</small>
	<a id=ref></a>
	<table style="width: 100%"><tr><td style="text-align: left">
	<h2>References</h2>
	<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>
	
	<p class=hang>C. Soderlund (2019) Transcriptome computational workbench (TCW): 
			analysis of single and comparative transcriptomes. 
			<br><a href="https://doi.org/10.1101/733311" class="ext" target="_blank">BioRxiv</a>
			<br><i>Describes the TCW v3 package.</i>
	</p>
	
	<p  class=hang>C. Soderlund, W. Nelson and S. Goff (2014)
		Allele Workbench: transcriptome pipeline and interactive graphics for allele-specific expression.
		<br><a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0115740" class="ext" target="_blank">PLOS ONE </a> 
		<br><i>Describes a pipeline that can be used with TCW, plus a new GO trim algorithm.</i>
	</p>		
	<p  class=hang>C. Soderlund, W. Nelson, M. Willer and D. Gang (2013) TCW: Transcriptome Computational Workbench. 
		<br><a href="http://dx.plos.org/10.1371/journal.pone.0069401" class="ext" target="_blank">PLOS ONE</a>	
		<br><i>Describes the TCW v1 package.</i>
	</p>	
			
	<p  class=hang>C. Soderlund, E. Johnson, M. Bomhoff, and A. Descour (2009) 
				PAVE: Program for Assembling and Viewing ESTs. 
		<br><a href="http://www.biomedcentral.com/1471-2164/10/400" class="ext" target="_blank">BMC Genomics</a>
		<br><i>Describes the assembly algorithm.</i>
	</p>	
<br>
<!--#include virtual="./ssi/end.html" -->
</body>
</HTML>

