<html>
<title>tra overview</title>
<body>
<center>
<h2>TCW overview for tra </h2>
<table width=700 border=1><tr><td>
<pre>
Project:  tra   #Seqs: 211   #Hits: 9,584   #GOs: 5,294    TPM  Seq-DE  GO-Enrich   Pairs
 
12-Aug-21 Build Database  sequences loaded from external source
12-Aug-21 Last Annotation with sTCW v3.3.0

INPUT
   Counts: 
      SEQID  ID          SIZE  #REPS  
      tra    Root     955,097      5  
      tra    Stem   1,865,201      5  
      tra    Oleaf    474,551      5  

   Sequences: 
      SEQID  SIZE  TITLE                                 AVG-len  MED-len  
      tra     211  TCW demo with transcripts and counts    1,079      861  

ANNOTATIONS
   Hit Statistics:
      Sequences with hits     210  (99.5%)     Bases covered by hit  180,729  (79.4%)  
      Unique hits           9,584              Total bases           227,683           
      Total sequence hits  13,224                                                      

   annoDBs (Annotation databases): 8   (see Legend below)
      ANNODB            ONLY  BITS  ANNO  UNIQUE  TOTAL   AVG  Rank  HAS (%Seqs)   AVG  COVER  COVER  
                                                         %SIM   =1   HIT          %SIM   >=50   >=90  
      SP-plants            0     0     0     781  1,311  65.3    |   193 (91.5%)  70.1  62.7%   3.6%  
      SP-invertebrates     0     0     1     329    590  48.3    |   126 (59.7%)  47.6  30.2%   0.8%  
      SP-fungi             0     0     0     640  1,067  46.7    |   118 (55.9%)  47.0  27.1%   0.8%  
      SP-bacteria          0     0     0     659    944  44.9    |    63 (29.9%)  45.8  25.4%     0%  
      SP-fullSubset        0     0     0     757  1,515  48.7    |   137 (64.9%)  48.7  33.6%   0.7%  
      TR-plants            7   207   207   4,122  4,858  77.4    |   210 (99.5%)  81.8  68.6%   8.6%  
      TR-invertebrates     0     3     2   2,165  2,701  52.3    |   179 (84.8%)  54.5  39.7%   2.8%  
      TF-PlantTFDB         0     0     0     131    238  47.0    |   123 (58.3%)  51.2   7.3%     0%  

   Top 15 species from total: 1,312
      SPECIES (25 char)         BITS   ANNO  TOTAL     SPECIES                 BITS   ANNO   TOTAL  
      Musa acuminata              73     53    325     Anthurium amnicola         2      4      93  
      Musa balbisiana             62     45    312     Vitis vinifera             2      3      91  
      Ensete ventricosum          19     14    199     Dendrobium catenatum       2      2      63  
      Elaeis guineensis           10     24    344     Macleaya cordata           2      2      43  
      Phoenix dactylifera          9     19    319     Nelumbo nucifera           1      3      95  
      Zingiber officinale          7      7     24     Apostasia shenzhenica      1      2      59  
      Ananas comosus               4     11    432     Ricinus communis           1      2      31  
      Meloidogyne enterolobii      3      0      9     Other                     12     19  10,785  

   Gene Ontology Statistics:
      Unique GOs          5,294              Unique hits with GOs  8,124  (84.8%)  
      Sequences with GOs    208  (98.6%)     Seq best hit has GOs    194  (91.9%)  
      Has goslim_plant       95                                                    
                                                                                   
      biological_process  3,858  (72.9%)     is_a                  8,481           
      molecular_function    821  (15.5%)     part_of                 898           
      cellular_component    615  (11.6%)                                           

EXPRESSION
   TPM: (% of 211)
               <2.0     2-5    5-10    10-50  50-100   100-1k     1k-5k     >=5k  
      Root   1(<1%)  0 (0%)  1(<1%)   1(<1%)  0 (0%)  60(28%)  101(48%)  47(22%)  
      Stem   3 (1%)  2 (1%)  1(<1%)   6 (3%)  7 (3%)  45(21%)  100(47%)  47(22%)  
      Oleaf  6 (3%)  1(<1%)  1(<1%)  11 (5%)  6 (3%)  40(19%)   96(45%)  50(24%)  

   Differential expression:  (% of 211)
             <1E-5    <1E-4   <0.001    <0.01     <0.05  
      RoSt  6 (3%)  17 (8%)  32(15%)  65(31%)   97(46%)  
      RoOl  1(<1%)  16 (8%)  37(18%)  74(35%)  110(52%)  
      StOl  0 (0%)   2 (1%)  12 (6%)  65(31%)  101(48%)  

   Gene ontology enrichment: (% of 5,294)
             <1E-5   <1E-4  <0.001    <0.01     <0.05  
      RoSt  0 (0%)  1(<1%)  3(<1%)  54 (1%)  289 (5%)  
      RoOl  0 (0%)  0 (0%)  3(<1%)  20(<1%)  187 (4%)  
      StOl  0 (0%)  0 (0%)  0 (0%)   0 (0%)   36 (1%)  

SEQUENCES
   Sequence lengths:
      <=100  101-500  501-1000  1001-2000  2001-3000  3001-4000  4001-5000  >5000  
      0(0%)  37(18%)   84(40%)    72(34%)      8(4%)      8(4%)      2(1%)  0(0%)  

   ORF lengths:
      <=100  101-500  501-1000  1001-2000  2001-3000  3001-4000  4001-5000  >5000  
      0(0%)  61(29%)   92(44%)    48(23%)      4(2%)      6(3%)      0(0%)  0(0%)  

   ORF Stats:   Average length 861
      Has Hit            210  (99.5%)    ORF=Hit         119  (56.4%)                               
      Is Longest ORF     200  (94.8%)    ORF>=300        191  (90.5%)    MultiFrame     3   (1.4%)  
      Markov Best Score  211 (100.0%)    Has Start|Stop  162  (76.8%)    Stops in Hit   6   (2.8%)  
      All of the above   200  (94.8%)    Has Start&Stop   53  (25.1%)    >=9 Ns in ORF  0     (0%)  
                        
   GC Content: 48.65%
      Pos1  18.3%              5UTR    CDS   3UTR              5UTR     CDS    3UTR  
      Pos2  13.7%    %GC      43.71  48.80  36.27    Length     21k    182k     25k  
      Pos3  16.6%    CpG-O/E   0.87   0.71   0.63    AvgLen   101.2   860.8   117.1  
                                    
   Similar pairs: 434
      Nucleotide              20  
      Translated nucleotide  415  
      Translated ORFs        369  

LOCATIONS
   Sequences with location:          12  unique locations: 12
   Sequences on positive strand:      7  negative strand:  5
   Sequences per group:
      1  2  3-4  5-7  8-10  11-20  21-30  >30  
      0  0    3    0     0      0      0    0  


-------------------------------------------------------------------
PROCESSING INFORMATION:
   AnnoDB Files:
      Type  Taxo           FILE                                DB DATE    ADD DATE   EXECUTE                                     
      sp    plants         uniprot_sprot_plants.fasta          27-Jan-21  12-Aug-21  diamond  --max-hsps 1 --masking 0 --top 20  
      sp    invertebrates  uniprot_sprot_invertebrates.fasta   27-Jan-21  12-Aug-21  diamond  --max-hsps 1 --masking 0 --top 20  
      sp    fungi          uniprot_sprot_fungi.fasta           27-Jan-21  12-Aug-21  diamond  --max-hsps 1 --masking 0 --top 20  
      sp    bacteria       uniprot_sprot_bacteria.fasta        27-Jan-21  12-Aug-21  diamond  --max-hsps 1 --masking 0 --top 20  
      sp    fullSubset     uniprot_sprot_fullSubset.fasta      27-Jan-21  12-Aug-21  diamond  --max-hsps 1 --masking 0 --top 20  
      tr    plants         uniprot_trembl_plants.fasta         27-Jan-21  12-Aug-21  diamond  --max-hsps 1 --masking 0 --top 20  
      tr    invertebrates  uniprot_trembl_invertebrates.fasta  27-Jan-21  12-Aug-21  diamond  --max-hsps 1 --masking 0 --top 20  
      tf    PlantTFDB      PlantTFDB_TCW.fa                    25-Oct-20  12-Aug-21  diamond  --max-hsps 1 --masking 0 --top 20  

   Gene Ontology: go-basic.obo-Feb2021  GOdb: go_demo  [GOs added with sTCW v3.3.0]
   GO Slim: goslim_plant

   ORF finder:
      Use ATG only for start site
      Rule 1: Use Good hit: E-value <=1E-10 or Sim >= 20%
      Rule 2: Use longest ORF if Log Len Ratio > 0.5
      Rule 3: Use best Markov score
              Train using best hits
      Good coverage: Hit overlap >= 95% with Sim 60% (internal params)

   Differential Expression computation: 
      Column       Method                         Conditions
      RoSt         edgeRglm.R CPM>1>=2            Root : Stem 
      RoOl         edgeRglm.R CPM>1>=2            Root : Oleaf 
      StOl         edgeRglm.R CPM>1>=2            Stem : Oleaf 

   GO enrichment computation: 
      Column       Method                         Cutoff
      RoSt         goSeqNoFDR.R                   5.0e-02
      RoOl         goSeqNoFDR.R                   5.0e-02
      StOl         goSeqNoFDR.R                   5.0e-02

-------------------------------------------------------------------
LEGEND:
   annoDB:
      ANNODB    is DBTYPE-TAXO, which is the DBtype and taxonomy
      ONLY      #Seqs that hit the annoDB and no others
      BITS      #Seqs with the overall best bitscore from the annoDB
      ANNO      #Seqs with the overall best annotation from the annoDB 
      UNIQUE    #Unique hits to the annoDB
      TOTAL     #Total seq-hit pairs for the annoDB
      AVG %SIM  Average percent similarity of the total seq-hit pairs
      HIT-SEQ   Percent of #Seqs that have at least one hit from the annoDB
      BEST HIT  The following columns refer to the best hit (Rank=1):
         AVG %SIM  Average percent similarity of the best hit seq-hit pairs
         Cover>=N  Percent of HIT-SEQ where the best hit has similarity>=N% and hit coverage>=N%

   #Seqs is listed at top of overview
   Best Annotation:
      Descriptions may not contain words such as 'uncharacterized protein'

</pre>
</body>
</html>
