<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
<title>Add or Edit annoDB</title>
 <body style="padding:20px;width:650px;font-family: Times;">
 <h2>Add annoDB</h2>
    Define the annoDBs (e.g. UniProt taxonomic databases) to search against. 
    UniProt is strongly supported as it will 
    also provide GO annotations if desired (from the .dat files).
    Additionally, it is highly recommended that you obtain <tt>Diamond</tt> 
    since it is significally faster than <tt>Blast</tt>. 
	<p>
	<table border=0 width="100%" cellpadding=3>
		<tr>
			<td valign="top">Taxonomy</td>
			<td>
				The taxonomy of an annoDB (e.g. 'plants') or it can be some other identifier. 
				<br><b><i>Naming Rules</i></b>:
				<ul>
				<li>The annoDB is named by its DBtype + Taxonomy, e.g. "SP-plants".
				
				<li>DBtype: If the annotation file uses definition lines like
				">sp|Q6V4H0|10HGO_CATRO", then the two letters before the "|" is used as the DBtype, 
				e.g. "SP" for SwissProt, "TR" for TrEMBL.
				Otherwise, the DBtype is "PR" for proteins files and "NT" for nucleotide files.
				
				<li>Often the annoDB name is abbreviated in viewSingleTCW (e.g. SPpla). Hence, it is important
				that the DBtype plus the first 3 letters of the taxonomy are unique.
				</ul>
				You can query viewSingleTCW for all hits of a given taxonomy and view the taxonomy
				as a column. 
			   </td>
		</tr>
			<td valign="top" nowrap>annoDB FASTA File</td>
			<td>For "Generate Hit Tabular File", the sequences will be searched against this file.
			For both the "Generate" and the "Use Existing" options, this is used to extract
			the description and species information.
		<tr>
			<td valign="top" nowrap>Date of annoDB download</td>
			<td>This will be used as the date of the annoDB. 
			<br>FORMAT: yyyy-mm-dd
			<br>If none is supplied, the date of the file on disk is used. 
			</td>
		</tr>
		
		<tr><td valign="top" nowrap>Generate Hit Tabular File
			 <td>The sequences will be searched against the FASTA file and loaded into the database.
		<tr>
			<td valign="top" nowrap>&nbsp;&nbsp;&nbsp;&nbsp;Search program</td>
			<td>
			<ul>
			<li><tt>Blast</tt> - this should be in your path or the path should be defined in HOSTS.cfg
			<li><tt>Diamond</tt> - if you want to use this problem, the path should be defined in HOSTS.cfg
		    </ul>
		    <i><b>It is strongly advised to use <tt>Diamond</tt></i></b> as it is so much faster; 
		    the TCW Diamond default parameters
		    have been tested to make it perform as well on transcript annotation with
		    protein databases. 
		    </td>
		</tr>
		
		<tr>
			<td valign="top" nowrap>&nbsp;&nbsp;&nbsp;&nbsp;Parameters</td>
			<td>Add or change the search parameters. TCW will automatically add to the parameters the number of CPUs to use
			and the output file name.
			<br>See the documentation for the search program for explanation of their parameters. 
			For example, the Diamond parameter "--masking 0" turns off masking; TCW tests show that
			this can often get exact hits that are not found with masking turned on, but it can also
			get more false positives.
			<br><i><b>TCW does not check the correctness of changed parameters.</b></i> 
			</td>
		</tr>
		
		<tr><td valign="top" nowrap>Use Existing Hit Tabular File 
			 <td>If the search has already been performed, load the existing file (see below).
			 
		<tr>
			<td valign="top" nowrap>&nbsp;&nbsp;&nbsp;&nbsp;Hit Tabular File</td>
			<td>User supplied tabular file of sequences searched against annoDB FASTA file </td>
		</tr>
		
	</table>
	
    
    <h4>Use Existing Hit Tabular File</h4>
    If the search has previously been performed by TCW, it will recognize that the file
    exists and will prompt you as to whether you want to re-use it. That is, you do not 
    need to enter the file here.
    
    <p>The corresponding annoDB FASTA file still needs to be provided to extract
			 the description and species.
			 
	<h2>Edit annoDB</h2>
	Only the taxonomy can be changed. Do not edit the 'annoDB FASTA File' name in the sTCW.cfg
	file or the change will not work. 
</body>
</html>