<html>
	<title>runSingleTCW</title>
	<body style="padding:20px;width:650px;font-family:Times">
	
		
<h2>singleTCW Main Window</h2>
<br>See www.agcol.arizona.edu/software/tcw/doc/stcw/UserGuide.html for a description on building 
a TCW database. It has highly recommended you run the demo before building your own. This
"Help" provides details.
<p>The main data type is the "sequence", which could be transcripts
with optional read counts, proteins with optional spectral counts, genes with locations,
or TCW assembled contigs.  If sequences (e.g. ESTs) are assembled,
the resulting contig consensus sequence will have read counts from the assembly.
<h3>Add Project</h3> 
You will be prompted for a <i>project-name</i> -- keep it short but descriptive. When you select
"Ok", a directory named <i>project-name</i> will be created
under the /projects directories, 
where information is stored for the building the project. Additionally, values will be automatically
entered for the following two:

<table border=0 cellspacing=5>
<tr><td valign="top" nowrap>singleTCW ID<td> The <i>project-name</i> is entered, which you may change.
Keep this name short but descriptive, e.g. about 3 letters.
<tr><td valign="top">Database<td>The <i>project-name</i> will be entered as the database name with a
"sTCW_" prefix -- do not remove the prefix. 
</table>
Though it is not required, you may put your dataset files in your project directory
under /projects (e.g.
see /TCW/projects/demoTra). You may create this directory yourself, which will
automatically be listed on the pull-down "Project" selector.

<h3>Existing project</h3>
<table border=0 cellspacing=5>
<tr><td valign="top">Project<td>
Allows you to select a project, where your projects are the directories under
the /projects directory. You can either click on the "Select..." to view a drop-down
of projects, or select the arrows to scroll through them.
<tr><td valign="top">Save<td> 
Save all information you have added. The information is also saved
whenever you execute "Keep" from any window, 
or a function ("Build Database", "Instantiate" and "Annotate").
Information about the datasets is stored in /projects/<i>project-name</i>/LIB.cfg and
information about the instantiation and annotation is stored in 
/projects/<i>project-name</i>/sTCW.cfg.
 
<tr><td valign="top">Copy<td>

1. It creates a copy of the current 
/projects/<i>project-name</i> directories with the same name
but with "Cp" appended to them. 
<br>2. All the sub-directories and files in the 
/projects/<i>project-name</i> directory will be copied with the exception of
the /hitResults and /orfResults directories.
<br>3. The <i>singleTCW ID</i> and <i>sTCW database</i>
values will have "Cp" appended to them. 
<br>4. The three Steps will need to be run, where it will create an identical sTCW database unless
you change any of the variables first.

<tr><td valign="top">Remove</td><td valign="top"> 
For the current project only, the options to remove are:
<br>1. existing annotation from the sTCW database
<br>2. the sTCW database
<br>3. the hit files from disk
<br>4. all project files from disk (i.e. the current project
directory under /project)</td></tr>

<tr><td valign="top">Overview<td> Shows the overview page of the selected project. 
There will only be an overview page if "Build Database" has been executed.

</table>

<h3>Step 1. Build Database</h3>
Add one or more Sequence Datasets with the "Add" button. You can make changes with
the "Edit" button, but once you have executed "Build Database", you can only change the
Attribute data (e.g. tissue type).
<table border=0 cellspacing=5>
<tr><td valign="top">Add<td>When you select this button, a new panel will appear with a Help to give more
detail. Briefly, you must enter the name of a sequence file; you may also enter the name of the file of counts
(for conditions).  
The "Keep" will take you back to this main window,
where the conditions will be shown in the second table, and the Reps column will be initialized to 1.
<tr><td valign="top" nowrap>Define Replicates<td> This function
will allows you to group replicates per condition. 
<tr><td valign="top" nowrap>Edit Attributes<td>After you have defined replicates (if needed), 
then edit the attributes of the conditions
 with "Edit Attributes". Attributes can be added or edited after the database is created.
</table>

<h3>Step 2. Instantiation</h3>

<table border=0 cellspacing=5>
<tr><td valign="top" nowrap>Use Sequence Names from File<td><td>When the input sequence will not be assembled, 
this option may be checked to use
the original sequence names, though the names <b>must</b> be composed of characters, digits or "_", and may not
be greater than 25 characters.
<tr><td valign="top">Skip Assembly<td>When the input sequences will be assembled, uncheck this box. The Options may be used
to define details of the assembly, though the default parameters are well-tested.
<tr><td  valign="top">#CPU<td>For assembly, if your machine has more than one CPU (thread/processor),
make sure that the "#CPUs" is set to the number in order to speed up processing.
</table>

The TPM (Transcripts Per Kilobase Million) is computed.
To compute RPKM (Reads Per Kilobase Million) instead, see 'execAssm -h'.


<h3>Step 3. Annotate</h3>
Add your annoDBs and Options (i.e. select the GO database), then execute "Step 3: Annotate".
<p>
Before running this step, 
you download the protein and/or nucleotide databases to search against (these are referred to as 
<i>annoDBs</i>), the resulting files are called <i>hit files</i>. 
If you are using UniProt, you may also create the <i>GO database</i>, which
provides GO, EC, PFam and KEGG identifiers. To download UniProt and GO, and create the GO database,
use the interface <b>runAS</b>.
<p>
If you use databases other then UniProt, you need to make sure there that the descriptor line
for each sequence (i.e. > lines) is formated correctly. Genbank formats are accepted,
plus a generic format (see the online Annotation Setup).
<p>You have a choice of using <tt>blast</tt>, <tt>usearch</tt> or <tt>diamond</tt> for searching the database sequences
against an annoDB. See <tt>HOSTS.cfg</tt> for further information on locations of these programs.
You can either request
a particular search program when you add an annoDB, or use the default "TCW Select" to let
TCW select an appropriate program (see Help on Add/Edit page). 

<br>When you select "Step 3: Annotate", there will be a popup with the choice:
<ul>
<li>Answer - Answer prompts at command line.
<br>All parameters are written out to the terminal. There may be questions if
you have run this before, e.g. whether you want to re-use a previous hit file.
<li>Use Defaults - Run with defaults
<br>The defaults are to add to any existing annotation and to re-use any previous
hit files.
</ul>

<p>Annotation Interface:  a few details.
<ul>
<li>Import AnnoDBs: you may select the TCW.anno file created in <tt>runAS</tt>, or a sTCW.cfg
file of another project; in both cases, the UniProts listed in the file will be used to populate
the annoDB table.
<li>Click a line once to selected it, than a second time to check or un-check the box.
<li>The annoDBs can be re-order with the "Move Up" and "Move Down" according
to their importance, where they will be shown in that order in the <tt>viewSingleTCW</tt> overview.
</ul>

<p>Add to annotation: After the initial annotation, you can add additional annoDBs, as follows:
<ol>
<li>Uncheck all annoDBs that have been added (they are in italics). You need to click the
line twice, once to highlight it and once to uncheck.
<li>Add any additional annoDBs and check them.
<li>Select in "Options" if you want to add similarity pairs.
<li>Execute "Step 3: Annotate".
<br>You will be prompt at the command line as to whether you want to 
<br><tt>
Add to existing annotation [a], Delete annotation [d], Exit [e]: a
</tt>
</ol>


<p>Remove annotation:
<ol>
<li>All annoDBs from database: Use the "Remove..." button at the top to remove the existing annotation from the database.
<li>Remove one annoDB from database: There is no way to do this. You must (a) remove all annotation,
(b) uncheck any annoDBs you do not want, (3) "Step 3: Annotate" to reload the annoDBs, which will
use the existing hit files so will not take as long (but can still take a few hours on a large database).
<li>From the table: (a) "Remove All" removes all entries from the table that are not in the database.
(b) Select a entry in the table followed by "Remove" on the right.
</ol>

<h4>ORF only</h4>
Execute the ORF finder on all sequences. This must be executed after "Step 2: Instantiate".

<h4>GO only</h4>
If "Annotate Sequences" was already run and you just want to add GO annotation to it, or update previous
GO annotation, use this function. TCW GO annotation requires a reference mySQL "GO database"
which must be created before hand using <tt>runAS</tt> or the supplied perl scripts. 
Note, this also adds Pfam, KEGG, Interpro and EC for the UniProts hit in your database. 

<h3>Options along the bottom</h3>

<h4>Launch viewSingleTCW</h4>
Launches this program to view the resulting project.

<h4>Add Remarks or Locations</h4>
You can add remarks or locations from a file, which can be displayed in a column in any table and can
be searched on viewSingleTCW. Click this "Add Remarks or Locations" button
to see additional Help.

<h3>Troubleshooting</h3>
<ol>
<li>If you have added a "Sequence Dataset" to the interface, you can not change anything except the
attributes, even if "Load Data" has not be executed. To make a change, remove the dataset and re-add. 
<li>After a TCW database has been created, runSingleTCW can be used to edit
the attributes (i.e. tissue, etc). This ONLY works right from the TCW directory that has 
the /project subdirectories, as it needs to read the LIB.cfg and sTCW.cfg 
project files. 
<li>runSingleTCW will detect certain file format problems, etc, 
but some errors are not found until "Load Data" is executed.
If that happens, remove the mySQL database, fix the error and restart.
<li>If something appears wrong and you do not know how to fix it, you can remove
the /projects/&lt;your project name&gt;/LIB.cfg file and restart. 
<li>If any process seems to "hang", i.e. no output to terminal for a very long time,
either close the runSingleTCW window or ctrl-C. Then just restart. NOTE: on Macs,
pop-up windows get hidden easy and freeze the main singleTCW window until you acknowledge
the pop-up, so look around for it...
</ol>

