<!DOCTYPE html>
<html>
<!--#set var=TITLE value="Tour: viewMulti Alignment" -->
<!--#include virtual="../../ssi/head.html" -->
<body  onLoad="init('align'); init2('viewM')">
<a id=top></a>
<!--#include virtual="../ssi/start2t.html" -->
<!--#include virtual="./ssi/view_mtcw.html" -->

<div style="text-align: center"><big><ttc>Alignment</ttc></big></div>
<table style="padding: 2;  border-spacing:0; border-collapse:collapse;  margin-left: auto; margin-right: auto;">
<tr><td style="background-color: #666666; padding-bottom:4px; "> 
<a href="#pair" 	class="white">Pairwise</a> 	<mark class="white">|</mark>
<a href="#msadb" 	class="white">MSAdb</a> 	<mark class="white">|</mark>
<a href="#msa" 		class="white">MSA...</a>
</td></tr>
</table>

<a id=table></a>
<h3>Alignments</h3> 


<table>
<tr>
	<td class="top"><small>Cluster table with MSA</small>
		<br><a href="img/clusterTable.png"><img  src="img/clusterTable.png" alt="" style="border: 1px solid black; width: 450px"></a>
	<td class="top"><small>Pair table with Pairwise</small>
		<br><a href="img/pairTable.png"><img src="img/pairTable.png" alt="" style="border: 1px solid black; width: 450px"></a>

</table>
The <ttc>Hit Table</ttc> and <ttc>Sequence Details</ttc> also have the <ttc>Pairwise...</ttc> option, and the <ttc>Sequence Table</ttc>
has the <ttc>MSA...</ttc> option.
<p>If there are only protein (AA) sequences in the database, only the AA options described below will be available.

<a id=pair></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h3>Pairwise...</h3> 
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>


<table style="width:60%">
<tr>
	<td>
	The <ttc>Pairwise...</ttc> drop-down has the options shown on the right, where all options use dynamic programming for the selected sequence(s).
	<p>The "for one pair" options are enabled when only one pair is selected.
	<td><a href="img/pairOpt.png"><img  src="img/pairOpt.png" alt="" style="border: 1px solid black; width: 200px"></a>
</table>

<br>AA, CDS and NT alignments of a pair. The text alignment can be viewed by clicking the AA, CDS or NT buttons.
<br><a href="img/pairAlign.png"><img  src="img/pairAlign.png" alt="" style="border: 1px solid black; width: 700px"></a>
<br><br>

<table>
<tr><td>
	Menu for the CDS text alignment: 
	<br><a href="img/alignMenu.png"><img  src="img/alignMenu.png" alt="" style="border: 1px solid black; width: 300px"></a>
	<td class=top>
	Text alignment of the CDS showing matches. 
	<br>The '|' are mutations in synonymous codons, the '*' mutations in nonsynonymous codons.
	<br><a href="img/alignText.png"><img src="img/alignText.png" alt="" style="border: 1px solid black; width: 600px"></a>
</table>

<a id=msadb></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h3>MSAdb</h3> 
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

When the database is built, the cluster  MSA is computed in stored in the database.
<br><a href="img/MSAdb.png"><img  src="img/MSAdb.png" alt="" style="border: 1px solid black; width: 700px"></a>
<p><a href="img/MSAzappo.png"><img  src="img/MSAzappo.png" alt="" style="border: 1px solid black; width: 700px"></a>

<a id=msa></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h3>MSA...</h3> 
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>


<table style="width: 70%">
<tr>
	<td>The <ttc>MSA...</ttc> drop-down has the options shown on the right, where all options execute the MSA program for the set of selected sequences.
	<p>AA is the translated ORF, CDS is the un-translated ORF, and NT is the entire transcript.
	<td><img src="img/msaOpt.png" alt="" style="border: 1px solid black; width: 120px">	
</table>
The display is the same as shown above for MSAdb.

<table style="width: 100%"><tr><td style="text-align: left">
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<!--#include virtual="../../ssi/end.html" -->
</body>
</html>
