<html>
<title>mTCW overview</title>

<style>
.nw {vertical-align: top; white-space: nowrap;}	
.top {vertical-align: top; }	
table.y, .y td, th { border: 1px solid black; border-collapse: collapse; border-spacing: 5px;}
th, td {padding: 5px;}
</style>
</head>

<body style="font-family: Arial, sans-serif;  font-size: 14px;">
<table style="width:70%; margin-left: auto; margin-right: auto;">
<tr><td>&nbsp;
<tr><td>&nbsp;

<tr><td>
<h2>Explain mTCW overview</h2>
If the mTCWdb was created with only AA sequences, the Pairs statistics will not be present.
The top line may have any of the labels "PCC Stats Multi KaKs"; these indicate that the database
has the corresponding information.

<p>Names in the "Compute" column refer to table (Cluster, Pair, Seq) Columns unless stated otherwise.

<table border=1 cellpadding=5>
<tr><th>Label<th>Description<th>Compute
<tr><td nowrap valign="top"><b>DATASETS</b>
<td colspan=2>
This first section describes the data imported from the sTCWdbs. 

<tr><td nowrap valign="top"><b>CLUSTER SETS</b>
<td colspan=2>One or more sets of clusters will be present in the mTCW. The columns are in the Cluster table.
<tr><td nowrap valign="top" colspan=3><i>Sizes</i>
<tr><td nowrap valign="top">&nbsp;&nbsp;[Range]
<td>The range of cluster sizes per cluster set. These may be computed from viewing all
Clusters sorted on the 'Count' column.
<td>Count
<tr><td colspan=3><i>Statistics</i>
<tr><td>&nbsp;&nbsp;conLen<td>Average consensus length of the cluster<td>Avg(conLen)<sup>1</sup>
<tr><td>&nbsp;&nbsp;sdLen<td>Average standard deviation of the lengths of 
the amino acid sequences in the cluster.<td>Avg(sdLen)<sup>1</sup>
<tr><td>&nbsp;&nbsp;Score1<td>The MSA scoring method is shown in the PROCESSING section.  
 <td>Avg(Score1)<sup>1</sup>
<tr><td>&nbsp;&nbsp;SD<td>Standard deviation of Score1. 
<td>StdDev(Score1)<sup>1</sup>
<tr><td>&nbsp;&nbsp;Score2<td>The MSA scoring method is shown in the PROCESSING section.
<td>Avg(Score2)<sup>1</sup>
<tr><td>&nbsp;&nbsp;SD<td>Standard deviation of Score2. 
<td>StdDev(Score2)<sup>1</sup>

<tr><td nowrap valign="top"><b>PAIRS</b>
	<td colspan=2>Similar pairs were identified from the comparing the 
	sequences using a search program (BLAST or DIAMOND), 
	where AA is from the amino acid search and NT is from the nucleotide search (NT may not exist).
	The columns are in the Pairs table. 

<tr><td nowrap valign="top" colspan=3><i>Hits</i> (from hit file) The following is the same for NT.
    
<tr><td nowrap valign="top">&nbsp;&nbsp;Diff
	<td>The number of hits from different datasets.
	<td> Pair Filter: Has AA, Different sets.
<tr><td nowrap valign="top">&nbsp;&nbsp;Same
	<td>The number of hits from the same dataset. 
	<td>Pair Filter: Has AA, Same sets.
<tr><td nowrap valign="top">&nbsp;&nbsp;Similarity	
	<td>Percent similarity of all alignments 
	<br>(Average percent similarity of each alignment).
	<td>see<sup>1b</sup> 
	<br>((Avg(%AAsim))<sup>1</sup>
	
<tr><td nowrap valign="top">&nbsp;&nbsp;Coverage	
	<td>Percent overlap of all alignments 
	<br>(Average percent alignment of the length of each sequence in the pair).
	
	<td>see<sup>1b</sup>
	<br>((Avg(%AAcov1)+
	<br>Avg(%AAcov2))/2)<sup>1</sup>
	
<tr><td nowrap valign="top" colspan=3><i>Aligned</i> (aligned using dynamic programming, Filter: Has Stats)

<tr><td>&nbsp;&nbsp;CDS:<td>The number of aligned CDS bases including gaps but not overhangs.<td>Sum(Align)<sup>1</sup>
<tr><td>&nbsp;&nbsp;5UTR:<td>Same as CDS but for 5UTR.<td>see<sup>2</sup>
<tr><td>&nbsp;&nbsp;3UTR:<td>Same as CDS but for 3UTR.<td>see<sup>2</sup>
    <td colspan=2> 


<tr><td colspan=3><i>Codon column</i>
<tr><td nowrap valign="top">&nbsp;&nbsp;Codons
    <td>Number of aligned codons excluding gaps. 
	<td>Sum(Calign)<sup>1</sup>
	
<tr><td nowrap valign="top">&nbsp;&nbsp;Exact
    <td>Percent of the number of codons that are exactly the same.
	<td>%Cexact; see<sup>1b</sup>
	
<tr><td nowrap valign="top">&nbsp;&nbsp;Synonymous
    <td>Percent of the number of codons that are synonymous (different codon, same amino acid).
    <td>%Csyn; see<sup>1b</sup>
    
<tr><td nowrap valign="top">&nbsp;&nbsp;&nbsp;&nbsp;Fourfold
    <td>Percent of the number codons that are fourfold (4d), which are two synonymous codons 
    where the ith position allows any of the 4 bases. 
    <td>%C4d; see<sup>1b</sup>
    
<tr><td nowrap valign="top">&nbsp;&nbsp;&nbsp;&nbsp;Twofold
    <td>Percent of the number of codons that are fourfold (2d), which are two synonymous codons 
    where the ith position allows any of the 2 bases. 
    <td>%C2d; see<sup>1b</sup>
    
<tr><td nowrap valign="top">&nbsp;&nbsp;Nonsynonymous
    <td>Percent of the number of codons that are nonsynonynous (different amino acid).
    <td>%CnonSyn; see<sup>1b</sup> 
        
<tr><td colspan=3><i>Amino acid column</i> (same number as Codons)
     
<tr><td nowrap valign="top">&nbsp;&nbsp;Exact
    <td>Percent of the AA characters that are the same.
    <td>%Aexact; see<sup>1b</sup>
    
<tr><td nowrap valign="top">&nbsp;&nbsp;Substitution&gt;0
    <td>Percent of AA characters that are substitutions with BLOSUM62&gt;0.
    <td>%Apos; see<sup>1b</sup>
    
<tr><td nowrap valign="top">&nbsp;&nbsp;Substitution&lt;=0
    <td>Percent of AA characters that are substitution with BLOSUM62&lt;=0.
	<td>%Aneg; see<sup>1b</sup>

<tr><td nowrap valign="top" colspan=3><i>Nucleotides columns </i>
    (References to CDS, 5UTR, 3UTR are the numbers by 'Aligned:')

<tr><td nowrap valign="top">&nbsp;&nbsp;CDS Diff
    <td>Percent of CDS bases that are not the same, i.e. ((Gap+SNP)/Align)*100.0. 
       <td>%Cdiff; see<sup>1b</sup>
  
  <tr><td nowrap valign="top">&nbsp;&nbsp;&nbsp;&nbsp;Gaps
    <td>Percent of CDS bases that are Gaps, i.e. (Gaps/Align)*100.0
    <td>see<sup>1b</sup>
    
   <tr><td nowrap valign="top">&nbsp;&nbsp;&nbsp;&nbsp;SNPs
    <td>Percent of CDS bases that are SNPs, i.e (SNP/Align)*100.0.
    <td>see<sup>1b</sup>
           
<tr><td nowrap valign="top">&nbsp;&nbsp;5UTR Diff
    <td>Percent of 5UTR bases that are different. 
       <td>%5diff; see<sup>1</sup>
         
<tr><td nowrap valign="top">&nbsp;&nbsp;3UTR Diff
    <td>Percent of 3UTR bases that are not different. 
    <td>%3diff; see<sup>1</sup>
         
<tr><td colspan=3><i>Columns: Pos1   Pos2   Pos3  Total</i><sup>4</sup>
<tr><td nowrap valign="top">&nbsp;&nbsp;Transition (ts)
    <td>Percent of SNPs that are transitions in each of the three codon positions.
       <td>see<sup>2</sup>
<tr><td nowrap valign="top">&nbsp;&nbsp;Transversion (tv)
    <td>Percent of SNPs that are transversion in each of the three codon positions.
       <td>see<sup>2</sup>
<tr><td nowrap valign="top">&nbsp;&nbsp;ts/tv
    <td>The total number of transitions divided the total number of transversions.
        <td>ts/tv; see<sup>2</sup>
    
<tr><td colspan=3><i>Columns: GC  CpG-Nt  CpG-Cd </i><sup>4</sup>
<tr><td nowrap valign="top">&nbsp;&nbsp;Both
    <td>Percent of CDS bases where both (union) sequences have a GC base. 
    Same for the CpG sites, where CpG-Nt (nucleotide) and CpG-Cd (codon, CpG does not cross codon boundaries).
    <td>see<sup>3</sup>
 <tr><td nowrap valign="top">&nbsp;&nbsp;Either
    <td>Percent of CDS bases where either or both (intersection) sequences have a GC base. 
    Same for the CpG sites, where CpG-Nt (nucleotide) and CpG-Cd (codon, CpG does not cross codon boundaries).
    <td>see<sup>3</sup>
     
  <tr><td nowrap valign="top">&nbsp;&nbsp;Jaccard
    <td>The total number of 'both' divided by the total number of 'either'.
    There is a pair column for the GC and codon-based CpG-Cd, but not the nucleotide-base CpG-Nt.
    <td>GC-JI, 
    <br>CpG-JI; see<sup>3</sup>
    
<tr><td nowrap valign="top" colspan=3><i>KaKs method</i><sup>5</sup>
<tr><td><i>Ka/Ks</i><td>Selective strength. A score &lt;1 is purifying, &gt;1 is positive, 
and significantly &gt;1 implies that at least some of the mutations are probably advantageous. 
It is rare for KaKs to be exactly 1, so the following rules are used:
<pre>
If KaKs &gt;= 0.995 and KaKs &lt; 1.006, then KaKs~1
If KaKs &gt;= 1.006, then KaKs &gt; 1
If KaKs &lt;  0.995, then KaKs &lt; 1
</pre>
<td>Set Filter according to rules on the left.

<tr><td><i>Quartiles</i><td>Of the KaKs values. It uses the method of splitting the list in half; 
Q1 is the median of the lower half and Q3 is the median of the upper
half.

<td>Q2=Median(KaKs)<sup>1</sup>
<tr><td nowrap valign="top" colspan=3><i>Average
<tr><td>&nbsp;&nbsp;Ka<td>Nonsynonymous substitution rate<td>Avg(Ka)<sup>1</sup> 
<tr><td>&nbsp;&nbsp;Ks<td>Synonymous substitution rate<td>Avg(Ks)<sup>1</sup>
<tr><td>&nbsp;&nbsp;P-value<td>KaKs p-value<td>Avg(p-value)<sup>1</sup>

<tr><td><i>P-value</i><td>The 'other' count includes pairs where the KaKs_calculator did not provide
a p-value though did provide a KaKs value.<td>p-value

<tr><td nowrap valign="top"><b>SEQUENCES</b>
<td colspan=2>
The columns are in the Sequence table; use Sequence Filters to select a dataset
to view its corresponding results.

<tr><td><i>Average Lengths</i>
<td>The ORFs were computed for the sTCWdbs and imported along with their translated sequence.
<td>5UTR Len, 3UTR Len, CDS Len; see<sup>1b</sup>

<tr><td><i>%GC</i>
<td>The average percent of GC for 5'UTR, CDS and 3'UTR.
The only %GC column is for the entire sequence.
<td>-

<tr><td><i>CpG O/E</i>
<td>The CpG observed/expected for the 5'UTR, CDS and 3'UTR [(#CpG/(#G*#C))*Len].

<td>5UTR CpG, CDS CpG, 3UTR CpG; see<sup>1b</sup> 

<tr><td><i>Counts</i>
<td colspan=2>The total raw counts from each dataset. These can be verified from the singleTCWs.
The TPM values are in the sequence table, but not the counts, though these can be viewed
in the Sequence Detail panel.

<tr><td><i>Differential Expression</i>
<td  colspan=2>The total DE from each dataset. These can be verified from the singleTCWs.

</table>
<table border=0 cellpadding=3> 
<tr><td  valign=top><sup>1</sup>
<td><small>For a Pair table, select the  "Table.../Show Column Stats" which provides "Avg" and "Sum".
</small>
<tr><td  valign=top><sup>1b</sup>
<td><small>Percents for the Overview are from summing the numerator and denominator before  dividing. 
<br>The Average of a percent column of 
the "Table.../Show Column Stats" is taking the the average of the percentages. 
<br>For example, in the mTCW demo, 
the <i>Exact number of all aligned bases</i> is 58.6%, whereas the Column
stats "Average" is the <i>average of all align to length percentages</i>, which is 60.286. 
<p>The only way to view the number of Exact, Synonymous, etc codons is, for a given pair:
<br>(a) Sequence Table, (b) Pairwise..., (c) AA,CDS,NT, (d) Align CDS..., (e) Match
(f) Codon Exact (and other counts) at the top.
</small>
<tr><td valign=top><sup>2</sup> <td><small>The only way to view the number of aligned UTR 
bases is, for a given pair: 
<br>(a) Sequence Table, (b) Pairwise..., (c) 5UTR,CDS,3UTR, (d) Align 5UTR..., (e) Match
(f) CROP is the number of aligned UTR bases.
</small>
<tr><td  valign=top><sup>3</sup> <td><small>
(a) Sequence Table, (b) Pairwise..., (c) 5UTR,CDS,3UTR, (d) Align CDS..., 
(e) ts/tv, (f) The ts and tv numbers are at the top.
</small>
<tr><td  valign=top><sup>4</sup> <td><small>
(a) Sequence Table, (b) Pairwise..., (c) 5UTR,CDS,3UTR, (d) Align CDS..., (e) CpG, 
(f) The GC and CpG numbers are at the top, where
<br>"All CpG" is "CpG-Nt" and "By Codon" is "CpG-Cd".
</small>

<tr><td  valign=top><sup>5</sup> <td><small>The KaKs_calculator (Zhang et al. 2006) will typically be used to 
    compute the KaKs values, where the method
    used is shown on the "KaKs Method" line. 
</small>
</table>
</center>
</html>