<html>
<title>Search</title>
<body style="padding:20px;width:650px;font-family:Times">
<h2>Searching</h2>
Search sequence(s) against mTCW NT-sequences, mTCW AA-sequences, or a selected database
of protein sequences. 
<ol>
<li><i>Query:</i> Input one or more sequence in FASTA format into the entry box. 
<ul>
<li>If your sequence contains characters other than A,G,C,T, or N, it will
be treated as a protein sequence.
<li>From the various tables and detail panel, you can use the <b>Copy...</b> command 
to copy a hit sequence or the seqID sequence, and <b>Paste</b> it here.
</ul>
<p>
<li><i>Subject:</i> There will be the following options: 
<ul>
<li>NT-Seqs (nucleotide sequences in the database; these will not exist if the sTCWdb were all AA), 
<li>AA-Seqs (amino acid sequences in the database), 
<li>AA-DB  (selected protein file with FASTA suffix: .fa, .fasta, .fna, .ffn, .faa, or .frn)
</ul>
<p>
<li>You have a choice to use <tt>Blast</tt> and <tt>Diamond</tt>. 
<tt>Diamond</tt> can only be used with an AA subject.

<p> 
<li>You may enter/change
parameters, though you must <i>make sure they are correct</i> since TCW does not verify them.
A list of some of the possible parameters are shown in the "Search Parameters" Help page.

<p>
<li>Select <b>RUN SEARCH</b> to execute the search.
The results are shown in the bottom half of the panel, and are also written
to a file in the "/ResultHits" directory. 

<p>
<li>If tabular format is chosen, then you can
select a row followed by <b>Copy Selected Subject</b>, which will copy to the clipboard
the subject value. This value can be pasted in an appropriate <b>Filters</b> text box
to find more information about it.


<p>
<li><b>Delete Files</b>: Files written by the TCW search command will go into the "/ResultHits"
directory. These files will end with ".fa" or be formated files. These files will be
deleted when you execute this command. Result files will not be deleted. If you select a protein database that is
not in the "/ResultsHits" directory, the formatted files are not deleted. 
</ol>
 

<h3>No Results Found</h3>
If you do not get the results you expect:
<ol>
<li>Remember that the original annotation is typically 1000's of sequences compared with
1000's of annotation sequences, whereas this compares just a few sequences against 1000's, which
causes the e-values to be different.
<br><br>
<li>Select <b>Trace</b> and <b>RUN SEARCH</b> again. The command is written to 
the terminal; copy it and run it from the terminal. If there was
any error (e.g. in the parameters), it will tell you the error. 
</ol>

