<html>
<body style="padding:20px;width:650px;">
<title>Pair Table</title>

<h2>Pair Table</h2>
There is a pair for a given two sequences if they have a NT (nucleotide)
of AA (amino acid) hit, which are computed when the mTCWdb is created. 
There are only coding statistics and KaKs values
for pairs in clusters (it depends how the database was built -- these values
may only exist for certain cluster sets).

<h3>Functions at the top</h3>

<u>Sequences</u> Select a pair followed by this function to view the two sequences
in the sequence table, where the alignment can be viewed.
<p>
<u>Clusters</u> Select a pair followed by this function to view the clusters for the two sequences.
<p>
<u>Copy...</u> For the selected pair, copy to clipboard a column (SeqID1, SeqID2, HitID or Hit Description).
<p>

<u>Table...</u>
<ul>
<li><b>Show Columns Stats:</b> Popup a table of statistics for the numeric columns in the table. 
<br>&nbsp;
<li><b>Show Table Stats:</b> (NT-mTCW only) Popup a table of the coding and KaKs statistics for the pairs in the
table. <i>This only works for pairs that have coding sequence statistics.</i>
<br>&nbsp;
<li><b>Copy Table</b>: Copies the table to the clipboard either the entire table. 
<br>&nbsp;
<li><b>Export Table</b>: Exports the table to file. 
<br>&nbsp;
<li><b>Export both AA/NT sequences</b>: Exports both AA (amino acid) or NT (nucleotide) sequences for all pairs in the table.
<br>&nbsp;
<li><b>Export 1st AA/NT sequences</b>: Exports the 1st AA or NT sequence of the pair for all pairs in the table.
<br>&nbsp;

<li><b>Export Pair GOs:</b> list GOs associated with the pair sequences in the table. 
If either
sequence in the pair has a hit with a GO the passes the below filters, the pair is
counted as having the GO. NOTE: it does NOT use the Pair "Best Anno Hit". The first three are filters::
<ol>
<li><i>GO Level</i>: 
	<ol type='a'>
	<li>Level 1 is the three ontologies (top level). Level 2 are the GO terms directly beneath
the top level, etc. 
	<li>If you leave this blank, it will output all GOs. 
	<li>If you just want a couple of levels, e.g. 2 and 3, first export level 2, then export level 3
to the same file and use the append option (there will be a pop-up asking whether you want to append).
	</ol>

<li><i>Any Ontology</i>: This toggles through the three main ontologies. If you select one, it will only 
export the GOs in that ontology.

<li><i>E-value</i>: If you leave this blank, it will ignore the E-value. Otherwise,
at least one sequence in the cluster must have a seq-hit E-value&le; this value.

<li><i>Count limit</i>: If a GO term has less than this count of GO-seq assignments,
it is not written to file.
<li><i>File Columns</i>:
	<ol type='a'>
	<li>Verbose: outputs columns (GO is the column for the ontology abbreviation)
	<tt>
	GO-ID	#Pair	Level	GO	GO-Name
	</tt>
	<li>Desc/Count: outputs the name and count by ontology.
	<br>This can be used in Excel to create graphs.
	<li>GO/Count: only outputs the <tt>GO ID</tt> and <tt>#Pairs</tt>. 
	<br>This can be used as input to input to REVIGO (http://revigo.irb.hr)
	</ol>
</ul>

<h2>Columns</h2>

<ul> 
<li>Put the mouse over a column name in the <b>Select Columns</b> interface, and a full explanation of the column is shown
in the lower left corner.
<li>Columns can be added/removed using the <b>Select Columns</b> button below the table. 
<li>Columns can
also be rearranged (by dragging the headers) and sorted (by clicking the headers). 

<li>One or more pair rows can be selected using mouse click, shift-click, or ctrl-click.
</ul>
The columns available depend on whether the mTCW database was build with nucleotide or protein
sequences. 
The sections are as follows, where most columns are adequately explain with the mouse-over,
except the following:

<table border=0 cellspacing=3>
<tr><td><b>General</b><td>&nbsp;
<tr><td valign="top">&nbsp;&nbsp;HitID<td>The best shared hit from the pair
(see the <a href=http://www.agcol.arizona.edu/software/tcw/doc/mtcw/UserGuide.html#step2>online documentation</a>
for a description of how this is computed). 
<br>The <tt>HitID</tt> "*NoShare" indicates the pair does not share any description.
<br>The <tt>HitID</tt> "*Novel" indicates that neither sequence of the pair has any annotation.

<tr><td valign="top">&nbsp;&nbsp;PCC<td>Pearson Correlation Coefficient of the TPM values.
This only has values>0 if the datasets have matching condition names.

<tr><td valign="top"><b>Hit</b><td>Results from protein and DNA sequence self-blast (if available). All
columns are directly from <tt>Blast</tt> or <tt>Diamond</tt> output except the 'cov' columns. 

<tr><td valign="top" nowrap>&nbsp;&nbsp;Cov1, Cov2<td>The percent coverage, which is computed from the
'align' divided by sequence length. <i>The Cov (percent coverage) can be &gt;1;</i> this happens when
the alignment covers the whole sequence with gaps.

<tr><td valign="top"><b>KaKs</b><td> (NT-mTCW only)
Ka is the synonymous substitution rate. Ks is the nonsynonymous
substitution rate. KaKs is Ka/Ks. The p-value applies to the KaKs value. 

</table>

</html>