<html>
<body style="padding:20px;width:650px;">
<title>Sequence Table</title>
<h2>Sequence Table</h2>

This table can be viewed from <b>Samples Sequences</b> or <b>Filters Sequence</b>, or 
selecting a row in the cluster, pairs or hit table.

<h3>Row of functions at the top</h3>

<u>Details</u> For the selected sequence: show counts, TPM values, 
and annotations. 

<p><u>Clusters</u> For the selected sequence(s): 
Show the table of clusters containing the sequence.

<p><u>Pairs</u> For the selected sequence(s): 
Show the table of pairs containing at least one of the selected sequences.

<p><u>MSA...</u> Select a set of sequences followed by this button, which provides a drop-down of actions.
All the following actions display the alignment as a tab on the left, and provides the alignment file in the <tt>ResultAlign</tt> directory.
<ul>
<li><b>MUSCLE-AA</b>: 
Run MUSCLE to build a multiple alignment 
for all amino acid sequences from the selected set. This can be very time consuming
when run on many sequences. The command executed is:
<pre>
/Ext/linux/muscle/muscle -in ResultAlign/inSeq.fa -out ResultAlign/outAln.fa
</pre>

<li><b>MAFFT-AA</b>: Same as above.
The command executed is:
<pre>
/Ext/linux/mafft/mafft.bat --auto --reorder --thread N ResultAlign/inSeq.fa
</pre>
Where N is the number of available threads (processors/CPUs).
<br><br>
<li><b>MAFFT-CDS</b>: (NT-mTCW only) Same as above but with the CDS sequences.
<br><br>
<li><b>MAFFT-NT</b>: (NT-mTCW only) Same as above but with the NT sequences.

</ul>

<p><u>Copy...</u> For the selected sequence: 
Copies the seqID, AA, NT or CDS sequence, Best hitID, Best description or Best hit sequence.

<p><u>Table...</u>
<ul>
<li><b>Show Column Stats</b>: Popup a table of statistics for the numeric columns in the table. 
<br>&nbsp;

<li><b>Copy Table</b>: Copies the table to the clipboard. 
<br>&nbsp;

<li><b>Export Table (.tsv)</b>: Exports the table to file in tab delimited file. 
<br>&nbsp;

<li><b>Export AA sequences (.fa)</b>: Export the translated CDS sequences..
<br>&nbsp;

<li><b>Export CDS sequences (.fa)</b>: (NT-mTCW only) Export the CDS sequences.
<br>&nbsp;

<li><b>Export NT sequences (.fa)</b>: (NT-mTCW only) Export the full nucleotide sequences.
<br>&nbsp;

<li><b>Export all GOs</b> (If GOs are in the database): Output all assigned and inherited GOs for all seq-hit pairs 
for all sequences in the table. 
The first three are filters::
<ol>
<li><i>GO Level</i>: 
<ol type='a'>
<li>Level 1 is the three ontologies (top level). Level 2 are the GO terms directly beneath
the top level, etc (see GO annotations 'GO Help' for more explanation). 
<li>If you leave this blank, it will output all GOs. 
<li>If you just want a couple of levels, e.g. 2 and 3, first export level 2, then export level 3
to the same file and use the append option (there will be a pop-up asking whether you want to append).
</ol>

<li><i>Any Ontology</i>: This toggles through the three main ontology. If you select one, it will only 
export the GOs in that ontology.

<li><i>E-value</i>: If you leave this blank, it will ignore the E-value. Otherwise,
the sequence must have at least one seq-hit with E-value&le; this value.

<li><i>File Columns</i>:
<ol type='a'>
<li>Verbose: outputs columns
<tt>
GO-ID	#Seqs	Level	Ont	GO-Name
</tt>
<li>Count only: only outputs the <tt>GO-ID</tt> and <tt>#Seqs</tt>. 
<br>This can be used as input to input to REVIGO (http://revigo.irb.hr)
</ol>
</ul>
</ul>

<h2>Columns</h2>
<ul> 
<li>Columns can be added/removed using the <b>Select Columns</b> button below the table. 
<li>Columns can
also be rearranged (by dragging the headers) and sorted (by clicking the headers). 
<li>Put the mouse over a column heading in the <b>Select Columns</b> interface, and a full explanation of the column is shown
in the lower left corner.
</ul>

</table>

