<!DOCTYPE html>
<html>
<!--#set var=TITLE value="Tour: runDE" -->
<!--#include virtual="../../ssi/head.html" -->
<body  onLoad="init('runDE')">
<a id=top></a>
<!--#include virtual="../ssi/start2t.html" -->

<div style="text-align: center"><big><ttp>runDE</ttp></big> - Differential Expression</div>
<table  class="tablex" style="margin-left: auto;margin-right: auto;">
<tr><td class="white" style="text-align:left; background-color:#666666; padding-bottom:4px;">
<a href="#main" class="white">runDE</a> <mark class="white">|</mark> 
<a href="#batch" class="white">Batch</a> <mark class="white">|</mark> 
<a href="#R" class="white">R-scripts</a> <mark class="white">|</mark> 
<a href="#refs"  class="white" >References</a>
</td></tr>
</table>	
<a id=main></a>
<p><ttp>runDE</ttp> adds differential expression (DE) to a singleTCW database.
It uses JRI (Java-R interface) to R methods. 
TCW supplies R-scripts for <ttp>edgeR<sup>1</sup></ttp>, <ttp>DEseq2<sup>2</sup></ttp> and <ttp>GOseq<sup>3</sup></ttp>. 
See <a href="../../stcw/DiffExp.html" class="ext" target="_blank">Differential Expression Guide</a> for details. 

<table class="tablex">
<tr>
	<td class=top>		
	<p>1. Select one condition from group 1 and another from group 2.
	<p>2. Select the R-script.
	<p>3. Only select <ttc>Fixed Dispersion</ttc> if no replicates. 
	<p>4. Click <ttc>Save results in p-value column</ttc> to save the results in the database. If the two conditions
	have been selected, it will provide a default name (that can be over-written).
	<p>5. <ttc>Group 1 - Group 2</ttc> will run the DE method and (if Save selected) add the results to the database.
	<p>6. When all columns have been added, select <ttc>Execute</ttc> below the <ttl>GO enrichment</ttl> label.
	<p><p>On the terminal window, it will be left in R so that you can execute
	R commands. 		
	</td>
	<td>
		<a id="Fig1"></a>
		<a  href="img/runDE.png">
		<img src="img/runDE.png" alt="" style="border: 1px solid black; width: 450px"></a>
	</td>
</tr>
</table>
	
<a id=batch></a>
<h4>Execute option</h4>

<ttc>All Pairs for Group 1</ttc>: All conditions in Group 1 will be paired for DE analysis. For example, if Rhiz, Root, and Stem are selected, then DE columns RhRo, RhSt and RoSt will be computed.

<p><ttc>Group 1 - Group 2</ttc>: shown above.
<p><ttc>All Pairs from File</ttc>: A file will be read where the first column is Group 1, the second is Group 2 and the third is the column name. All entries will be computed and added to the database.
<p><ttc>P-values from File</ttc>: If your desired method to compute p-value is not written in R, then you can compute them
outside of the TCW environment and import them with this option.
<a id=R></a>
<h4>R-scripts</h4>
A script of R commands can be provided, which uses the variables created by 
<ttp>runDE</ttp> to compute DE and GO enrichment. <ttp>RunDE</ttp> 
writes the data into the R environment (e.g. the count data is written to a matrix called 
<ttx>countData</ttx>), executes the R script using the <ttx>source</ttx> command, and 
reads the results from the variable called <ttx>results</ttx>.  

<p>The <tt>edgeRglm.R</tt> and
<tt>goSeqBH.R</tt> scripts are supplied as example. For GO enrichment, the <tt>goSeqBH.R</tt> script
is provided; for the demo only, the <tt>goSeqNoFDR.R</tt> if provided.
<p>

<a id=refs></a>
<table style="width: 100%"><tr><td style="text-align: left">
<h3>References</h3>
<td style="text-align: right"><a href="#top">Go to top</a></td></tr></table>

<ol>
<li><ttp>edgeR</ttp> - Robinson MD, McCarthy DJ, Smyth GK (2010) edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics 26: 139-140.
<li><ttp>DESeq2</ttp> - Love MI, Huber W and Anders S (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology 15:550.
<li><ttp>GOSeq</ttp> - Young MD, Wakefield MJ, Smyth GK, Oshlack A (2010) Gene ontology analysis for RNA-seq: accounting for selection bias. Genome Biology 11: R14.
</ol>
<!--#include virtual="../../ssi/end.html" -->
</body>
</html>

