<html>
<title>NnR overview</title>
<body>
<center>
<h2>TCW overview for NnR </h2>
<table width=700 border=1><tr><td>
<pre>
Project:  NnR   #Seqs: 26,685   #Hits: 1,085,108   #GOs: 24,988    TPM  Seq-DE  GO-Enrich 
 
14-Jun-21 Build Database  sequences loaded from external source
05-Jul-21 Last Annotation with sTCW v3.2.6

INPUT
   Counts: 
      SEQID  ID            SIZE  #REPS  
      NnR    Tip    299,383,735      1  
      NnR    Zone   301,658,446      1  
      NnR    Rhiz   153,385,980      5  
      NnR    Root   146,652,269      5  
      NnR    Stem   146,995,438      5  
      NnR    OLeaf  140,444,093      5  

   Sequences: 
      SEQID    SIZE  TITLE         SPECIES           SOURCE  AVG-len  MED-len  
      NnR    26,685  Sacred lotus  Nelumbo nucifera  NCBI      1,467    1,229  

ANNOTATIONS
   Hit Statistics:
      Sequences with hits     25,972  (97.3%)     Bases covered by hit  28,919,385  (73.8%)  
      Unique hits          1,085,108              Total bases           39,167,358           
      Total sequence hits  2,255,922                                                         

   annoDBs (Annotation databases): 12   (see Legend below)
      ANNODB             ONLY    BITS    ANNO  UNIQUE   TOTAL   AVG  HIT-SEQ (#SEQ)  BEST  AVG  COVER  COVER  
                                                               %SIM                  HIT  %SIM   >=50   >=90  
      SP-plants            11      36      55  27,395  240.0k  49.3  19,737 (74.0%)       60.8  55.7%   1.8%  
      SP-fungi              0       0       0  13,844  77,928  38.2   8,300 (31.1%)       40.1  13.5%   0.3%  
      SP-viruses            0       0       0     517   8,922  32.0   1,671 ( 6.3%)       34.6   1.4%     0%  
      SP-bacteria           0       0       0  27,984  68,005  38.7   5,083 (19.0%)       38.0  11.5%  <0.1%  
      SP-invertebrates      0       0       1   8,998  70,860  39.2   8,233 (30.9%)       42.3  16.0%   0.4%  
      SP-fullSubset         0       0       0  28,261  130.6k  39.9   9,683 (36.3%)       41.8  15.0%   0.3%  
      TR-plants         2,980  25,280  25,043  455.4k  635.0k  69.1  25,959 (97.3%)       92.4  79.9%  54.7%  
      TR-fungi              0       4       7  173.5k  297.5k  40.5  14,218 (53.3%)       40.9  14.6%   0.2%  
      TR-viruses            0       0       0  10,313  57,049  34.7   4,592 (17.2%)       36.7   6.4%  <0.1%  
      TR-bacteria           0       2      23  130.9k  232.6k  40.6  13,188 (49.4%)       46.1  26.6%   1.1%  
      TR-invertebrates      0       1       4  172.7k  323.4k  41.6  16,405 (61.5%)       43.9  18.7%   0.6%  
      TF-PlantTFDB          2     649     839  35,322  114.0k  49.4  12,968 (48.6%)       57.0  21.6%  11.3%  

   Top 15 species from total: 17,358
      SPECIES (25 char)        BITS    ANNO   TOTAL     SPECIES                    BITS   ANNO    TOTAL  
      Nelumbo nucifera       24,365  20,699  65,112     Hevea brasiliensis           29     25    4,727  
      Macleaya cordata          110     847  17,588     Actinidia rufa               26     79    6,015  
      Vitis vinifera             97     303  35,804     Thalictrum thalictroides     24    416    8,244  
      Davidia involucrata        64     288  27,969     Fagus sylvatica              24     43    8,206  
      Camellia sinensis          49      38  13,532     Malus domestica              23     29    8,454  
      Nyssa sinensis             47     120  16,850     Citrus unshiu                23     15    4,783  
      Cinnamomum micranthum      45     247   9,565     Kingdonia uniflora           23      3    6,024  
      Aquilegia coerulea         32      57  15,397     Other                       991  2,763  2007.7k  

   Gene Ontology Statistics:
      Unique GOs          24,988              Unique hits with GOs  849,012  (78.2%)  
      Sequences with GOs  24,756  (92.8%)     Seq best hit has GOs   20,253  (75.9%)  
      Has goslim_plant        97                                                      
                                                                                      
      biological_process  16,752  (67.0%)     is_a                   40,548           
      molecular_function   5,938  (23.8%)     part_of                 4,041           
      cellular_component   2,298  ( 9.2%)                                             

EXPRESSION
   TPM: (% of 26,685)
                    <2.0         2-5        5-10       10-50      50-100      100-1k     1k-5k     >=5k  
      Tip     7,269(27%)  2,777(10%)  2,765(10%)  9,192(34%)  2,724(10%)  1,895 (7%)   61(<1%)   2(<1%)  
      Zone    6,976(26%)  2,883(11%)  2,815(11%)  9,371(35%)  2,669(10%)  1,918 (7%)   51(<1%)   2(<1%)  
      Rhiz   10,649(40%)  2,895(11%)  2,792(10%)  6,865(26%)  1,743 (7%)  1,633 (6%)  102(<1%)   6(<1%)  
      Root    8,276(31%)  3,038(11%)  3,045(11%)  8,205(31%)  2,144 (8%)  1,899 (7%)   74(<1%)   4(<1%)  
      Stem   11,705(44%)  3,378(13%)  3,002(11%)  5,639(21%)  1,444 (5%)  1,392 (5%)  111(<1%)  14(<1%)  
      OLeaf  11,876(45%)  3,574(13%)  3,035(11%)  5,616(21%)  1,228 (5%)  1,238 (5%)   98(<1%)  20(<1%)  

   Differential expression:  (% of 26,685)
                  <1E-5        <1E-4       <0.001        <0.01        <0.05  
      RhRo   4,443(17%)   6,495(24%)   8,938(33%)  11,797(44%)  14,032(53%)  
      RhSt   3,363(13%)   6,181(23%)   9,269(35%)  12,155(46%)  14,192(53%)  
      RhOL   1,859 (7%)   4,023(15%)   7,545(28%)  11,392(43%)  13,934(52%)  
      RoSt   7,006(26%)   9,793(37%)  12,446(47%)  14,876(56%)  16,492(62%)  
      RoOL   3,467(13%)   5,754(22%)   8,642(32%)  11,939(45%)  14,380(54%)  
      StOL     330 (1%)   1,188 (4%)   3,124(12%)   6,107(23%)   9,192(34%)  
      RoSt1  7,186(27%)  10,037(38%)  12,984(49%)  16,052(60%)  18,356(69%)  
      RoSt2  7,214(27%)  10,014(38%)  12,945(49%)  15,976(60%)  18,276(68%)  

   Gene ontology enrichment: (% of 24,988)
               <1E-5     <1E-4    <0.001     <0.01       <0.05  
      RhRo    6(<1%)   27(<1%)   78(<1%)  263 (1%)    936 (4%)  
      RhSt  188 (1%)  226 (1%)  305 (1%)  440 (2%)    665 (3%)  
      RhOL   73(<1%)   94(<1%)  118(<1%)  230 (1%)    428 (2%)  
      RoSt   81(<1%)  129 (1%)  243 (1%)  538 (2%)  1,081 (4%)  
      RoOL   49(<1%)   66(<1%)   98(<1%)  213 (1%)    484 (2%)  
      StOL  103(<1%)  119(<1%)  152 (1%)  219 (1%)    335 (1%)  

SEQUENCES
   Sequence lengths:
       <=100     101-500    501-1000   1001-2000   2001-3000  3001-4000  4001-5000    >5000  
      1(<1%)  4,584(17%)  6,214(23%)  9,452(35%)  4,222(16%)  1,388(5%)    507(2%)  317(1%)  

   Quality:
      Sequences with #n>0:   1,328  ( 5.0%)  
      Sequences with #n>10:     11  (<0.1%)  

   ORF lengths:
         <=100     101-500    501-1000   1001-2000  2001-3000  3001-4000  4001-5000     >5000  
      104(<1%)  6,860(26%)  7,842(29%)  8,640(32%)  2,263(8%)    662(2%)    184(1%)  130(<1%)  

   Total ORFs: 26,685 (100.0%)
      Has Hit            25,970  (97.3%)    Average ORF len            1105                                    
      Is Longest ORF     25,537  (95.7%)    ORF>=300         23,902  (89.6%)    MultiFrame     1,167   (4.4%)  
      Markov Best Score  25,775  (96.6%)    Has Start&Stop   21,642  (81.1%)    >=9 Ns in ORF      2    (<1%)  
      All of the above   24,648  (92.4%)    Has Start|Stop   25,565  (95.8%)    Stops in Hit     721   (2.7%)  
                        

   GC Content: 44.78%
      Pos1  17.1%              5UTR    CDS   3UTR              5UTR      CDS    3UTR  
      Pos2  13.6%    %GC      43.12  45.35  37.58    Length  4,229k  29,476k  5,463k  
      Pos3  14.6%    CpG-O/E   0.70   0.52   0.43    AvgLen   158.5   1104.6   204.7  
                                    

-------------------------------------------------------------------
PROCESSING INFORMATION:
   AnnoDB Files:
      Type  Taxo           FILE                                DB DATE    ADD DATE   EXECUTE                
      sp    plants         uniprot_sprot_plants.fasta          07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      sp    fungi          uniprot_sprot_fungi.fasta           07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      sp    viruses        uniprot_sprot_viruses.fasta         07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      sp    bacteria       uniprot_sprot_bacteria.fasta        07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      sp    invertebrates  uniprot_sprot_invertebrates.fasta   07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      sp    fullSubset     uniprot_sprot_fullSubset.fasta      07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      tr    plants         uniprot_trembl_plants.fasta.gz      07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      tr    fungi          uniprot_trembl_fungi.fasta.gz       07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      tr    viruses        uniprot_trembl_viruses.fasta        07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      tr    bacteria       uniprot_trembl_bacteria.fasta.gz    07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      tr    invertebrates  uniprot_trembl_invertebrates.fasta  07-Jun-21  14-Jun-21  diamond  --max-hsps 1  
      tf    PlantTFDB      PlantTFDB_pep.fa                    27-Aug-20  14-Jun-21  diamond  --max-hsps 1  

   Gene Ontology: go-basic.obo-May2021  GOdb: go_Jun2021  [GOs added with sTCW v3.2.5]
   GO Slim: goslim_plant

   ORF finder:
      Use ATG only for start site
      Rule 1: Use Good hit: E-value <=1E-10 or Sim >= 20%
      Rule 2: Use longest ORF if Log Len Ratio > 0.5
      Rule 3: Use best Markov score
              Train using best hits
      Good coverage: Hit overlap >= 95% with Sim 60% (internal params)

   DE (Differential Expression) computation: 
      Column       Method               Conditions
      RhRo         edgeRglm.R           Rhiz : Root 
      RhSt         edgeRglm.R           Rhiz : Stem 
      RhOL         edgeRglm.R           Rhiz : OLeaf 
      RoSt         edgeRglm.R           Root : Stem 
      RoOL         edgeRglm.R           Root : OLeaf 
      StOL         edgeRglm.R           Stem : OLeaf 
      RoSt1        edgeRglm.R           Root : Stem 
      RoSt2        edgeRglm.R           Root : Stem 

   GO enrichment computation: 
      Column       Method               Cutoff
      RhRo         goSeqBH.R            0.0500
      RhSt         goSeqBH.R            0.0500
      RhOL         goSeqBH.R            0.0500
      RoSt         goSeqBH.R            0.0500
      RoOL         goSeqBH.R            0.0500
      StOL         goSeqBH.R            0.0500

-------------------------------------------------------------------
LEGEND:
   annoDB:
      ANNODB    is DBTYPE-TAXO, which is the DBtype and taxonomy
      ONLY      #Seqs that hit the annoDB and no others
      BITS      #Seqs with the overall best bitscore from the annoDB
      ANNO      #Seqs with the overall best annotation from the annoDB 
      UNIQUE    #Unique hits to the annoDB
      TOTAL     #Total seq-hit pairs for the annoDB
      AVG %SIM  Average percent similarity of the total seq-hit pairs
      HIT-SEQ   Percent of #Seqs that have at least one hit from the annoDB
      BEST HIT  The following columns refer to the best hit (Rank=1):
         AVG %SIM  Average percent similarity of the best hit seq-hit pairs
         Cover>=N  Percent of HIT-SEQ where the best hit has similarity>=N% and hit coverage>=N%

   #Seqs is listed at top of overview
   Best Annotation:
      Descriptions may not contain words such as 'uncharacterized protein'

</pre>
</body>
</html>
