<html>
<title>Nn overview</title>
<body>
<center>
<h2>TCW overview for Nn </h2>
<table width=700 border=1><tr><td>
<pre>
Project:  Nn   #Seqs: 26,685   #Hits: 593,083   #GOs: 24,582    TPM  Seq-DE  GO-Enrich   Pairs
 
06-Nov-20 Build Database  sequences loaded from external source
28-Mar-21 Last Annotation with sTCW v3.2.0    Updated with v3.2.2

INPUT
   Counts: 
      SEQID  ID            SIZE  #REPS  
      NnR    Tip    299,383,735      1  
      NnR    Zone   301,658,446      1  
      NnR    Rhiz   153,385,980      5  
      NnR    Root   146,652,269      5  
      NnR    Stem   146,995,438      5  
      NnR    OLeaf  140,444,093      5  

   Sequences: 
      SEQID    SIZE  TITLE         SPECIES           SOURCE  AVG-len  MED-len  
      NnR    26,685  Sacred lotus  Nelumbo nucifera  NCBI      1,467    1,229  

ANNOTATIONS
   Hit Statistics:
      Sequences with hits     25,970  (97.3%)     Bases covered by hit  28,913,634  (73.8%)  
      Unique hits            593,083              Total bases           39,167,358           
      Total sequence hits  1,344,487                                                         

   annoDBs (Annotation databases): 8   (see Legend below)
      ANNODB             ONLY    BITS    ANNO  UNIQUE   TOTAL   AVG  HIT-SEQ (#SEQ)  BEST  AVG  COVER  COVER  
                                                               %SIM                  HIT  %SIM   >=50   >=90  
      SP-plants            11      37      56  27,301  239.7k  49.3  19,711 (73.9%)       60.9  55.7%   1.8%  
      SP-invertebrates      0       0       1   8,963  70,718  39.2   8,229 (30.8%)       42.3  16.0%   0.4%  
      SP-fungi              0       1       1  13,824  77,828  38.1   8,299 (31.1%)       40.0  13.5%   0.3%  
      SP-viruses            0       0       0     517   8,922  32.0   1,671 ( 6.3%)       34.6   1.4%     0%  
      SP-bacteria           0       1       1  27,967  67,926  38.7   5,077 (19.0%)       38.0  11.5%  <0.1%  
      SP-fullSubset         0       0       0  28,247  130.6k  39.9   9,683 (36.3%)       41.8  15.0%   0.3%  
      TR-plants         3,907  25,282  25,062  450.9k  634.7k  69.0  25,957 (97.3%)       92.4  79.9%  54.7%  
      TF-tf                 2     649     849  35,322  114.0k  49.4  12,968 (48.6%)       57.0  21.6%  11.3%  

   Top 15 species from total: 4,872
      SPECIES (25 char)        BITS    ANNO   TOTAL     SPECIES              BITS   ANNO    TOTAL  
      Nelumbo nucifera       24,396  20,765  65,502     Fagus sylvatica        28     47    8,970  
      Macleaya cordata          116     873  18,111     Hevea brasiliensis     28     24    5,192  
      Vitis vinifera            105     321  37,684     Malus domestica        26     29    8,903  
      Davidia involucrata        70     298  29,036     Citrus unshiu          23     14    5,347  
      Nyssa sinensis             49     124  17,698     Prunus dulcis          20    250    8,222  
      Cinnamomum micranthum      47     282  10,191     Durio zibethinus       20     37   12,465  
      Aquilegia coerulea         39      76  16,225     Ananas comosus         20     30    4,475  
      Camellia sinensis          35      38   7,679     Other                 948  2,762  1088.8k  

   Gene Ontology Statistics:
      Unique GOs          24,582              Unique hits with GOs  410,922  (69.3%)  
      Sequences with GOs  24,670  (92.4%)     Seq best hit has GOs   20,283  (76.0%)  
      Has goslim_plant        97                                                      
                                                                                      
      biological_process  16,666  (67.8%)     is_a                   39,994           
      molecular_function   5,640  (22.9%)     part_of                 4,057           
      cellular_component   2,276  ( 9.3%)                                             

EXPRESSION
   TPM: (% of 26,685)
                    <2.0         2-5        5-10       10-50      50-100      100-1k     1k-5k     >=5k  
      Tip     7,269(27%)  2,777(10%)  2,765(10%)  9,192(34%)  2,724(10%)  1,895 (7%)   61(<1%)   2(<1%)  
      Zone    6,976(26%)  2,883(11%)  2,815(11%)  9,371(35%)  2,669(10%)  1,918 (7%)   51(<1%)   2(<1%)  
      Rhiz   10,649(40%)  2,895(11%)  2,792(10%)  6,865(26%)  1,743 (7%)  1,633 (6%)  102(<1%)   6(<1%)  
      Root    8,276(31%)  3,038(11%)  3,045(11%)  8,205(31%)  2,144 (8%)  1,899 (7%)   74(<1%)   4(<1%)  
      Stem   11,705(44%)  3,378(13%)  3,002(11%)  5,639(21%)  1,444 (5%)  1,392 (5%)  111(<1%)  14(<1%)  
      OLeaf  11,876(45%)  3,574(13%)  3,035(11%)  5,616(21%)  1,228 (5%)  1,238 (5%)   98(<1%)  20(<1%)  

   Differential expression: (% of 26,685)
                 <1E-5       <1E-4       <0.001        <0.01        <0.05  
      RzRo  4,443(17%)  6,495(24%)   8,938(33%)  11,797(44%)  14,032(53%)  
      RzSt  3,363(13%)  6,181(23%)   9,269(35%)  12,155(46%)  14,192(53%)  
      RzOL  1,859 (7%)  4,023(15%)   7,545(28%)  11,392(43%)  13,934(52%)  
      RoSt  7,006(26%)  9,793(37%)  12,446(47%)  14,876(56%)  16,492(62%)  
      RoOL  3,467(13%)  5,754(22%)   8,642(32%)  11,939(45%)  14,380(54%)  
      StOL    330 (1%)  1,188 (4%)   3,124(12%)   6,107(23%)   9,192(34%)  

   GOs enrichment: (% of 24,582)
               <1E-5     <1E-4    <0.001     <0.01       <0.05  
      RzRo    7(<1%)   28(<1%)   61(<1%)  236 (1%)    961 (4%)  
      RzSt  179 (1%)  225 (1%)  283 (1%)  415 (2%)    643 (3%)  
      RzOL   64(<1%)   87(<1%)  123 (1%)  221 (1%)    429 (2%)  
      RoSt   67(<1%)  119(<1%)  225 (1%)  512 (2%)  1,044 (4%)  
      RoOL   49(<1%)   63(<1%)  100(<1%)  210 (1%)    495 (2%)  
      StOL  102(<1%)  114(<1%)  147 (1%)  201 (1%)    331 (1%)  

SEQUENCES
   Sequence lengths:
       <=100     101-500    501-1000   1001-2000   2001-3000  3001-4000  4001-5000    >5000  
      1(<1%)  4,584(17%)  6,214(23%)  9,452(35%)  4,222(16%)  1,388(5%)    507(2%)  317(1%)  

   Quality:
      Sequences with #n>0:   1,328  ( 5.0%)  
      Sequences with #n>10:     11  (<0.1%)  

   ORF lengths:
         <=100     101-500    501-1000   1001-2000  2001-3000  3001-4000  4001-5000     >5000  
      105(<1%)  6,859(26%)  7,843(29%)  8,640(32%)  2,264(8%)    661(2%)    183(1%)  130(<1%)  

   Total ORFs: 26,685 (100.0%)
      Has Hit            25,968 (97.3%)    ORF>=300         23,902 (89.6%)    MultiFrame     1,060  (4.0%)  
      Is Longest ORF     25,535 (95.7%)    Has Start&Stop   21,646 (81.1%)    >=9 Ns in ORF      2   (<1%)  
      Markov Best Score  25,777 (96.6%)    Has Start|Stop   25,568 (95.8%)    Stops in Hit     718  (2.7%)  
      All of the above   24,645 (92.4%)    Average ORF len           1104         

   GC Content: 44.78%
      Pos1  17.1%              5UTR    CDS   3UTR              5UTR      CDS    3UTR  
      Pos2  13.6%    %GC      43.12  45.35  37.58    Length  4,235k  29,470k  5,462k  
      Pos3  14.6%    CpG-O/E   0.70   0.52   0.43    AvgLen   158.7   1104.4   204.7  
                                    
   Similar pairs: 100
      Translated nucleotide  100  


-------------------------------------------------------------------
PROCESSING INFORMATION:
   AnnoDB Files:
      Type  Taxo           FILE                                  DB DATE    ADD DATE   EXECUTE                 
      sp    plants         uniprot_sprot_plants.fasta.gz         17-Feb-21  18-Mar-21  Load Nn_SPpla.dmnd.tab  
      sp    invertebrates  uniprot_sprot_invertebrates.fasta.gz  17-Feb-21  18-Mar-21  Load Nn_SPinv.dmnd.tab  
      sp    fungi          uniprot_sprot_fungi.fasta.gz          17-Feb-21  18-Mar-21  Load Nn_SPfun.dmnd.tab  
      sp    viruses        uniprot_sprot_viruses.fasta.gz        17-Feb-21  18-Mar-21  Load Nn_SPvir.dmnd.tab  
      sp    bacteria       uniprot_sprot_bacteria.fasta.gz       17-Feb-21  18-Mar-21  Load Nn_SPbac.dmnd.tab  
      sp    fullSubset     uniprot_sprot_fullSubset.fasta.gz     17-Feb-21  18-Mar-21  Load Nn_SPful.dmnd.tab  
      tr    plants         uniprot_trembl_plants.fasta.gz        17-Feb-21  18-Mar-21  Load Nn_TRpla.dmnd.tab  
      tf    tf             PlantTFDB_TCW.fa                      25-Jan-21  18-Mar-21  Load Nn_TFtf.dmnd.tab   

   Gene Ontology: go-basic.obo-Feb2021  GOdb: go_Mar2021  [GOs added with sTCW v3.2.0]
   GO Slim: goslim_plant

   ORF finder:
      Use ATG only for start site
      Rule 1: Use Good hit: E-value <=1E-10 or Sim >= 20%
      Rule 2: Use longest ORF if Log Len Ratio > 0.5
      Rule 3: Use best Markov score
              Train using best hits
      Good coverage: Hit overlap >= 95% with Sim 60% (internal params)

   DE (Differential Expression) computation: 
      Column       Method               Conditions
      RzRo         source(edgeRglm.R)   Rhiz : Root 
      RzSt         source(edgeRglm.R)   Rhiz : Stem 
      RzOL         source(edgeRglm.R)   Rhiz : OLeaf 
      RoSt         source(edgeRglm.R)   Root : Stem 
      RoOL         source(edgeRglm.R)   Root : OLeaf 
      StOL         source(edgeRglm.R)   Stem : OLeaf 

   GO enrichment computation: 
      Column       Method               Cutoff
      RzRo         goSeqBH.R            0.0500
      RzSt         goSeqBH.R            0.0500
      RzOL         goSeqBH.R            0.0500
      RoSt         goSeqBH.R            0.0500
      RoOL         goSeqBH.R            0.0500
      StOL         goSeqBH.R            0.0500

-------------------------------------------------------------------
LEGEND:
   annoDB:
      ANNODB    is DBTYPE-TAXO, which is the DBtype and taxonomy
      ONLY      #Seqs that hit the annoDB and no others
      BITS      #Seqs with the overall best bitscore from the annoDB
      ANNO      #Seqs with the overall best annotation from the annoDB 
      UNIQUE    #Unique hits to the annoDB
      TOTAL     #Total seq-hit pairs for the annoDB
      AVG %SIM  Average percent similarity of the total seq-hit pairs
      HIT-SEQ   Percent of #Seqs that have at least one hit from the annoDB
      BEST HIT  The following columns refer to the best hit (Rank=1):
         AVG %SIM  Average percent similarity of the best hit seq-hit pairs
         Cover>=N  Percent of HIT-SEQ where the best hit has similarity>=N% and hit coverage>=N%

   #Seqs is listed at top of overview
   Best Annotation:
      Descriptions may not contain words such as 'uncharacterized protein'

</pre>
</body>
</html>
